
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Section 1 &#8212; ISMB Tutorial Book</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b76e3c8a" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=384b581d" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=efea14e4"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=36754332"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Section1/Section1';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="ISMB Tutorial Book - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="ISMB Tutorial Book - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    An Introduction to Multi Modal Analysis using Networks
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../GettingStarted.html">Getting Started</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../Section1_intro.html">Introduction to Network Basics</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="Part1.html"><font color="darkblue">Retrieval of Protein-Protein Interactions from STRING</font></a></li>








<li class="toctree-l2"><a class="reference internal" href="Part2.html">Section 2</a></li>







</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../Section2_intro.html">Introduction to Network Analysis</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../Section2/Part1.html">Section 2: Network Analysis</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../Section3_intro.html">Introduction to Graph Neural Networks</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part1.html"><font color="darkblue"> Part 1 : Message Passing </font></a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part2.html"><font color="darkblue"> Part 2 : Problem Formulation / Patient Similarity Network Generation </font></a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part3.html"><font color="darkblue"> Part 3 : Smoking Network Classification using Graph Convolutional Networks</font></a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part4.html"><font color="darkblue"> Part 4 : Similarity Network Fusion with Graph Convolutional Networks</font></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../ConceptGlossary.html">Concept Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../functions_definitions.html">Custom Function Definitions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/Barry8197/ISMB_Test" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/Barry8197/ISMB_Test/issues/new?title=Issue%20on%20page%20%2FSection1/Section1.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/Section1/Section1.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Section 1</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Section 1</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#structuring-biomedical-data-using-networks-workshop-network-generation">Structuring Biomedical Data Using Networks  workshop - Network Generation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#table-of-contents">Table of Contents</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#libraries">Libraries</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data">Data</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tcga-luad-project">TCGA-LUAD Project</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata-analysis">Metadata analysis</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#expression-data-analysis">Expression data analysis</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data-filtering-and-preprocessing">Data Filtering and Preprocessing</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#thresholding-based-on-gene-expression-levels">Thresholding Based on Gene Expression Levels</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-retention-at-various-thresholds">Gene Retention at Various Thresholds</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#thresholding-based-on-gene-expression-variance">Thresholding Based on Gene Expression Variance</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#instead-of-gene-expression">instead of Gene Expression</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-retention-at-different-variance-levels">Gene Retention at Different Variance Levels</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-symbol-conversion-ensg-to-hgcn">Gene Symbol Conversion: ENSG to HGCN</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#checking-for-missing-data-and-outliers">Checking for Missing Data and Outliers</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-heterogeneity">Data heterogeneity</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#correlation-matrix-calculation">Correlation Matrix Calculation</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#consider-loading-in-precomputed-correlation-matrices-and-skipping-the-next-cell">Consider loading in precomputed correlation matrices and skipping the next cell</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-check-grouping-ordering">TODO: check grouping, ordering ?</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#graph-construction-from-correlation-matrices">Graph Construction from Correlation Matrices</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-list-comprehension-for-graph-generation">TODO: list comprehension for Graph generation ??</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#show-densities-what-treshold-is-useful-to-get-rid-of-the-fully-connected-network">show densities, what treshold is useful to get rid of the fully connected network</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#network-cleaning">Network Cleaning</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#sparsification-methods">Sparsification methods</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-show-connected-components-of-the-plot-above">TODO: show connected components of the plot above</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#interactive-visualization">Interactive Visualization</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-add-hover-info-to-plotly-plot-gene-info">TODO: add hover info to plotly plot (gene info)</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-introduce-clustering-kmeans-etc">TODO: introduce clustering kmeans. etc</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#storing-generated-networks">Storing generated networks</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#patient-similarity-network">Patient similarity network</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#dnameth-pn">DNAmeth PN</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#setting-up-pip-network-using-string">Setting up PIP network using String</a></li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="section-1">
<h1>Section 1<a class="headerlink" href="#section-1" title="Link to this heading">#</a></h1>
<section id="structuring-biomedical-data-using-networks-workshop-network-generation">
<h2>Structuring Biomedical Data Using Networks  workshop - Network Generation<a class="headerlink" href="#structuring-biomedical-data-using-networks-workshop-network-generation" title="Link to this heading">#</a></h2>
</section>
<section id="table-of-contents">
<h2>Table of Contents<a class="headerlink" href="#table-of-contents" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p><strong>Biomedical Networks</strong></p>
<ul class="simple">
<li><p>1.1. Libraries</p></li>
<li><p>1.2. TCGA-LUAD project</p></li>
<li><p>1.3. Metadata Analysis</p></li>
<li><p>1.4. Expression Data Analysis</p></li>
</ul>
</li>
<li><p><strong>Data Filtering and Preprocessing</strong></p>
<ul class="simple">
<li><p>2.1. Thresholding Based on Gene Expression Levels</p></li>
<li><p>2.2. Gene Retention at Various Thresholds</p></li>
<li><p>2.3. Thresholding Based on Gene Expression Variance</p></li>
<li><p>2.4. Gene Retention at Different Variance Levels</p></li>
</ul>
</li>
<li><p><strong>Gene Symbol Conversion</strong></p>
<ul class="simple">
<li><p>3.1. Checking for Missing Data and Outliers</p></li>
<li><p>3.2. Addressing Data Heterogeneity</p></li>
</ul>
</li>
<li><p><strong>Correlation Matrix Calculation</strong></p></li>
<li><p><strong>Graph Construction from Correlation Matrices</strong></p></li>
<li><p><strong>Network Cleaning</strong></p></li>
<li><p><strong>Sparsification Methods for Networks</strong></p></li>
<li><p><strong>Interactive Network Visualization</strong></p></li>
<li><p><strong>Storing and Managing Generated Networks</strong></p></li>
<li><p><strong>Patient Similarity Network Construction</strong></p></li>
<li><p><strong>DNA Methylation Network Analysis</strong></p></li>
<li><p><strong>Setting up Protein Interaction Network Using STRING Database</strong></p></li>
</ol>
</section>
<section id="libraries">
<h2>Libraries<a class="headerlink" href="#libraries" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># standard libraries</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pickle</span>

<span class="c1"># scientific and data manipulation libraries</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="kn">from</span> <span class="nn">sklearn.feature_selection</span> <span class="kn">import</span> <span class="n">mutual_info_regression</span>
<span class="kn">from</span> <span class="nn">astropy.stats</span> <span class="kn">import</span> <span class="n">median_absolute_deviation</span>
<span class="kn">import</span> <span class="nn">mygene</span>
<span class="kn">import</span> <span class="nn">astropy</span>

<span class="c1"># graph and network libraries</span>
<span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
<span class="kn">from</span> <span class="nn">pyvis.network</span> <span class="kn">import</span> <span class="n">Network</span>

<span class="c1"># visualization libraries</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">plotly.graph_objects</span> <span class="k">as</span> <span class="nn">go</span>
<span class="kn">import</span> <span class="nn">plotly.io</span> <span class="k">as</span> <span class="nn">pio</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="data">
<h1>Data<a class="headerlink" href="#data" title="Link to this heading">#</a></h1>
<section id="tcga-luad-project">
<h2>TCGA-LUAD Project<a class="headerlink" href="#tcga-luad-project" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p><strong>Title</strong>: The Cancer Genome Atlas Lung Adenocarcinoma (TCGA-LUAD)</p></li>
<li><p><strong>Main Focus</strong>: Study of lung adenocarcinoma (a common type of lung cancer)</p></li>
<li><p><strong>Data Collected</strong>: Comprehensive genomic, epigenomic, transcriptomic, and proteomic data from lung adenocarcinoma samples</p></li>
<li><p><strong>Disease Types</strong>:</p>
<ul>
<li><p>Acinar Cell Neoplasms</p></li>
<li><p>Adenomas and Adenocarcinomas</p></li>
<li><p>Cystic, Mucinous, and Serous Neoplasms</p></li>
</ul>
</li>
<li><p><strong>Number of Cases</strong>: 585</p></li>
<li><p><strong>Data Accessibility</strong>: Available on the NIH-GDC Data Portal</p></li>
<li><p><strong>Link</strong>: <a class="reference external" href="https://portal.gdc.cancer.gov/projects/TCGA-LUAD">TCGA-LUAD Project Page</a></p></li>
</ul>
<p>Our main focus in this notebook is on RNA-Seq based, already preprocessed gene expression data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># main data directory for the project</span>

<span class="n">data_dir</span> <span class="o">=</span> <span class="s1">&#39;../../ismb_data&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># load the data from the pickle file</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s2">&quot;ISMB_TCGA_GE.pkl&quot;</span><span class="p">),</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>dict_keys([&#39;datExpr&#39;, &#39;datMeta&#39;])
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="metadata-analysis">
<h1>Metadata analysis<a class="headerlink" href="#metadata-analysis" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;datMeta&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>patient</th>
      <th>race</th>
      <th>gender</th>
      <th>sample_type</th>
      <th>cigarettes_per_day</th>
      <th>Smoked</th>
      <th>sizeFactor</th>
      <th>replaceable</th>
    </tr>
    <tr>
      <th>_row</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TCGA-38-7271</th>
      <td>TCGA-38-7271</td>
      <td>white</td>
      <td>female</td>
      <td>Primary Tumor</td>
      <td>1.3699</td>
      <td>Smoker</td>
      <td>0.5841</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-55-7914</th>
      <td>TCGA-55-7914</td>
      <td>white</td>
      <td>female</td>
      <td>Primary Tumor</td>
      <td>0.274</td>
      <td>Smoker</td>
      <td>0.9873</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-95-7043</th>
      <td>TCGA-95-7043</td>
      <td>white</td>
      <td>female</td>
      <td>Primary Tumor</td>
      <td>2.1918</td>
      <td>Smoker</td>
      <td>0.5439</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-73-4658</th>
      <td>TCGA-73-4658</td>
      <td>white</td>
      <td>female</td>
      <td>Primary Tumor</td>
      <td>1.3699</td>
      <td>Smoker</td>
      <td>0.7715</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-86-8076</th>
      <td>TCGA-86-8076</td>
      <td>white</td>
      <td>male</td>
      <td>Primary Tumor</td>
      <td>0</td>
      <td>Never</td>
      <td>1.313</td>
      <td>True</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TCGA-86-8073</th>
      <td>TCGA-86-8073</td>
      <td>white</td>
      <td>male</td>
      <td>Primary Tumor</td>
      <td>2.1918</td>
      <td>Smoker</td>
      <td>1.9733</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-MN-A4N4</th>
      <td>TCGA-MN-A4N4</td>
      <td>white</td>
      <td>male</td>
      <td>Primary Tumor</td>
      <td>0</td>
      <td>Never</td>
      <td>1.0464</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-53-7626</th>
      <td>TCGA-53-7626</td>
      <td>white</td>
      <td>female</td>
      <td>Primary Tumor</td>
      <td>1.9178</td>
      <td>Smoker</td>
      <td>1.7208</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-44-A47G</th>
      <td>TCGA-44-A47G</td>
      <td>white</td>
      <td>female</td>
      <td>Primary Tumor</td>
      <td>1.5342</td>
      <td>Smoker</td>
      <td>0.8926</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TCGA-55-6969</th>
      <td>TCGA-55-6969</td>
      <td>white</td>
      <td>male</td>
      <td>Primary Tumor</td>
      <td>0</td>
      <td>Never</td>
      <td>1.4732</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>498 rows × 8 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;datMeta&quot;</span><span class="p">][</span><span class="s1">&#39;sample_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([&#39;Primary Tumor&#39;, &#39;Solid Tissue Normal&#39;], dtype=object)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Set up the figure and axes for a 2-column layout</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">18</span><span class="p">,</span> <span class="mi">18</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Metadata Distributions&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Plot 1: Distribution of Race</span>
<span class="n">sns</span><span class="o">.</span><span class="n">countplot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;race&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Distribution of Race&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Race&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Count&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>

<span class="c1"># Plot 2: Gender Distribution</span>
<span class="n">sns</span><span class="o">.</span><span class="n">countplot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;gender&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;magma&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Gender Distribution&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Gender&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Count&#39;</span><span class="p">)</span>

<span class="c1"># Plot 3: Sample Type Distribution</span>
<span class="n">sns</span><span class="o">.</span><span class="n">countplot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;sample_type&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;plasma&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Sample Type Distribution&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Sample Type&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Count&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>

<span class="c1"># Plot 4: Distribution of Cigarettes Per Day</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">][</span><span class="s1">&#39;cigarettes_per_day&#39;</span><span class="p">],</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Distribution of Cigarettes Per Day&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Cigarettes Per Day&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>

<span class="c1"># Plot 5: Smoking Status by Gender</span>
<span class="n">sns</span><span class="o">.</span><span class="n">countplot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;Smoked&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;gender&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">],</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;coolwarm&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Smoking Status by Gender&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Smoking Status&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Count&#39;</span><span class="p">)</span>
<span class="n">axes</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s1">&#39;Gender&#39;</span><span class="p">)</span>

<span class="n">axes</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/skamp/.conda/envs/ismbenv/lib/python3.12/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div>
</div>
<img alt="../_images/5ce11f62a0e9f129a0d0026ba63ebf236b902ab58471d9f4bc1ec6921907dbd7.png" src="../_images/5ce11f62a0e9f129a0d0026ba63ebf236b902ab58471d9f4bc1ec6921907dbd7.png" />
</div>
</div>
</section>
<section id="expression-data-analysis">
<h1>Expression data analysis<a class="headerlink" href="#expression-data-analysis" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">expression_data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;datExpr&quot;</span><span class="p">]</span>
<span class="n">expression_data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>_row</th>
      <th>ENSG00000000003.15</th>
      <th>ENSG00000000419.13</th>
      <th>ENSG00000000457.14</th>
      <th>ENSG00000000460.17</th>
      <th>ENSG00000000938.13</th>
      <th>ENSG00000000971.16</th>
      <th>ENSG00000001036.14</th>
      <th>ENSG00000001084.13</th>
      <th>ENSG00000001167.14</th>
      <th>ENSG00000001460.18</th>
      <th>...</th>
      <th>ENSG00000288586.1</th>
      <th>ENSG00000288596.2</th>
      <th>ENSG00000288598.1</th>
      <th>ENSG00000288611.1</th>
      <th>ENSG00000288612.1</th>
      <th>ENSG00000288638.1</th>
      <th>ENSG00000288658.1</th>
      <th>ENSG00000288663.1</th>
      <th>ENSG00000288670.1</th>
      <th>ENSG00000288675.1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TCGA-38-7271</th>
      <td>11.3668</td>
      <td>10.3673</td>
      <td>9.7884</td>
      <td>8.2552</td>
      <td>10.6173</td>
      <td>12.8447</td>
      <td>10.8836</td>
      <td>10.4437</td>
      <td>10.8112</td>
      <td>9.1178</td>
      <td>...</td>
      <td>4.4411</td>
      <td>5.6909</td>
      <td>3.8172</td>
      <td>5.5378</td>
      <td>4.5573</td>
      <td>5.4255</td>
      <td>5.1694</td>
      <td>4.0073</td>
      <td>7.97</td>
      <td>4.664</td>
    </tr>
    <tr>
      <th>TCGA-55-7914</th>
      <td>11.5434</td>
      <td>10.5282</td>
      <td>9.7292</td>
      <td>7.9951</td>
      <td>8.4858</td>
      <td>10.4696</td>
      <td>10.4737</td>
      <td>7.9065</td>
      <td>11.8983</td>
      <td>8.5842</td>
      <td>...</td>
      <td>4.768</td>
      <td>8.0948</td>
      <td>4.2182</td>
      <td>3.5233</td>
      <td>4.975</td>
      <td>2.4821</td>
      <td>6.0753</td>
      <td>4.975</td>
      <td>8.8543</td>
      <td>4.3015</td>
    </tr>
    <tr>
      <th>TCGA-95-7043</th>
      <td>11.411</td>
      <td>11.2018</td>
      <td>9.4449</td>
      <td>8.3546</td>
      <td>7.3211</td>
      <td>10.7856</td>
      <td>11.6185</td>
      <td>14.1836</td>
      <td>10.8329</td>
      <td>8.8155</td>
      <td>...</td>
      <td>4.2248</td>
      <td>7.3539</td>
      <td>5.0121</td>
      <td>4.2248</td>
      <td>4.8311</td>
      <td>4.7304</td>
      <td>5.5053</td>
      <td>4.9246</td>
      <td>7.6064</td>
      <td>3.8623</td>
    </tr>
    <tr>
      <th>TCGA-73-4658</th>
      <td>12.2149</td>
      <td>10.3249</td>
      <td>9.234</td>
      <td>7.7537</td>
      <td>11.0656</td>
      <td>12.3631</td>
      <td>10.9023</td>
      <td>10.06</td>
      <td>11.2092</td>
      <td>9.0663</td>
      <td>...</td>
      <td>4.2146</td>
      <td>6.0253</td>
      <td>3.8231</td>
      <td>4.0984</td>
      <td>4.6727</td>
      <td>7.1082</td>
      <td>7.1615</td>
      <td>3.6533</td>
      <td>7.3223</td>
      <td>5.2312</td>
    </tr>
    <tr>
      <th>TCGA-86-8076</th>
      <td>11.2882</td>
      <td>10.2095</td>
      <td>9.8186</td>
      <td>7.9844</td>
      <td>10.5213</td>
      <td>13.7848</td>
      <td>11.2318</td>
      <td>10.8315</td>
      <td>10.5784</td>
      <td>9.6428</td>
      <td>...</td>
      <td>4.7844</td>
      <td>6.8675</td>
      <td>4.5078</td>
      <td>2.4821</td>
      <td>5.1468</td>
      <td>3.3897</td>
      <td>4.2833</td>
      <td>4.4553</td>
      <td>8.1011</td>
      <td>4.8254</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TCGA-86-8073</th>
      <td>10.7676</td>
      <td>10.5503</td>
      <td>9.0277</td>
      <td>8.2811</td>
      <td>9.9471</td>
      <td>11.8641</td>
      <td>11.237</td>
      <td>14.2843</td>
      <td>10.9497</td>
      <td>8.8266</td>
      <td>...</td>
      <td>4.2611</td>
      <td>7.157</td>
      <td>4.2187</td>
      <td>5.2566</td>
      <td>5.276</td>
      <td>2.4821</td>
      <td>6.695</td>
      <td>5.1967</td>
      <td>7.8098</td>
      <td>4.3418</td>
    </tr>
    <tr>
      <th>TCGA-MN-A4N4</th>
      <td>11.656</td>
      <td>11.0739</td>
      <td>9.5578</td>
      <td>8.7883</td>
      <td>8.9532</td>
      <td>11.0791</td>
      <td>10.5202</td>
      <td>11.331</td>
      <td>11.3659</td>
      <td>8.8701</td>
      <td>...</td>
      <td>4.3317</td>
      <td>7.6037</td>
      <td>5.0518</td>
      <td>7.4063</td>
      <td>6.5835</td>
      <td>3.0746</td>
      <td>5.9828</td>
      <td>4.9634</td>
      <td>7.9208</td>
      <td>5.1745</td>
    </tr>
    <tr>
      <th>TCGA-53-7626</th>
      <td>10.8912</td>
      <td>10.1742</td>
      <td>9.6307</td>
      <td>8.0179</td>
      <td>10.6328</td>
      <td>11.247</td>
      <td>10.4815</td>
      <td>9.8428</td>
      <td>10.5795</td>
      <td>9.86</td>
      <td>...</td>
      <td>5.6458</td>
      <td>6.6984</td>
      <td>4.646</td>
      <td>3.1345</td>
      <td>4.8113</td>
      <td>3.8962</td>
      <td>5.9768</td>
      <td>4.2816</td>
      <td>8.4192</td>
      <td>4.4566</td>
    </tr>
    <tr>
      <th>TCGA-44-A47G</th>
      <td>11.32</td>
      <td>9.9137</td>
      <td>9.3312</td>
      <td>7.8703</td>
      <td>11.3029</td>
      <td>12.8687</td>
      <td>11.2659</td>
      <td>12.4677</td>
      <td>10.9404</td>
      <td>9.4188</td>
      <td>...</td>
      <td>4.6125</td>
      <td>8.0939</td>
      <td>5.1263</td>
      <td>2.4821</td>
      <td>5.2619</td>
      <td>5.0279</td>
      <td>8.793</td>
      <td>5.173</td>
      <td>7.951</td>
      <td>4.465</td>
    </tr>
    <tr>
      <th>TCGA-55-6969</th>
      <td>10.5764</td>
      <td>10.9022</td>
      <td>9.3425</td>
      <td>8.8147</td>
      <td>9.6115</td>
      <td>12.3155</td>
      <td>11.216</td>
      <td>11.1273</td>
      <td>10.311</td>
      <td>9.123</td>
      <td>...</td>
      <td>5.9603</td>
      <td>7.2623</td>
      <td>4.5931</td>
      <td>4.1326</td>
      <td>4.5495</td>
      <td>3.1863</td>
      <td>3.9293</td>
      <td>4.828</td>
      <td>7.5868</td>
      <td>4.1326</td>
    </tr>
  </tbody>
</table>
<p>498 rows × 22637 columns</p>
</div></div></div>
</div>
<p>498 rows × 22637 columns is a quite large matrix, thus we have to consider reducing the dimensionality.</p>
</section>
<section id="data-filtering-and-preprocessing">
<h1>Data Filtering and Preprocessing<a class="headerlink" href="#data-filtering-and-preprocessing" title="Link to this heading">#</a></h1>
<ul class="simple">
<li><p>Filter by Variance: Select genes with high variance across samples, as low variance genes might not contribute significantly to the analysis.</p></li>
<li><p>Filter by Mean Expression: Select genes with high mean expression levels</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Ensure all columns are numeric</span>
<span class="n">expression_data</span> <span class="o">=</span> <span class="n">expression_data</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">,</span> <span class="n">errors</span><span class="o">=</span><span class="s1">&#39;coerce&#39;</span><span class="p">)</span>

<span class="c1"># Drop columns that could not be converted to numeric (if any)</span>
<span class="n">expression_data</span> <span class="o">=</span> <span class="n">expression_data</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">expression_data</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(498, 22637)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># checking for duplicate rows and columns</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of duplicate indices: </span><span class="si">{</span><span class="n">expression_data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>  <span class="c1"># Should be 0</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of duplicate columns: </span><span class="si">{</span><span class="n">expression_data</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">duplicated</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>  <span class="c1"># Should be 0</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of duplicate indices: 0
Number of duplicate columns: 0
</pre></div>
</div>
</div>
</div>
<section id="thresholding-based-on-gene-expression-levels">
<h2>Thresholding Based on Gene Expression Levels<a class="headerlink" href="#thresholding-based-on-gene-expression-levels" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot the distribution of gene expression levels and choose a threshold to exclude low-expressed genes</span>

<span class="c1"># Calculate the mean expression level for each gene</span>
<span class="n">gene_means</span> <span class="o">=</span> <span class="n">expression_data</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Plot the distribution of gene expression levels</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">gene_means</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Mean Gene Expression&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Distribution of Gene Expression Levels&#39;</span><span class="p">)</span>

<span class="c1"># Choose a threshold (example threshold at the mean of the gene means)</span>
<span class="n">threshold</span> <span class="o">=</span> <span class="n">gene_means</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">threshold</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Threshold = </span><span class="si">{</span><span class="n">threshold</span><span class="si">:</span><span class="s1">.2f</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/skamp/.conda/envs/ismbenv/lib/python3.12/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div>
</div>
<img alt="../_images/0dabf73f9955a770531b851edf909fb1591f1c5700f8b45c64d3fd2bf2dd4913.png" src="../_images/0dabf73f9955a770531b851edf909fb1591f1c5700f8b45c64d3fd2bf2dd4913.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to filter out low-expressed genes</span>
<span class="k">def</span> <span class="nf">filter_low_expression_genes</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">1.0</span><span class="p">):</span>
    <span class="c1"># Calculate the mean expression for each gene</span>
    <span class="n">gene_means</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="c1"># Filter out genes with mean expression below the threshold</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">gene_means</span> <span class="o">&gt;=</span> <span class="n">threshold</span>
    <span class="n">filtered_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">mask</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">filtered_data</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="gene-retention-at-various-thresholds">
<h2>Gene Retention at Various Thresholds<a class="headerlink" href="#gene-retention-at-various-thresholds" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define a range of thresholds</span>
<span class="n">thresholds</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)</span>

<span class="c1"># List to store the number of genes retained at each threshold</span>
<span class="n">num_genes</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># Assuming df_renamed is your DataFrame with gene expression data</span>
<span class="k">for</span> <span class="n">threshold</span> <span class="ow">in</span> <span class="n">thresholds</span><span class="p">:</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">filter_low_expression_genes</span><span class="p">(</span><span class="n">expression_data</span><span class="p">,</span> <span class="n">threshold</span><span class="p">)</span>
    <span class="n">num_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df_filtered</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

<span class="c1"># Plot the results</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">thresholds</span><span class="p">,</span> <span class="n">num_genes</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Threshold&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Genes Retained&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Number of Genes Retained at Different Thresholds&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/f0dba6767841754602fe39e1c2102093b5452542f746036fcd80e22dd5064acb.png" src="../_images/f0dba6767841754602fe39e1c2102093b5452542f746036fcd80e22dd5064acb.png" />
</div>
</div>
<p>Based on these two diagrams, let’s choose threshold 8.11</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assign the filtered data to a new variable</span>
<span class="c1"># df_filtered_expression = filter_low_expression_genes(expression_data, threshold=9)</span>
<span class="c1"># print(f&quot;Filtered data shape: {df_filtered_expression.shape}&quot;)  # Check the new shape to confirm filtering</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="thresholding-based-on-gene-expression-variance">
<h2>Thresholding Based on Gene Expression Variance<a class="headerlink" href="#thresholding-based-on-gene-expression-variance" title="Link to this heading">#</a></h2>
<section id="instead-of-gene-expression">
<h3>instead of Gene Expression<a class="headerlink" href="#instead-of-gene-expression" title="Link to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>Genes with higher variance across samples are likely to be more biologically relevant.</p></li>
<li><p>Low-variance genes often represent housekeeping genes or noise.</p></li>
<li><p>High-dimensional data can be computationally expensive to process.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">filter_high_variance_genes</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">threshold</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filters out genes (columns) with variance below the specified threshold.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    data (pd.DataFrame): DataFrame with gene expression data where rows are samples and columns are genes.</span>
<span class="sd">    threshold (float): Minimum variance level to keep a gene.</span>

<span class="sd">    Returns:</span>
<span class="sd">    pd.DataFrame: Filtered DataFrame with genes having variance above the threshold.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Calculate the variance for each gene (column)</span>
    <span class="n">gene_variances</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="c1"># Create a boolean mask to filter out genes with variance below the threshold</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">gene_variances</span> <span class="o">&gt;=</span> <span class="n">threshold</span>
    <span class="c1"># Apply the mask to filter the DataFrame</span>
    <span class="n">filtered_data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">mask</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">filtered_data</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="gene-retention-at-different-variance-levels">
<h2>Gene Retention at Different Variance Levels<a class="headerlink" href="#gene-retention-at-different-variance-levels" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define a range of variance thresholds</span>
<span class="n">variance_thresholds</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)</span>

<span class="c1"># List to store the number of genes retained at each threshold</span>
<span class="n">num_genes</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># Assuming df_renamed is your DataFrame with gene expression data</span>
<span class="k">for</span> <span class="n">threshold</span> <span class="ow">in</span> <span class="n">variance_thresholds</span><span class="p">:</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">filter_high_variance_genes</span><span class="p">(</span><span class="n">expression_data</span><span class="p">,</span> <span class="n">threshold</span><span class="p">)</span>
    <span class="n">num_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df_filtered</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

<span class="c1"># Calculate the variance for each gene</span>
<span class="n">gene_variances</span> <span class="o">=</span> <span class="n">expression_data</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="c1"># Calculate the 75th percentile of the variance distribution</span>
<span class="n">variance_threshold</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">gene_variances</span><span class="p">,</span> <span class="mi">75</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Chosen Variance Threshold: </span><span class="si">{</span><span class="n">variance_threshold</span><span class="w"> </span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Plot the results</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">variance_thresholds</span><span class="p">,</span> <span class="n">num_genes</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">variance_threshold</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;75th Percentile Threshold = </span><span class="si">{</span><span class="n">variance_threshold</span><span class="si">:</span><span class="s1">.2f</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Variance Threshold&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Genes Retained&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Number of Genes Retained at Different Variance Thresholds&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Chosen Variance Threshold: 1.2018901296749187
</pre></div>
</div>
<img alt="../_images/77c9b87def99e294b820d61d6ba9d10f7ab02bf738ae3b9fcc2dc259c60b7814.png" src="../_images/77c9b87def99e294b820d61d6ba9d10f7ab02bf738ae3b9fcc2dc259c60b7814.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assign the filtered data to a new variable</span>
<span class="c1"># Focusing on the top quartile (75th percentile) of genes with the highest variance</span>
<span class="n">df_filtered_variance</span> <span class="o">=</span> <span class="n">filter_high_variance_genes</span><span class="p">(</span><span class="n">expression_data</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="mf">1.2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Filtered data shape: </span><span class="si">{</span><span class="n">df_filtered_variance</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>  <span class="c1"># Check the new shape to confirm filtering</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Filtered data shape: (498, 5671)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># df_filtered_variance.describe()</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="gene-symbol-conversion-ensg-to-hgcn">
<h1>Gene Symbol Conversion: ENSG to HGCN<a class="headerlink" href="#gene-symbol-conversion-ensg-to-hgcn" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">rename_ensembl_to_gene_names</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">chunk_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Renames Ensembl gene IDs to gene names using mygene.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    df (pd.DataFrame): DataFrame with Ensembl gene IDs as columns.</span>
<span class="sd">    chunk_size (int): Number of Ensembl IDs to query at a time.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    pd.DataFrame: DataFrame with gene names as columns, excluding genes that couldn&#39;t be mapped.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1"># Make a copy of the DataFrame to avoid modifying the original</span>
    <span class="n">df_copy</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># Remove the `.number` suffix from ENSG IDs</span>
    <span class="n">df_copy</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">df_copy</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Initialize mygene client</span>
    <span class="n">mg</span> <span class="o">=</span> <span class="n">mygene</span><span class="o">.</span><span class="n">MyGeneInfo</span><span class="p">()</span>

    <span class="c1"># Split ENSG IDs into smaller chunks</span>
    <span class="k">def</span> <span class="nf">chunks</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Yield successive n-sized chunks from lst.&quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">lst</span><span class="p">),</span> <span class="n">n</span><span class="p">):</span>
            <span class="k">yield</span> <span class="n">lst</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span> <span class="o">+</span> <span class="n">n</span><span class="p">]</span>

    <span class="n">ensg_ids</span> <span class="o">=</span> <span class="n">df_copy</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">gene_mappings</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">chunks</span><span class="p">(</span><span class="n">ensg_ids</span><span class="p">,</span> <span class="n">chunk_size</span><span class="p">):</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">mg</span><span class="o">.</span><span class="n">querymany</span><span class="p">(</span><span class="n">chunk</span><span class="p">,</span> <span class="n">scopes</span><span class="o">=</span><span class="s1">&#39;ensembl.gene&#39;</span><span class="p">,</span> <span class="n">fields</span><span class="o">=</span><span class="s1">&#39;symbol&#39;</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">)</span>
        <span class="n">gene_mappings</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

    <span class="c1"># Create a mapping from ENSG to gene symbol, handle missing mappings</span>
    <span class="n">ensg_to_gene</span> <span class="o">=</span> <span class="p">{</span><span class="n">item</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">]:</span> <span class="n">item</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;symbol&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">gene_mappings</span><span class="p">}</span>
    
    <span class="c1"># Log the unmapped genes</span>
    <span class="n">unmapped_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">ensg_ids</span> <span class="k">if</span> <span class="n">ensg_to_gene</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">unmapped_genes</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unmapped genes: </span><span class="si">{</span><span class="n">unmapped_genes</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Filter out genes that couldn&#39;t be mapped</span>
    <span class="n">mapped_columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">ensg_to_gene</span><span class="p">[</span><span class="n">ensg</span><span class="p">]</span> <span class="k">for</span> <span class="n">ensg</span> <span class="ow">in</span> <span class="n">df_copy</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">ensg_to_gene</span><span class="p">[</span><span class="n">ensg</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]</span>

    <span class="c1"># Filter the DataFrame to only include columns that have been mapped</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">df_copy</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">df_copy</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">ensg_to_gene</span><span class="o">.</span><span class="n">keys</span><span class="p">())]</span>

    <span class="c1"># Further filter to ensure we have the same number of columns as mapped gene names</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="n">ensg</span> <span class="k">for</span> <span class="n">ensg</span> <span class="ow">in</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">ensg_to_gene</span><span class="p">[</span><span class="n">ensg</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]]</span>

    <span class="c1"># Assign new column names</span>
    <span class="n">df_filtered</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">ensg_to_gene</span><span class="p">[</span><span class="n">ensg</span><span class="p">]</span> <span class="k">for</span> <span class="n">ensg</span> <span class="ow">in</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>

    <span class="c1"># Handle duplicate gene names by aggregating them (e.g., by taking the mean)</span>
    <span class="n">df_final</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">df_filtered</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">.</span><span class="n">T</span>

    <span class="k">return</span> <span class="n">df_final</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># convert the ensembl gene IDs to gene names</span>
<span class="n">df_renamed</span> <span class="o">=</span> <span class="n">rename_ensembl_to_gene_names</span><span class="p">(</span><span class="n">df_filtered_variance</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1 input query terms found no hit:	[&#39;ENSG00000168078&#39;]
1 input query terms found no hit:	[&#39;ENSG00000180525&#39;]
2 input query terms found no hit:	[&#39;ENSG00000236740&#39;, &#39;ENSG00000239467&#39;]
9 input query terms found no hit:	[&#39;ENSG00000256618&#39;, &#39;ENSG00000259820&#39;, &#39;ENSG00000259834&#39;, &#39;ENSG00000261068&#39;, &#39;ENSG00000261534&#39;, &#39;ENS
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Unmapped genes: [&#39;ENSG00000166104&#39;, &#39;ENSG00000168078&#39;, &#39;ENSG00000180525&#39;, &#39;ENSG00000189229&#39;, &#39;ENSG00000197476&#39;, &#39;ENSG00000203635&#39;, &#39;ENSG00000204584&#39;, &#39;ENSG00000213058&#39;, &#39;ENSG00000214776&#39;, &#39;ENSG00000214797&#39;, &#39;ENSG00000215146&#39;, &#39;ENSG00000217801&#39;, &#39;ENSG00000218048&#39;, &#39;ENSG00000218175&#39;, &#39;ENSG00000218274&#39;, &#39;ENSG00000218617&#39;, &#39;ENSG00000220154&#39;, &#39;ENSG00000222000&#39;, &#39;ENSG00000223829&#39;, &#39;ENSG00000223931&#39;, &#39;ENSG00000224251&#39;, &#39;ENSG00000224585&#39;, &#39;ENSG00000224945&#39;, &#39;ENSG00000226025&#39;, &#39;ENSG00000226286&#39;, &#39;ENSG00000226468&#39;, &#39;ENSG00000226668&#39;, &#39;ENSG00000226816&#39;, &#39;ENSG00000227619&#39;, &#39;ENSG00000227755&#39;, &#39;ENSG00000228113&#39;, &#39;ENSG00000228477&#39;, &#39;ENSG00000228509&#39;, &#39;ENSG00000228613&#39;, &#39;ENSG00000229312&#39;, &#39;ENSG00000229896&#39;, &#39;ENSG00000229970&#39;, &#39;ENSG00000230623&#39;, &#39;ENSG00000230628&#39;, &#39;ENSG00000230699&#39;, &#39;ENSG00000230882&#39;, &#39;ENSG00000231331&#39;, &#39;ENSG00000231533&#39;, &#39;ENSG00000231703&#39;, &#39;ENSG00000231881&#39;, &#39;ENSG00000231892&#39;, &#39;ENSG00000231990&#39;, &#39;ENSG00000232006&#39;, &#39;ENSG00000232949&#39;, &#39;ENSG00000234156&#39;, &#39;ENSG00000234648&#39;, &#39;ENSG00000235563&#39;, &#39;ENSG00000235584&#39;, &#39;ENSG00000235721&#39;, &#39;ENSG00000235939&#39;, &#39;ENSG00000235994&#39;, &#39;ENSG00000236212&#39;, &#39;ENSG00000236283&#39;, &#39;ENSG00000236449&#39;, &#39;ENSG00000236740&#39;, &#39;ENSG00000237550&#39;, &#39;ENSG00000237813&#39;, &#39;ENSG00000238279&#39;, &#39;ENSG00000239467&#39;, &#39;ENSG00000244306&#39;, &#39;ENSG00000247134&#39;, &#39;ENSG00000248347&#39;, &#39;ENSG00000248545&#39;, &#39;ENSG00000249001&#39;, &#39;ENSG00000249007&#39;, &#39;ENSG00000249199&#39;, &#39;ENSG00000249743&#39;, &#39;ENSG00000249746&#39;, &#39;ENSG00000250049&#39;, &#39;ENSG00000250509&#39;, &#39;ENSG00000250770&#39;, &#39;ENSG00000250978&#39;, &#39;ENSG00000251095&#39;, &#39;ENSG00000251532&#39;, &#39;ENSG00000253217&#39;, &#39;ENSG00000253227&#39;, &#39;ENSG00000253288&#39;, &#39;ENSG00000253339&#39;, &#39;ENSG00000253361&#39;, &#39;ENSG00000254054&#39;, &#39;ENSG00000254143&#39;, &#39;ENSG00000254290&#39;, &#39;ENSG00000254367&#39;, &#39;ENSG00000254632&#39;, &#39;ENSG00000254851&#39;, &#39;ENSG00000254988&#39;, &#39;ENSG00000255026&#39;, &#39;ENSG00000255085&#39;, &#39;ENSG00000255182&#39;, &#39;ENSG00000255310&#39;, &#39;ENSG00000255345&#39;, &#39;ENSG00000255366&#39;, &#39;ENSG00000255462&#39;, &#39;ENSG00000255847&#39;, &#39;ENSG00000255910&#39;, &#39;ENSG00000255983&#39;, &#39;ENSG00000256417&#39;, &#39;ENSG00000256443&#39;, &#39;ENSG00000256615&#39;, &#39;ENSG00000256618&#39;, &#39;ENSG00000256673&#39;, &#39;ENSG00000256969&#39;, &#39;ENSG00000257474&#39;, &#39;ENSG00000257512&#39;, &#39;ENSG00000257604&#39;, &#39;ENSG00000257826&#39;, &#39;ENSG00000257829&#39;, &#39;ENSG00000257894&#39;, &#39;ENSG00000258053&#39;, &#39;ENSG00000258077&#39;, &#39;ENSG00000258168&#39;, &#39;ENSG00000258342&#39;, &#39;ENSG00000258346&#39;, &#39;ENSG00000258661&#39;, &#39;ENSG00000258676&#39;, &#39;ENSG00000258752&#39;, &#39;ENSG00000258837&#39;, &#39;ENSG00000258844&#39;, &#39;ENSG00000259087&#39;, &#39;ENSG00000259094&#39;, &#39;ENSG00000259342&#39;, &#39;ENSG00000259345&#39;, &#39;ENSG00000259353&#39;, &#39;ENSG00000259425&#39;, &#39;ENSG00000259723&#39;, &#39;ENSG00000259820&#39;, &#39;ENSG00000259834&#39;, &#39;ENSG00000259925&#39;, &#39;ENSG00000260197&#39;, &#39;ENSG00000260289&#39;, &#39;ENSG00000260328&#39;, &#39;ENSG00000260337&#39;, &#39;ENSG00000260430&#39;, &#39;ENSG00000260597&#39;, &#39;ENSG00000260604&#39;, &#39;ENSG00000260710&#39;, &#39;ENSG00000260711&#39;, &#39;ENSG00000260877&#39;, &#39;ENSG00000260997&#39;, &#39;ENSG00000261068&#39;, &#39;ENSG00000261104&#39;, &#39;ENSG00000261123&#39;, &#39;ENSG00000261172&#39;, &#39;ENSG00000261296&#39;, &#39;ENSG00000261305&#39;, &#39;ENSG00000261502&#39;, &#39;ENSG00000261534&#39;, &#39;ENSG00000261578&#39;, &#39;ENSG00000261600&#39;, &#39;ENSG00000261762&#39;, &#39;ENSG00000261786&#39;, &#39;ENSG00000261804&#39;, &#39;ENSG00000262539&#39;, &#39;ENSG00000262580&#39;, &#39;ENSG00000262714&#39;, &#39;ENSG00000262873&#39;, &#39;ENSG00000262877&#39;, &#39;ENSG00000262920&#39;, &#39;ENSG00000262979&#39;, &#39;ENSG00000263301&#39;, &#39;ENSG00000263893&#39;, &#39;ENSG00000264112&#39;, &#39;ENSG00000264659&#39;, &#39;ENSG00000265298&#39;, &#39;ENSG00000265750&#39;, &#39;ENSG00000266302&#39;, &#39;ENSG00000266830&#39;, &#39;ENSG00000266869&#39;, &#39;ENSG00000266968&#39;, &#39;ENSG00000267072&#39;, &#39;ENSG00000267383&#39;, &#39;ENSG00000267466&#39;, &#39;ENSG00000267506&#39;, &#39;ENSG00000267640&#39;, &#39;ENSG00000267886&#39;, &#39;ENSG00000267892&#39;, &#39;ENSG00000268362&#39;, &#39;ENSG00000268416&#39;, &#39;ENSG00000268555&#39;, &#39;ENSG00000268833&#39;, &#39;ENSG00000268903&#39;, &#39;ENSG00000269028&#39;, &#39;ENSG00000269353&#39;, &#39;ENSG00000269918&#39;, &#39;ENSG00000269954&#39;, &#39;ENSG00000269981&#39;, &#39;ENSG00000270460&#39;, &#39;ENSG00000270605&#39;, &#39;ENSG00000270933&#39;, &#39;ENSG00000270977&#39;, &#39;ENSG00000271392&#39;, &#39;ENSG00000271959&#39;, &#39;ENSG00000272079&#39;, &#39;ENSG00000272141&#39;, &#39;ENSG00000272279&#39;, &#39;ENSG00000272379&#39;, &#39;ENSG00000272405&#39;, &#39;ENSG00000272502&#39;, &#39;ENSG00000272512&#39;, &#39;ENSG00000272622&#39;, &#39;ENSG00000272626&#39;, &#39;ENSG00000272720&#39;, &#39;ENSG00000272825&#39;, &#39;ENSG00000272865&#39;, &#39;ENSG00000273108&#39;, &#39;ENSG00000273151&#39;, &#39;ENSG00000273179&#39;, &#39;ENSG00000273203&#39;, &#39;ENSG00000273230&#39;, &#39;ENSG00000273295&#39;, &#39;ENSG00000273301&#39;, &#39;ENSG00000273328&#39;, &#39;ENSG00000273760&#39;, &#39;ENSG00000273906&#39;, &#39;ENSG00000274002&#39;, &#39;ENSG00000274414&#39;, &#39;ENSG00000274979&#39;, &#39;ENSG00000275678&#39;, &#39;ENSG00000276093&#39;, &#39;ENSG00000276851&#39;, &#39;ENSG00000276980&#39;, &#39;ENSG00000277287&#39;, &#39;ENSG00000277840&#39;, &#39;ENSG00000278263&#39;, &#39;ENSG00000278266&#39;, &#39;ENSG00000278484&#39;, &#39;ENSG00000278864&#39;, &#39;ENSG00000278912&#39;, &#39;ENSG00000278962&#39;, &#39;ENSG00000279017&#39;, &#39;ENSG00000279286&#39;, &#39;ENSG00000279511&#39;, &#39;ENSG00000279518&#39;, &#39;ENSG00000279662&#39;, &#39;ENSG00000279806&#39;, &#39;ENSG00000279878&#39;, &#39;ENSG00000279930&#39;, &#39;ENSG00000280002&#39;, &#39;ENSG00000280027&#39;, &#39;ENSG00000280032&#39;, &#39;ENSG00000280079&#39;, &#39;ENSG00000280138&#39;, &#39;ENSG00000280202&#39;, &#39;ENSG00000280217&#39;, &#39;ENSG00000280285&#39;, &#39;ENSG00000280334&#39;, &#39;ENSG00000280399&#39;, &#39;ENSG00000280435&#39;, &#39;ENSG00000283098&#39;, &#39;ENSG00000283142&#39;, &#39;ENSG00000283413&#39;, &#39;ENSG00000283486&#39;, &#39;ENSG00000283631&#39;, &#39;ENSG00000283633&#39;, &#39;ENSG00000283907&#39;, &#39;ENSG00000284636&#39;, &#39;ENSG00000284928&#39;, &#39;ENSG00000285108&#39;, &#39;ENSG00000285128&#39;, &#39;ENSG00000285417&#39;, &#39;ENSG00000285476&#39;, &#39;ENSG00000285517&#39;, &#39;ENSG00000285601&#39;, &#39;ENSG00000285820&#39;, &#39;ENSG00000285847&#39;, &#39;ENSG00000285933&#39;, &#39;ENSG00000286009&#39;, &#39;ENSG00000286015&#39;, &#39;ENSG00000286073&#39;, &#39;ENSG00000286133&#39;, &#39;ENSG00000286191&#39;, &#39;ENSG00000286256&#39;, &#39;ENSG00000286299&#39;, &#39;ENSG00000286342&#39;, &#39;ENSG00000286355&#39;, &#39;ENSG00000286358&#39;, &#39;ENSG00000286403&#39;, &#39;ENSG00000286449&#39;, &#39;ENSG00000286456&#39;, &#39;ENSG00000286519&#39;, &#39;ENSG00000286587&#39;, &#39;ENSG00000286722&#39;, &#39;ENSG00000286750&#39;, &#39;ENSG00000286797&#39;, &#39;ENSG00000286802&#39;, &#39;ENSG00000286818&#39;, &#39;ENSG00000286830&#39;, &#39;ENSG00000286892&#39;, &#39;ENSG00000286944&#39;, &#39;ENSG00000286989&#39;, &#39;ENSG00000287003&#39;, &#39;ENSG00000287051&#39;, &#39;ENSG00000287052&#39;, &#39;ENSG00000287055&#39;, &#39;ENSG00000287059&#39;, &#39;ENSG00000287097&#39;, &#39;ENSG00000287185&#39;, &#39;ENSG00000287238&#39;, &#39;ENSG00000287245&#39;, &#39;ENSG00000287252&#39;, &#39;ENSG00000287271&#39;, &#39;ENSG00000287338&#39;, &#39;ENSG00000287352&#39;, &#39;ENSG00000287390&#39;, &#39;ENSG00000287426&#39;, &#39;ENSG00000287430&#39;, &#39;ENSG00000287460&#39;, &#39;ENSG00000287550&#39;, &#39;ENSG00000287626&#39;, &#39;ENSG00000287627&#39;, &#39;ENSG00000287656&#39;, &#39;ENSG00000287673&#39;, &#39;ENSG00000287725&#39;, &#39;ENSG00000287770&#39;, &#39;ENSG00000287778&#39;, &#39;ENSG00000287804&#39;, &#39;ENSG00000287839&#39;, &#39;ENSG00000287860&#39;, &#39;ENSG00000287879&#39;, &#39;ENSG00000287887&#39;, &#39;ENSG00000287906&#39;, &#39;ENSG00000287920&#39;, &#39;ENSG00000287929&#39;, &#39;ENSG00000288075&#39;, &#39;ENSG00000288218&#39;]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_filtered_variance</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(498, 5671)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_renamed</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(498, 5338)
</pre></div>
</div>
</div>
</div>
<p>Different number of rows indicate that some of the genes werent mapped to a gene ID or multiple were mapped to the same gene ID.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_renamed</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>A2M</th>
      <th>A2ML1</th>
      <th>A4GALT</th>
      <th>AACSP1</th>
      <th>AADAC</th>
      <th>AADACP1</th>
      <th>AARD</th>
      <th>AASS</th>
      <th>AATBC</th>
      <th>AATK</th>
      <th>...</th>
      <th>ZNF90</th>
      <th>ZP3</th>
      <th>ZPLD1</th>
      <th>ZPLD2P</th>
      <th>ZSCAN1</th>
      <th>ZSCAN23</th>
      <th>ZSCAN31</th>
      <th>ZSCAN4</th>
      <th>ZWINT</th>
      <th>ZYG11A</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TCGA-38-7271</th>
      <td>17.7492</td>
      <td>4.4411</td>
      <td>10.1862</td>
      <td>3.5845</td>
      <td>9.6415</td>
      <td>3.5845</td>
      <td>8.0089</td>
      <td>11.2890</td>
      <td>6.8710</td>
      <td>8.6552</td>
      <td>...</td>
      <td>6.1774</td>
      <td>6.9127</td>
      <td>4.9403</td>
      <td>3.5845</td>
      <td>3.8172</td>
      <td>5.7385</td>
      <td>9.0817</td>
      <td>3.2709</td>
      <td>9.2615</td>
      <td>4.0073</td>
    </tr>
    <tr>
      <th>TCGA-55-7914</th>
      <td>14.8513</td>
      <td>4.4530</td>
      <td>8.9312</td>
      <td>3.6762</td>
      <td>6.5685</td>
      <td>3.6762</td>
      <td>6.8983</td>
      <td>10.5920</td>
      <td>7.9894</td>
      <td>8.2660</td>
      <td>...</td>
      <td>9.9162</td>
      <td>7.9426</td>
      <td>2.4821</td>
      <td>2.4821</td>
      <td>5.0222</td>
      <td>4.3796</td>
      <td>10.9565</td>
      <td>3.0918</td>
      <td>10.1334</td>
      <td>8.3219</td>
    </tr>
    <tr>
      <th>TCGA-95-7043</th>
      <td>14.1691</td>
      <td>4.5026</td>
      <td>9.0834</td>
      <td>3.8623</td>
      <td>5.5634</td>
      <td>8.0183</td>
      <td>3.6226</td>
      <td>10.5600</td>
      <td>7.5223</td>
      <td>8.8214</td>
      <td>...</td>
      <td>5.3150</td>
      <td>5.8224</td>
      <td>3.6226</td>
      <td>3.2988</td>
      <td>3.8623</td>
      <td>5.6728</td>
      <td>10.8095</td>
      <td>2.4821</td>
      <td>10.9149</td>
      <td>6.7593</td>
    </tr>
    <tr>
      <th>TCGA-73-4658</th>
      <td>16.7238</td>
      <td>3.1704</td>
      <td>9.1443</td>
      <td>2.4821</td>
      <td>4.3206</td>
      <td>3.6533</td>
      <td>4.8856</td>
      <td>10.5574</td>
      <td>6.0253</td>
      <td>7.8940</td>
      <td>...</td>
      <td>4.6727</td>
      <td>6.9805</td>
      <td>5.4212</td>
      <td>3.6533</td>
      <td>3.4468</td>
      <td>3.6533</td>
      <td>8.9874</td>
      <td>3.4468</td>
      <td>9.7828</td>
      <td>3.4468</td>
    </tr>
    <tr>
      <th>TCGA-86-8076</th>
      <td>15.6783</td>
      <td>4.7422</td>
      <td>9.5150</td>
      <td>3.2271</td>
      <td>7.2321</td>
      <td>5.1146</td>
      <td>7.2174</td>
      <td>10.3567</td>
      <td>7.2467</td>
      <td>10.4765</td>
      <td>...</td>
      <td>4.8254</td>
      <td>7.9800</td>
      <td>3.6417</td>
      <td>3.9275</td>
      <td>3.6417</td>
      <td>5.0478</td>
      <td>9.5810</td>
      <td>3.0118</td>
      <td>8.6640</td>
      <td>4.7844</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TCGA-86-8073</th>
      <td>16.3773</td>
      <td>4.0817</td>
      <td>9.4234</td>
      <td>6.6519</td>
      <td>6.7641</td>
      <td>4.2611</td>
      <td>7.1467</td>
      <td>11.3497</td>
      <td>7.0724</td>
      <td>8.7741</td>
      <td>...</td>
      <td>8.6449</td>
      <td>9.2445</td>
      <td>5.9388</td>
      <td>6.8492</td>
      <td>5.8132</td>
      <td>7.2074</td>
      <td>10.2410</td>
      <td>5.1126</td>
      <td>10.1390</td>
      <td>8.2478</td>
    </tr>
    <tr>
      <th>TCGA-MN-A4N4</th>
      <td>14.9411</td>
      <td>6.7210</td>
      <td>11.9107</td>
      <td>8.8283</td>
      <td>3.4947</td>
      <td>3.0746</td>
      <td>4.1736</td>
      <td>9.6396</td>
      <td>6.6811</td>
      <td>9.3538</td>
      <td>...</td>
      <td>10.1845</td>
      <td>7.5083</td>
      <td>5.0939</td>
      <td>5.7178</td>
      <td>6.2061</td>
      <td>6.9167</td>
      <td>9.4672</td>
      <td>3.3143</td>
      <td>10.0583</td>
      <td>8.4566</td>
    </tr>
    <tr>
      <th>TCGA-53-7626</th>
      <td>16.7343</td>
      <td>5.2575</td>
      <td>9.5379</td>
      <td>2.4821</td>
      <td>5.7120</td>
      <td>5.0910</td>
      <td>5.6287</td>
      <td>10.7802</td>
      <td>9.3334</td>
      <td>9.3085</td>
      <td>...</td>
      <td>5.4842</td>
      <td>7.9278</td>
      <td>3.8286</td>
      <td>3.7563</td>
      <td>3.8286</td>
      <td>6.0927</td>
      <td>10.0507</td>
      <td>4.2816</td>
      <td>9.4627</td>
      <td>7.4014</td>
    </tr>
    <tr>
      <th>TCGA-44-A47G</th>
      <td>15.6622</td>
      <td>3.7345</td>
      <td>9.8660</td>
      <td>4.1045</td>
      <td>6.2214</td>
      <td>4.9759</td>
      <td>4.2980</td>
      <td>10.3975</td>
      <td>7.8841</td>
      <td>10.1369</td>
      <td>...</td>
      <td>4.9218</td>
      <td>7.7402</td>
      <td>5.9326</td>
      <td>4.7450</td>
      <td>3.9944</td>
      <td>5.3854</td>
      <td>7.4843</td>
      <td>4.4650</td>
      <td>9.5817</td>
      <td>4.6804</td>
    </tr>
    <tr>
      <th>TCGA-55-6969</th>
      <td>14.8136</td>
      <td>4.3058</td>
      <td>9.7702</td>
      <td>6.8425</td>
      <td>5.2016</td>
      <td>3.9293</td>
      <td>7.1699</td>
      <td>8.8889</td>
      <td>6.1528</td>
      <td>7.6565</td>
      <td>...</td>
      <td>6.4289</td>
      <td>7.9499</td>
      <td>4.0012</td>
      <td>6.4059</td>
      <td>3.7694</td>
      <td>3.6792</td>
      <td>8.5268</td>
      <td>2.4821</td>
      <td>10.8228</td>
      <td>8.6068</td>
    </tr>
  </tbody>
</table>
<p>498 rows × 5338 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_renamed</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s2">&quot;expression_data_filtered.csv&quot;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="checking-for-missing-data-and-outliers">
<h2>Checking for Missing Data and Outliers<a class="headerlink" href="#checking-for-missing-data-and-outliers" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check for missing values in each column</span>
<span class="n">missing_values</span> <span class="o">=</span> <span class="n">df_renamed</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

<span class="c1"># Calculate the percentage of missing values for each column</span>
<span class="n">missing_percentage</span> <span class="o">=</span> <span class="p">(</span><span class="n">missing_values</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">))</span> <span class="o">*</span> <span class="mi">100</span>

<span class="c1"># Create a DataFrame to hold the missing data statistics</span>
<span class="n">missing_stats</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;Total Missing&#39;</span><span class="p">:</span> <span class="n">missing_values</span><span class="p">,</span> <span class="s1">&#39;Percentage&#39;</span><span class="p">:</span> <span class="n">missing_percentage</span><span class="p">})</span>

<span class="c1"># Filter to display only columns with missing values</span>
<span class="n">missing_stats</span> <span class="o">=</span> <span class="n">missing_stats</span><span class="p">[</span><span class="n">missing_stats</span><span class="p">[</span><span class="s1">&#39;Total Missing&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>

<span class="c1"># Print out the missing values statistics for columns with missing data</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">missing_stats</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Missing values in columns:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">missing_stats</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No missing values found in the dataset.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>No missing values found in the dataset.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate the mean of each column</span>
<span class="n">means</span> <span class="o">=</span> <span class="n">df_renamed</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>

<span class="c1"># Create a list of column IDs (integer index)</span>
<span class="n">column_ids</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df_renamed</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span>

<span class="c1"># Plotting</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">column_ids</span><span class="p">,</span> <span class="n">means</span><span class="p">)</span>

<span class="c1"># Labeling the axes</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Column ID&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Mean Expression&#39;</span><span class="p">)</span>

<span class="c1"># Adding a title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Scatter Plot of Expression Means vs. Column IDs&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/67eca73944cc6a35fff03f3961120fb99c487342f27de70a6d1b90bd6fbecfdb.png" src="../_images/67eca73944cc6a35fff03f3961120fb99c487342f27de70a6d1b90bd6fbecfdb.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Sort and get the top 10 and bottom 10 columns based on their mean values</span>
<span class="n">top_10_columns</span> <span class="o">=</span> <span class="n">means</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="n">bottom_10_columns</span> <span class="o">=</span> <span class="n">means</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>

<span class="c1"># Combine top 10 and bottom 10 columns into one DataFrame</span>
<span class="n">combined</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">top_10_columns</span><span class="p">,</span> <span class="n">bottom_10_columns</span><span class="p">])</span>

<span class="c1"># Plot the results</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">combined</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;bar&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;skyblue&#39;</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="mi">10</span> <span class="k">else</span> <span class="s1">&#39;lightcoral&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">20</span><span class="p">)])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Top 10 and Bottom 10 Genes Based on Mean Expression Values&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Mean Value&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/a1bd13c4db0d9fa2c9b2e4b9a609af3d7408f3e7921effd95821624aa0e8811e.png" src="../_images/a1bd13c4db0d9fa2c9b2e4b9a609af3d7408f3e7921effd95821624aa0e8811e.png" />
</div>
</div>
</section>
<section id="data-heterogeneity">
<h2>Data heterogeneity<a class="headerlink" href="#data-heterogeneity" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select numerical columns</span>
<span class="n">numerical_columns</span> <span class="o">=</span> <span class="n">df_renamed</span><span class="o">.</span><span class="n">select_dtypes</span><span class="p">(</span><span class="n">include</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;float64&#39;</span><span class="p">,</span> <span class="s1">&#39;int64&#39;</span><span class="p">])</span>

<span class="c1"># Calculate the mean of each column for the numerical columns only</span>
<span class="n">means</span> <span class="o">=</span> <span class="n">numerical_columns</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>

<span class="c1"># Get the top 10 and bottom 10 columns based on mean values</span>
<span class="n">top_8_columns</span> <span class="o">=</span> <span class="n">means</span><span class="o">.</span><span class="n">nlargest</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">bottom_8_columns</span> <span class="o">=</span> <span class="n">means</span><span class="o">.</span><span class="n">nsmallest</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="c1"># Combine top 10 and bottom 10 columns for plotting</span>
<span class="n">columns_to_plot</span> <span class="o">=</span> <span class="n">top_8_columns</span> <span class="o">+</span> <span class="n">bottom_8_columns</span>

<span class="c1"># Filter the numerical columns to only include those to plot</span>
<span class="n">filtered_numerical_columns</span> <span class="o">=</span> <span class="n">numerical_columns</span><span class="p">[</span><span class="n">columns_to_plot</span><span class="p">]</span>

<span class="c1"># Calculate the number of rows needed for subplots based on the number of selected columns</span>
<span class="n">num_plots</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered_numerical_columns</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="n">num_rows</span> <span class="o">=</span> <span class="p">(</span><span class="n">num_plots</span> <span class="o">//</span> <span class="mi">4</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">num_plots</span> <span class="o">%</span> <span class="mi">4</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>  <span class="c1"># Ensure there is an extra row if there are leftovers</span>

<span class="c1"># Plot histograms for each selected numerical column</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">num_rows</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">5</span> <span class="o">*</span> <span class="n">num_rows</span><span class="p">))</span>  <span class="c1"># Adjust width and height as needed</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Distribution of Top 8 &amp; Bottom 8 Genes Mean Expression&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">col</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">filtered_numerical_columns</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="o">.</span><span class="n">flatten</span><span class="p">()[</span><span class="n">i</span><span class="p">]</span>
    <span class="n">filtered_numerical_columns</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Value&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>

<span class="c1"># Hide any unused axes if the number of plots isn&#39;t a perfect multiple of 4</span>
<span class="k">if</span> <span class="n">num_plots</span> <span class="o">%</span> <span class="mi">4</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="n">axes</span><span class="o">.</span><span class="n">flatten</span><span class="p">()[</span><span class="n">num_plots</span><span class="p">:]:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">(</span><span class="n">rect</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">])</span>  <span class="c1"># Adjust layout to make room for the suptitle</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ca2e5f0b9c680b3a4fa6f83f6e658386de4fac46d9b240636bd3057e71587208.png" src="../_images/ca2e5f0b9c680b3a4fa6f83f6e658386de4fac46d9b240636bd3057e71587208.png" />
</div>
</div>
</section>
</section>
<section id="correlation-matrix-calculation">
<h1>Correlation Matrix Calculation<a class="headerlink" href="#correlation-matrix-calculation" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># function to calculate absolute biweight midcorrelation </span>
<span class="k">def</span> <span class="nf">abs_bicorr</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
    <span class="c1"># Select only numeric data</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">_get_numeric_data</span><span class="p">()</span>
    <span class="n">cols</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">columns</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="n">cols</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">na_value</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="o">.</span><span class="n">T</span>

    <span class="n">K</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span>
    <span class="n">correl</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">K</span><span class="p">,</span> <span class="n">K</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>

    <span class="c1"># Calculate biweight midcovariance</span>
    <span class="n">bicorr</span> <span class="o">=</span> <span class="n">astropy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">biweight_midcovariance</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">modify_sample_size</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">K</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">K</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">j</span><span class="p">:</span>
                <span class="n">correl</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">denominator</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">bicorr</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">bicorr</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="n">j</span><span class="p">])</span>
                <span class="k">if</span> <span class="n">denominator</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">correl</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">bicorr</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">/</span> <span class="n">denominator</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">correl</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># Or handle it in another appropriate way</span>

    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">correl</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="n">idx</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">cols</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to calculate mutual information</span>
<span class="k">def</span> <span class="nf">mutual_information</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
    <span class="n">mi</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">j</span><span class="p">:</span>
                <span class="n">mi</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">mutual_info_regression</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">mi</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>
    <span class="k">return</span> <span class="n">mi</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="consider-loading-in-precomputed-correlation-matrices-and-skipping-the-next-cell">
<h2>Consider loading in precomputed correlation matrices and skipping the next cell<a class="headerlink" href="#consider-loading-in-precomputed-correlation-matrices-and-skipping-the-next-cell" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the entire dictionary of correlation matrices from a pickle file</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s2">&quot;correlation_matrices.pkl&quot;</span><span class="p">),</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">correlation_matrices</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

<span class="c1"># Verify the loaded data</span>
<span class="nb">print</span><span class="p">(</span><span class="n">correlation_matrices</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>dict_keys([&#39;pearson&#39;, &#39;spearman&#39;, &#39;biweight_midcorrelation&#39;])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dictionary to store different correlation matrices</span>
<span class="n">correlation_matrices</span> <span class="o">=</span> <span class="p">{}</span>

<span class="c1"># Pearson correlation - O(n^2) complexity, fast for large datasets</span>
<span class="n">correlation_matrices</span><span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_renamed</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">)</span>

<span class="c1"># Spearman rank correlation -  O(n^2 log n) complexity, relatively fast but can be slower than Pearson</span>
<span class="n">correlation_matrices</span><span class="p">[</span><span class="s1">&#39;spearman&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_renamed</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;spearman&#39;</span><span class="p">)</span>

<span class="c1"># Biweight midcorrelation -  Robust but slower than Pearson and Spearman, suitable for datasets with outliers</span>
<span class="n">correlation_matrices</span><span class="p">[</span><span class="s1">&#39;biweight_midcorrelation&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">abs_bicorr</span><span class="p">(</span><span class="n">df_renamed</span><span class="p">)</span>

<span class="c1"># Mutual information -  O(n^3) complexity, very slow for large datasets</span>
<span class="c1"># correlation_matrices[&#39;mutual_information&#39;] = mutual_information(df_renamed)</span>

<span class="c1"># Print the keys of the correlation matrices to verify</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Correlation matrices calculated:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">correlation_matrices</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize correlation matrices using heatmaps</span>
<span class="k">def</span> <span class="nf">plot_correlation_matrices</span><span class="p">(</span><span class="n">correlation_matrices</span><span class="p">):</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>  <span class="c1"># Adjust to 1 row and 3 columns</span>
    <span class="n">axes</span> <span class="o">=</span> <span class="n">axes</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">matrix</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">correlation_matrices</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">matrix</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;coolwarm&#39;</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">xticklabels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">yticklabels</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">key</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s1"> Correlation Matrix&#39;</span><span class="p">)</span>
        
        <span class="c1"># Set square aspect ratio</span>
        <span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;box&#39;</span><span class="p">)</span>
    
    <span class="c1"># Hide any unused subplots</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">axes</span><span class="p">)):</span>
        <span class="n">axes</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="todo-check-grouping-ordering">
<h1>TODO: check grouping, ordering ?<a class="headerlink" href="#todo-check-grouping-ordering" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_correlation_matrices</span><span class="p">(</span><span class="n">correlation_matrices</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/1a2818348e764be2b27d76a2185c2961a62e0d44b6d82834abf5d33cb06ed428.png" src="../_images/1a2818348e764be2b27d76a2185c2961a62e0d44b6d82834abf5d33cb06ed428.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">correlation_matrices</span><span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(5338, 5338)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Save the entire dictionary of correlation matrices as a pickle file</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s2">&quot;correlation_matrices.pkl&quot;</span><span class="p">),</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">correlation_matrices</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="graph-construction-from-correlation-matrices">
<h1>Graph Construction from Correlation Matrices<a class="headerlink" href="#graph-construction-from-correlation-matrices" title="Link to this heading">#</a></h1>
</section>
<section id="todo-list-comprehension-for-graph-generation">
<h1>TODO: list comprehension for Graph generation ??<a class="headerlink" href="#todo-list-comprehension-for-graph-generation" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">create_graph_from_correlation</span><span class="p">(</span><span class="n">correlation_matrix</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.8</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a graph from a correlation matrix using a specified threshold.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    correlation_matrix (pd.DataFrame): DataFrame containing the correlation matrix.</span>
<span class="sd">    threshold (float): Threshold for including edges based on correlation value.</span>

<span class="sd">    Returns:</span>
<span class="sd">    G (nx.Graph): Graph created from the correlation matrix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">G</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>

    <span class="c1"># Add nodes</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">correlation_matrix</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">G</span><span class="o">.</span><span class="n">add_node</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>

    <span class="c1"># Add edges with weights above the threshold</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">correlation_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">correlation_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">j</span><span class="p">:</span>  <span class="c1"># Ignore the diagonal elements</span>
                <span class="n">weight</span> <span class="o">=</span> <span class="n">correlation_matrix</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>
                <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">weight</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">threshold</span><span class="p">:</span>
                    <span class="n">G</span><span class="o">.</span><span class="n">add_edge</span><span class="p">(</span><span class="n">correlation_matrix</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">correlation_matrix</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="n">j</span><span class="p">],</span> <span class="n">weight</span><span class="o">=</span><span class="n">weight</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">verify_edge_weights</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">num_samples</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Verifies that edges in the graph have the correct weights from the correlation matrix.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): The NetworkX graph.</span>
<span class="sd">    num_samples (int): Number of edges to sample and print for verification.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">edges</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
    <span class="n">sample</span> <span class="o">=</span> <span class="n">edges</span><span class="p">[:</span><span class="n">num_samples</span><span class="p">]</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">edges</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">num_samples</span> <span class="k">else</span> <span class="n">edges</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sample of </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span><span class="si">}</span><span class="s2"> edges with weights:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">sample</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Edge (</span><span class="si">{</span><span class="n">u</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">v</span><span class="si">}</span><span class="s2">) - weight: </span><span class="si">{</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;weight&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="show-densities-what-treshold-is-useful-to-get-rid-of-the-fully-connected-network">
<h1>show densities, what treshold is useful to get rid of the fully connected network<a class="headerlink" href="#show-densities-what-treshold-is-useful-to-get-rid-of-the-fully-connected-network" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># correlation_matrices is a dictionary containing correlation matrices</span>
<span class="n">pearson_graph</span> <span class="o">=</span> <span class="n">create_graph_from_correlation</span><span class="p">(</span><span class="n">correlation_matrices</span><span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">],</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">verify_edge_weights</span><span class="p">(</span><span class="n">pearson_graph</span><span class="p">,</span> <span class="n">num_samples</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sample of 10 edges with weights:
Edge (A2M, ABCA8) - weight: 0.8308274641648915
Edge (A2M, ADAMTS8) - weight: 0.8036704002698273
Edge (A2M, AOC3) - weight: 0.8767411078325781
Edge (A2M, C7) - weight: 0.8013688242083488
Edge (A2M, INMT) - weight: 0.8117391230502314
Edge (A2M, ITGA8) - weight: 0.8724141478160804
Edge (A2M, MFAP4) - weight: 0.8444697225886688
Edge (A2M, PKNOX2) - weight: 0.8062841461653377
Edge (A2M, SCN7A) - weight: 0.8382334505662205
Edge (A2M, SLIT2) - weight: 0.8041515706739126
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">biweight_graph</span> <span class="o">=</span> <span class="n">create_graph_from_correlation</span><span class="p">(</span><span class="n">correlation_matrices</span><span class="p">[</span><span class="s1">&#39;biweight_midcorrelation&#39;</span><span class="p">],</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to visualize the graph</span>
<span class="k">def</span> <span class="nf">visualize_graph</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Gene Co-expression Network&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualizes the graph using Matplotlib and NetworkX.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): Graph to visualize.</span>
<span class="sd">    title (str): Title of the plot.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">spring_layout</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>  <span class="c1"># k controls the distance between nodes</span>
    <span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_nodes</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">node_size</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">node_color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    <span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_edges</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the graph</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">pearson_graph</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Correlation Network (Threshold = 0.8)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/bc1bf43a3cb189bf4108490e06ef682753dfa20310d909fd6442b9255c24ecce.png" src="../_images/bc1bf43a3cb189bf4108490e06ef682753dfa20310d909fd6442b9255c24ecce.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">print_graph_info</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prints basic information about the graph.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): The NetworkX graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of nodes: </span><span class="si">{</span><span class="n">G</span><span class="o">.</span><span class="n">number_of_nodes</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of edges: </span><span class="si">{</span><span class="n">G</span><span class="o">.</span><span class="n">number_of_edges</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample nodes:&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">)[:</span><span class="mi">10</span><span class="p">])</span>  <span class="c1"># Print first 10 nodes as a sample</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample edges:&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">))[:</span><span class="mi">10</span><span class="p">])</span>  <span class="c1"># Print first 10 edges as a sample</span>
    
    <span class="c1"># Check for self-loops</span>
    <span class="n">self_loops</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">selfloop_edges</span><span class="p">(</span><span class="n">G</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">self_loops</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of self-loops: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">self_loops</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Self-loops:&quot;</span><span class="p">,</span> <span class="n">self_loops</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No self-loops in the graph.&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">calculate_average_clustering</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates and prints the average clustering coefficient of the graph.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): The NetworkX graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">avg_clustering</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">average_clustering</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Average clustering coefficient: </span><span class="si">{</span><span class="n">avg_clustering</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_graph_info</span><span class="p">(</span><span class="n">pearson_graph</span><span class="p">)</span>
<span class="n">calculate_average_clustering</span><span class="p">(</span><span class="n">pearson_graph</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 5338
Number of edges: 13449
Sample nodes: [&#39;A2M&#39;, &#39;A2ML1&#39;, &#39;A4GALT&#39;, &#39;AACSP1&#39;, &#39;AADAC&#39;, &#39;AADACP1&#39;, &#39;AARD&#39;, &#39;AASS&#39;, &#39;AATBC&#39;, &#39;AATK&#39;]
Sample edges: [(&#39;A2M&#39;, &#39;ABCA8&#39;, {&#39;weight&#39;: 0.8308274641648915}), (&#39;A2M&#39;, &#39;ADAMTS8&#39;, {&#39;weight&#39;: 0.8036704002698273}), (&#39;A2M&#39;, &#39;AOC3&#39;, {&#39;weight&#39;: 0.8767411078325781}), (&#39;A2M&#39;, &#39;C7&#39;, {&#39;weight&#39;: 0.8013688242083488}), (&#39;A2M&#39;, &#39;INMT&#39;, {&#39;weight&#39;: 0.8117391230502314}), (&#39;A2M&#39;, &#39;ITGA8&#39;, {&#39;weight&#39;: 0.8724141478160804}), (&#39;A2M&#39;, &#39;MFAP4&#39;, {&#39;weight&#39;: 0.8444697225886688}), (&#39;A2M&#39;, &#39;PKNOX2&#39;, {&#39;weight&#39;: 0.8062841461653377}), (&#39;A2M&#39;, &#39;SCN7A&#39;, {&#39;weight&#39;: 0.8382334505662205}), (&#39;A2M&#39;, &#39;SLIT2&#39;, {&#39;weight&#39;: 0.8041515706739126})]
No self-loops in the graph.
Average clustering coefficient: 0.1045798392696385
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">biweight_graph</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Biweight MidCorrelation Network (Threshold = 0.8)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="network-cleaning">
<h1>Network Cleaning<a class="headerlink" href="#network-cleaning" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Graph Information:&quot;</span><span class="p">)</span>
<span class="n">print_graph_info</span><span class="p">(</span><span class="n">pearson_graph</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Clustering Coefficient:&quot;</span><span class="p">)</span>
<span class="n">calculate_average_clustering</span><span class="p">(</span><span class="n">pearson_graph</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Graph Information:
Number of nodes: 5338
Number of edges: 13449
Sample nodes: [&#39;A2M&#39;, &#39;A2ML1&#39;, &#39;A4GALT&#39;, &#39;AACSP1&#39;, &#39;AADAC&#39;, &#39;AADACP1&#39;, &#39;AARD&#39;, &#39;AASS&#39;, &#39;AATBC&#39;, &#39;AATK&#39;]
Sample edges: [(&#39;A2M&#39;, &#39;ABCA8&#39;, {&#39;weight&#39;: 0.8308274641648915}), (&#39;A2M&#39;, &#39;ADAMTS8&#39;, {&#39;weight&#39;: 0.8036704002698273}), (&#39;A2M&#39;, &#39;AOC3&#39;, {&#39;weight&#39;: 0.8767411078325781}), (&#39;A2M&#39;, &#39;C7&#39;, {&#39;weight&#39;: 0.8013688242083488}), (&#39;A2M&#39;, &#39;INMT&#39;, {&#39;weight&#39;: 0.8117391230502314}), (&#39;A2M&#39;, &#39;ITGA8&#39;, {&#39;weight&#39;: 0.8724141478160804}), (&#39;A2M&#39;, &#39;MFAP4&#39;, {&#39;weight&#39;: 0.8444697225886688}), (&#39;A2M&#39;, &#39;PKNOX2&#39;, {&#39;weight&#39;: 0.8062841461653377}), (&#39;A2M&#39;, &#39;SCN7A&#39;, {&#39;weight&#39;: 0.8382334505662205}), (&#39;A2M&#39;, &#39;SLIT2&#39;, {&#39;weight&#39;: 0.8041515706739126})]
No self-loops in the graph.

Clustering Coefficient:
Average clustering coefficient: 0.1045798392696385
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">clean_graph</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">degree_threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">keep_largest_component</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Cleans the graph by performing several cleaning steps:</span>
<span class="sd">    - Removes unconnected nodes (isolates)</span>
<span class="sd">    - Removes self-loops</span>
<span class="sd">    - Removes nodes with a degree below a specified threshold</span>
<span class="sd">    - Keeps only the largest connected component (optional)</span>

<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): The NetworkX graph to clean.</span>
<span class="sd">    degree_threshold (int): Minimum degree for nodes to keep.</span>
<span class="sd">    keep_largest_component (bool): Whether to keep only the largest connected component.</span>

<span class="sd">    Returns:</span>
<span class="sd">    G (nx.Graph): Cleaned graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Remove self-loops</span>
    <span class="n">G</span><span class="o">.</span><span class="n">remove_edges_from</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">selfloop_edges</span><span class="p">(</span><span class="n">G</span><span class="p">))</span>

    <span class="c1"># Remove nodes with no edges (isolates)</span>
    <span class="n">G</span><span class="o">.</span><span class="n">remove_nodes_from</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">isolates</span><span class="p">(</span><span class="n">G</span><span class="p">)))</span>

    <span class="c1"># Remove nodes with degree below the threshold</span>
    <span class="n">low_degree_nodes</span> <span class="o">=</span> <span class="p">[</span><span class="n">node</span> <span class="k">for</span> <span class="n">node</span><span class="p">,</span> <span class="n">degree</span> <span class="ow">in</span> <span class="nb">dict</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">degree</span><span class="p">())</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">degree</span> <span class="o">&lt;</span> <span class="n">degree_threshold</span><span class="p">]</span>
    <span class="n">G</span><span class="o">.</span><span class="n">remove_nodes_from</span><span class="p">(</span><span class="n">low_degree_nodes</span><span class="p">)</span>

    <span class="c1"># Keep only the largest connected component</span>
    <span class="k">if</span> <span class="n">keep_largest_component</span><span class="p">:</span>
        <span class="n">largest_cc</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">connected_components</span><span class="p">(</span><span class="n">G</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="nb">len</span><span class="p">)</span>
        <span class="n">G</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">subgraph</span><span class="p">(</span><span class="n">largest_cc</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clean the graph by removing unconnected nodes</span>
<span class="n">pearson_graph_cleaned</span> <span class="o">=</span> <span class="n">clean_graph</span><span class="p">(</span><span class="n">pearson_graph</span><span class="p">,</span>
                                    <span class="n">degree_threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                                    <span class="n">keep_largest_component</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">pearson_graph_cleaned</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Correlation Network - Cleaned&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/14afac0911d491b474d925364fe4e55458fef0daa81595ee5e598a1bed51dd61.png" src="../_images/14afac0911d491b474d925364fe4e55458fef0daa81595ee5e598a1bed51dd61.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_graph_info</span><span class="p">(</span><span class="n">pearson_graph_cleaned</span><span class="p">)</span>

<span class="n">calculate_average_clustering</span><span class="p">(</span><span class="n">pearson_graph_cleaned</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 1114
Number of edges: 13449
Sample nodes: [&#39;A2M&#39;, &#39;ABCA8&#39;, &#39;ABCC9&#39;, &#39;ACKR1&#39;, &#39;ACKR2&#39;, &#39;ACOXL&#39;, &#39;ACOXL-AS1&#39;, &#39;ACTG2&#39;, &#39;ADAM12&#39;, &#39;ADAMTS12&#39;]
Sample edges: [(&#39;A2M&#39;, &#39;ABCA8&#39;, {&#39;weight&#39;: 0.8308274641648915}), (&#39;A2M&#39;, &#39;ADAMTS8&#39;, {&#39;weight&#39;: 0.8036704002698273}), (&#39;A2M&#39;, &#39;AOC3&#39;, {&#39;weight&#39;: 0.8767411078325781}), (&#39;A2M&#39;, &#39;C7&#39;, {&#39;weight&#39;: 0.8013688242083488}), (&#39;A2M&#39;, &#39;INMT&#39;, {&#39;weight&#39;: 0.8117391230502314}), (&#39;A2M&#39;, &#39;ITGA8&#39;, {&#39;weight&#39;: 0.8724141478160804}), (&#39;A2M&#39;, &#39;MFAP4&#39;, {&#39;weight&#39;: 0.8444697225886688}), (&#39;A2M&#39;, &#39;PKNOX2&#39;, {&#39;weight&#39;: 0.8062841461653377}), (&#39;A2M&#39;, &#39;SCN7A&#39;, {&#39;weight&#39;: 0.8382334505662205}), (&#39;A2M&#39;, &#39;SLIT2&#39;, {&#39;weight&#39;: 0.8041515706739126})]
No self-loops in the graph.
Average clustering coefficient: 0.5011195529814455
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clean the graph by removing unconnected nodes</span>
<span class="n">pearson_graph_pruned</span> <span class="o">=</span> <span class="n">clean_graph</span><span class="p">(</span><span class="n">pearson_graph</span><span class="p">,</span>
                                    <span class="n">degree_threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                    <span class="n">keep_largest_component</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Correlation Network - Pruned&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/3cef0f71b20e1ce9dab0b998dc07cd3cc548ffaefc417cb26ff4c440dd0e63bd.png" src="../_images/3cef0f71b20e1ce9dab0b998dc07cd3cc548ffaefc417cb26ff4c440dd0e63bd.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#example</span>
<span class="n">nx</span><span class="o">.</span><span class="n">write_gml</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s2">&quot;gene_coexpression_network.gml&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">verify_edge_weights</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">num_samples</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sample of 10 edges with weights:
Edge (MARCO, CD300C) - weight: 0.8193032288783177
Edge (MARCO, HK3) - weight: 0.8074445362801838
Edge (MARCO, OSCAR) - weight: 0.8179747063198598
Edge (MARCO, PELATON) - weight: 0.8086577748898539
Edge (MARCO, VSIG4) - weight: 0.8109330062916816
Edge (CD19, BLK) - weight: 0.858415502990471
Edge (CD19, CD27) - weight: 0.845177134942465
Edge (CD19, CD79A) - weight: 0.883045209072919
Edge (CD19, CD79B) - weight: 0.8958401295059505
Edge (CD19, CLEC17A) - weight: 0.8778303265432529
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_graph_info</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>

<span class="n">calculate_average_clustering</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 305
Number of edges: 4033
Sample nodes: [&#39;MARCO&#39;, &#39;CD19&#39;, &#39;IGHV1-46&#39;, &#39;C5AR1&#39;, &#39;HLA-DQA1&#39;, &#39;HCK&#39;, &#39;IGKV3OR2-268&#39;, &#39;LILRB4&#39;, &#39;PPP1R16B&#39;, &#39;IGHV3-15&#39;]
Sample edges: [(&#39;MARCO&#39;, &#39;CD300C&#39;, {&#39;weight&#39;: 0.8193032288783177}), (&#39;MARCO&#39;, &#39;HK3&#39;, {&#39;weight&#39;: 0.8074445362801838}), (&#39;MARCO&#39;, &#39;OSCAR&#39;, {&#39;weight&#39;: 0.8179747063198598}), (&#39;MARCO&#39;, &#39;PELATON&#39;, {&#39;weight&#39;: 0.8086577748898539}), (&#39;MARCO&#39;, &#39;VSIG4&#39;, {&#39;weight&#39;: 0.8109330062916816}), (&#39;CD19&#39;, &#39;BLK&#39;, {&#39;weight&#39;: 0.858415502990471}), (&#39;CD19&#39;, &#39;CD27&#39;, {&#39;weight&#39;: 0.845177134942465}), (&#39;CD19&#39;, &#39;CD79A&#39;, {&#39;weight&#39;: 0.883045209072919}), (&#39;CD19&#39;, &#39;CD79B&#39;, {&#39;weight&#39;: 0.8958401295059505}), (&#39;CD19&#39;, &#39;CLEC17A&#39;, {&#39;weight&#39;: 0.8778303265432529})]
No self-loops in the graph.
Average clustering coefficient: 0.6766781614263463
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># visualize_interactive_network(pearson_graph_cleaned, output_file=&#39;pearson_network_cleaned.html&#39;)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="sparsification-methods">
<h1>Sparsification methods<a class="headerlink" href="#sparsification-methods" title="Link to this heading">#</a></h1>
<ul class="simple">
<li><p>Edge Sampling: Randomly removes a fraction of edges.</p></li>
<li><p>Thresholding: Removes edges with weights below a certain threshold.</p></li>
<li><p>Degree-based Sparsification</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">check_edge_attributes</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">attribute</span><span class="o">=</span><span class="s1">&#39;weight&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Checks if all edges have the specified attribute.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): The NetworkX graph.</span>
<span class="sd">    attribute (str): The edge attribute to check for.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    bool: True if all edges have the attribute, False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">missing_attributes</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">for</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">attribute</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">G</span><span class="p">[</span><span class="n">u</span><span class="p">][</span><span class="n">v</span><span class="p">]:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Edge (</span><span class="si">{</span><span class="n">u</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">v</span><span class="si">}</span><span class="s2">) is missing the &#39;</span><span class="si">{</span><span class="n">attribute</span><span class="si">}</span><span class="s2">&#39; attribute.&quot;</span><span class="p">)</span>
            <span class="n">missing_attributes</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">return</span> <span class="ow">not</span> <span class="n">missing_attributes</span>

<span class="k">def</span> <span class="nf">check_self_loops_and_isolates</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Checks and prints the number of self-loops and isolates in the graph.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): The NetworkX graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">self_loops</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">selfloop_edges</span><span class="p">(</span><span class="n">G</span><span class="p">))</span>
    <span class="n">isolates</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">isolates</span><span class="p">(</span><span class="n">G</span><span class="p">))</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of self-loops: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">self_loops</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">self_loops</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Self-loops: </span><span class="si">{</span><span class="n">self_loops</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of isolates: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">isolates</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">isolates</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Isolates: </span><span class="si">{</span><span class="n">isolates</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">plot_degree_distribution</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the degree distribution of the graph using Seaborn.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): The NetworkX graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">degrees</span> <span class="o">=</span> <span class="p">[</span><span class="n">d</span> <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">degree</span><span class="p">()]</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">degrees</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Degree Distribution&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Degree&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Graph Information:&quot;</span><span class="p">)</span>
<span class="n">print_graph_info</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Edge Attribute Check:&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="n">check_edge_attributes</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="s1">&#39;weight&#39;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;All edges have the &#39;weight&#39; attribute.&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Some edges are missing the &#39;weight&#39; attribute.&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Self-Loops and Isolates Check:&quot;</span><span class="p">)</span>
<span class="n">check_self_loops_and_isolates</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Degree Distribution:&quot;</span><span class="p">)</span>
<span class="n">plot_degree_distribution</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Clustering Coefficient:&quot;</span><span class="p">)</span>
<span class="n">calculate_average_clustering</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Graph Information:
Number of nodes: 305
Number of edges: 4033
Sample nodes: [&#39;MARCO&#39;, &#39;CD19&#39;, &#39;IGHV1-46&#39;, &#39;C5AR1&#39;, &#39;HLA-DQA1&#39;, &#39;HCK&#39;, &#39;IGKV3OR2-268&#39;, &#39;LILRB4&#39;, &#39;PPP1R16B&#39;, &#39;IGHV3-15&#39;]
Sample edges: [(&#39;MARCO&#39;, &#39;CD300C&#39;, {&#39;weight&#39;: 0.8193032288783177}), (&#39;MARCO&#39;, &#39;HK3&#39;, {&#39;weight&#39;: 0.8074445362801838}), (&#39;MARCO&#39;, &#39;OSCAR&#39;, {&#39;weight&#39;: 0.8179747063198598}), (&#39;MARCO&#39;, &#39;PELATON&#39;, {&#39;weight&#39;: 0.8086577748898539}), (&#39;MARCO&#39;, &#39;VSIG4&#39;, {&#39;weight&#39;: 0.8109330062916816}), (&#39;CD19&#39;, &#39;BLK&#39;, {&#39;weight&#39;: 0.858415502990471}), (&#39;CD19&#39;, &#39;CD27&#39;, {&#39;weight&#39;: 0.845177134942465}), (&#39;CD19&#39;, &#39;CD79A&#39;, {&#39;weight&#39;: 0.883045209072919}), (&#39;CD19&#39;, &#39;CD79B&#39;, {&#39;weight&#39;: 0.8958401295059505}), (&#39;CD19&#39;, &#39;CLEC17A&#39;, {&#39;weight&#39;: 0.8778303265432529})]
No self-loops in the graph.

Edge Attribute Check:
All edges have the &#39;weight&#39; attribute.

Self-Loops and Isolates Check:
Number of self-loops: 0
Number of isolates: 0

Degree Distribution:
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/skamp/.conda/envs/ismbenv/lib/python3.12/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div>
</div>
<img alt="../_images/98ed257f4b265bc811ece72b015764ef4b93ad9fb1e444ba65c6efce978ee9cd.png" src="../_images/98ed257f4b265bc811ece72b015764ef4b93ad9fb1e444ba65c6efce978ee9cd.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Clustering Coefficient:
Average clustering coefficient: 0.6766781614263463
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">handle_missing_weights</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">default_weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Assigns a default weight to edges that are missing the &#39;weight&#39; attribute.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    G (nx.Graph): The NetworkX graph.</span>
<span class="sd">    default_weight (float): The default weight to assign.</span>

<span class="sd">    Returns:</span>
<span class="sd">    G (nx.Graph): The graph with all edges having weights.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">():</span>
        <span class="k">if</span> <span class="s1">&#39;weight&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">G</span><span class="p">[</span><span class="n">u</span><span class="p">][</span><span class="n">v</span><span class="p">]:</span>
            <span class="n">G</span><span class="p">[</span><span class="n">u</span><span class="p">][</span><span class="n">v</span><span class="p">][</span><span class="s1">&#39;weight&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">default_weight</span>
    <span class="k">return</span> <span class="n">G</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># pearson_graph_pruned = handle_missing_weights(pearson_graph_pruned, default_weight=1.0)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">visualize_edge_weight_distribution</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualizes the distribution of edge weights using a histogram and KDE plot.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    edge_weights (list): List of edge weights.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">edge_weights</span> <span class="o">=</span> <span class="p">[</span><span class="n">G</span><span class="p">[</span><span class="n">u</span><span class="p">][</span><span class="n">v</span><span class="p">][</span><span class="s1">&#39;weight&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">()]</span>
    <span class="c1"># Histogram</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">edge_weights</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">kde</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Distribution of Edge Weights&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Edge Weight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the distribution of edge weights</span>
<span class="n">visualize_edge_weight_distribution</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/skamp/.conda/envs/ismbenv/lib/python3.12/site-packages/seaborn/_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  with pd.option_context(&#39;mode.use_inf_as_na&#39;, True):
</pre></div>
</div>
<img alt="../_images/37d213fe14fb8fcf9e29f9aebcd375cfb81490304954b2089277729fa589d497.png" src="../_images/37d213fe14fb8fcf9e29f9aebcd375cfb81490304954b2089277729fa589d497.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">analyze_and_plot_density</span><span class="p">(</span><span class="n">graph</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates and plots the density of the graph for a predefined series of thresholds.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>

<span class="sd">    Returns:</span>
<span class="sd">    densities (list of float): Densities of the graph at each threshold.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">thresholds</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.7</span> <span class="o">+</span> <span class="n">i</span> <span class="o">*</span> <span class="mf">0.01</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">31</span><span class="p">)]</span>
    <span class="n">densities</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">threshold</span> <span class="ow">in</span> <span class="n">thresholds</span><span class="p">:</span>
        <span class="n">filtered_edges</span> <span class="o">=</span> <span class="p">[(</span><span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">graph</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">if</span> <span class="n">d</span><span class="p">[</span><span class="s1">&#39;weight&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">threshold</span><span class="p">]</span>
        <span class="n">temp_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
        <span class="n">temp_graph</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">(</span><span class="n">filtered_edges</span><span class="p">)</span>
        <span class="n">densities</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nx</span><span class="o">.</span><span class="n">density</span><span class="p">(</span><span class="n">temp_graph</span><span class="p">))</span>

    <span class="c1"># Plot the densities</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">thresholds</span><span class="p">,</span> <span class="n">densities</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Threshold&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Density&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Density vs. Threshold&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">densities</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">densities</span> <span class="o">=</span> <span class="n">analyze_and_plot_density</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/b637e9e19ecaf06f129d576eaa8e69e8ed2f5bfad6413b2c8e6d77e7e787d97b.png" src="../_images/b637e9e19ecaf06f129d576eaa8e69e8ed2f5bfad6413b2c8e6d77e7e787d97b.png" />
</div>
</div>
</section>
<section id="todo-show-connected-components-of-the-plot-above">
<h1>TODO: show connected components of the plot above<a class="headerlink" href="#todo-show-connected-components-of-the-plot-above" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">threshold_sparsification</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">threshold</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sparsifies the graph by removing edges below the specified weight threshold.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>
<span class="sd">    threshold (float): The weight threshold.</span>

<span class="sd">    Returns:</span>
<span class="sd">    nx.Graph: The sparsified graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sparsified_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
    <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">add_nodes_from</span><span class="p">(</span><span class="n">graph</span><span class="o">.</span><span class="n">nodes</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
    <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">((</span><span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">d</span><span class="p">)</span> <span class="k">for</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">graph</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">if</span> <span class="n">d</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;weight&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">threshold</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sparsified_graph</span>

<span class="k">def</span> <span class="nf">top_percentage_sparsification</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">top_percentage</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sparsifies the graph by keeping the top percentage of edges by weight.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>
<span class="sd">    top_percentage (float): The percentage of top-weight edges to keep.</span>

<span class="sd">    Returns:</span>
<span class="sd">    nx.Graph: The sparsified graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sorted_edges</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">graph</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;weight&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">top_edges_count</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sorted_edges</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">top_percentage</span> <span class="o">/</span> <span class="mi">100</span><span class="p">)))</span>
    <span class="n">sparsified_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
    <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">add_nodes_from</span><span class="p">(</span><span class="n">graph</span><span class="o">.</span><span class="n">nodes</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
    <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">(</span><span class="n">sorted_edges</span><span class="p">[:</span><span class="n">top_edges_count</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">sparsified_graph</span>

<span class="k">def</span> <span class="nf">remove_by_closeness</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">threshold</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sparsifies the graph by removing nodes with closeness centrality below the specified threshold.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>
<span class="sd">    threshold (float): The closeness centrality threshold.</span>

<span class="sd">    Returns:</span>
<span class="sd">    nx.Graph: The sparsified graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">closeness</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">closeness_centrality</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>
    <span class="n">nodes_to_remove</span> <span class="o">=</span> <span class="p">[</span><span class="n">node</span> <span class="k">for</span> <span class="n">node</span><span class="p">,</span> <span class="n">centrality</span> <span class="ow">in</span> <span class="n">closeness</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">centrality</span> <span class="o">&lt;</span> <span class="n">threshold</span><span class="p">]</span>
    
    <span class="n">sparsified_graph</span> <span class="o">=</span> <span class="n">graph</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">remove_nodes_from</span><span class="p">(</span><span class="n">nodes_to_remove</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sparsified_graph</span>

<span class="k">def</span> <span class="nf">remove_by_degree</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">min_degree</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sparsifies the graph by removing nodes with degree below the specified threshold.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>
<span class="sd">    min_degree (int): The minimum degree threshold.</span>

<span class="sd">    Returns:</span>
<span class="sd">    nx.Graph: The sparsified graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">nodes_to_remove</span> <span class="o">=</span> <span class="p">[</span><span class="n">node</span> <span class="k">for</span> <span class="n">node</span><span class="p">,</span> <span class="n">degree</span> <span class="ow">in</span> <span class="nb">dict</span><span class="p">(</span><span class="n">graph</span><span class="o">.</span><span class="n">degree</span><span class="p">())</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">degree</span> <span class="o">&lt;</span> <span class="n">min_degree</span><span class="p">]</span>
    
    <span class="n">sparsified_graph</span> <span class="o">=</span> <span class="n">graph</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">remove_nodes_from</span><span class="p">(</span><span class="n">nodes_to_remove</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sparsified_graph</span>

<span class="k">def</span> <span class="nf">knn_sparsification</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">k</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sparsifies the graph by keeping only the top-k edges with the highest weights for each node.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>
<span class="sd">    k (int): The number of nearest neighbors to keep for each node.</span>

<span class="sd">    Returns:</span>
<span class="sd">    nx.Graph: The sparsified graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sparsified_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
    <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">add_nodes_from</span><span class="p">(</span><span class="n">graph</span><span class="o">.</span><span class="n">nodes</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
    
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">graph</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
        <span class="n">edges</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">graph</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;weight&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">add_edges_from</span><span class="p">(</span><span class="n">edges</span><span class="p">[:</span><span class="n">k</span><span class="p">])</span>
    
    <span class="k">return</span> <span class="n">sparsified_graph</span>

<span class="k">def</span> <span class="nf">spanning_tree_sparsification</span><span class="p">(</span><span class="n">graph</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sparsifies the graph by creating a minimum spanning tree.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>

<span class="sd">    Returns:</span>
<span class="sd">    nx.Graph: The sparsified graph.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">nx</span><span class="o">.</span><span class="n">minimum_spanning_tree</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="s1">&#39;weight&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>#TODO: pros and cons of different methods, show the distribution next to each network</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Initialize a dictionary to store graphs</span>
<span class="n">graphs</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pearson_graph_pruned</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>  <span class="c1"># Store the original graph</span>

<span class="c1"># Apply sparsification methods</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;threshold&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">threshold_sparsification</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.82</span><span class="p">)</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;top_10_percent&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">top_percentage_sparsification</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="n">top_percentage</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;closeness_below_0.3&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">remove_by_closeness</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;degree_below_3&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">remove_by_degree</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="n">min_degree</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;knn_5&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">knn_sparsification</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="n">k</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;spanning_tree&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spanning_tree_sparsification</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">])</span>


<span class="c1"># Example visualization</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="s1">&#39;Original Graph&#39;</span><span class="p">)</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;threshold&#39;</span><span class="p">],</span> <span class="s1">&#39;Thresholded Graph (weight &gt; 0.82)&#39;</span><span class="p">)</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;top_10_percent&#39;</span><span class="p">],</span> <span class="s1">&#39;Top 10</span><span class="si">% E</span><span class="s1">dges by Weight&#39;</span><span class="p">)</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;closeness_below_0.3&#39;</span><span class="p">],</span> <span class="s1">&#39;Closeness Centrality Below 0.3&#39;</span><span class="p">)</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;degree_below_3&#39;</span><span class="p">],</span> <span class="s1">&#39;Degree Below 3&#39;</span><span class="p">)</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;knn_5&#39;</span><span class="p">],</span> <span class="s1">&#39;K-Nearest Neighbors (k=5)&#39;</span><span class="p">)</span>
<span class="n">visualize_graph</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;spanning_tree&#39;</span><span class="p">],</span> <span class="s1">&#39;Minimum Spanning Tree&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/3d7ac31c2051d5a0e6a12a554ae4b48492d78da440b9706a8c00aca43b1f4094.png" src="../_images/3d7ac31c2051d5a0e6a12a554ae4b48492d78da440b9706a8c00aca43b1f4094.png" />
<img alt="../_images/097c3953caca9c8468709c64658e4178827f5945c8c083796dbd856eb1bdba25.png" src="../_images/097c3953caca9c8468709c64658e4178827f5945c8c083796dbd856eb1bdba25.png" />
<img alt="../_images/927f133af858e07d902644f2669c65051b7c63d556c1ae7f3b5b61c39f85da14.png" src="../_images/927f133af858e07d902644f2669c65051b7c63d556c1ae7f3b5b61c39f85da14.png" />
<img alt="../_images/dd3108a5da4c09b0f1602c5e092a5979c2eee3c0f40553c9fe2a7f3624972fc8.png" src="../_images/dd3108a5da4c09b0f1602c5e092a5979c2eee3c0f40553c9fe2a7f3624972fc8.png" />
<img alt="../_images/158b6d6d346542dfc480587a99849865e47c14839eef73440354275173b1516a.png" src="../_images/158b6d6d346542dfc480587a99849865e47c14839eef73440354275173b1516a.png" />
<img alt="../_images/32a84cfedf994d7ec48988c38e762681af70afdbabbd8efcc99c08d7cc0361a8.png" src="../_images/32a84cfedf994d7ec48988c38e762681af70afdbabbd8efcc99c08d7cc0361a8.png" />
<img alt="../_images/d4a8ecf26d92d513290d1614d8141ac12cf2ee6b0c899cec539af1f10458a04f.png" src="../_images/d4a8ecf26d92d513290d1614d8141ac12cf2ee6b0c899cec539af1f10458a04f.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">analyze_knn_effect</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">k_values</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Analyzes the effect of different k values on the network properties.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The original NetworkX graph.</span>
<span class="sd">    k_values (list): List of k values to use for sparsification.</span>

<span class="sd">    Returns:</span>
<span class="sd">    pd.DataFrame: DataFrame containing the analysis results.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;k&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;num_edges&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;avg_degree&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;avg_clustering&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;num_connected_components&#39;</span><span class="p">:</span> <span class="p">[],</span>
    <span class="p">}</span>
    
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">k_values</span><span class="p">:</span>
        <span class="n">sparsified_graph</span> <span class="o">=</span> <span class="n">knn_sparsification</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span>
        <span class="n">num_edges</span> <span class="o">=</span> <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">number_of_edges</span><span class="p">()</span>
        <span class="n">avg_degree</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">sparsified_graph</span><span class="o">.</span><span class="n">degree</span><span class="p">())</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">/</span> <span class="n">sparsified_graph</span><span class="o">.</span><span class="n">number_of_nodes</span><span class="p">()</span>
        <span class="n">avg_clustering</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">average_clustering</span><span class="p">(</span><span class="n">sparsified_graph</span><span class="p">)</span>
        <span class="n">num_connected_components</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">number_connected_components</span><span class="p">(</span><span class="n">sparsified_graph</span><span class="p">)</span>
        
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;k&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;num_edges&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num_edges</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;avg_degree&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">avg_degree</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;avg_clustering&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">avg_clustering</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;num_connected_components&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">num_connected_components</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">plot_knn_analysis</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the analysis of the effect of different k values on network properties.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    df (pd.DataFrame): DataFrame containing the analysis results.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
    
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;k&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;num_edges&#39;</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Number of Edges vs k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Edges&#39;</span><span class="p">)</span>
    
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;k&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;avg_degree&#39;</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Average Degree vs k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Average Degree&#39;</span><span class="p">)</span>
    
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;k&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;avg_clustering&#39;</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Average Clustering Coefficient vs k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Average Clustering Coefficient&#39;</span><span class="p">)</span>
    
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;k&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;num_connected_components&#39;</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Number of Connected Components vs k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Connected Components&#39;</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>


<span class="n">k_values</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">11</span><span class="p">))</span>  <span class="c1"># Different k values to analyze</span>
<span class="n">analysis_results</span> <span class="o">=</span> <span class="n">analyze_knn_effect</span><span class="p">(</span><span class="n">graphs</span><span class="p">[</span><span class="s1">&#39;original&#39;</span><span class="p">],</span> <span class="n">k_values</span><span class="p">)</span>

<span class="c1"># Plot the analysis results</span>
<span class="n">plot_knn_analysis</span><span class="p">(</span><span class="n">analysis_results</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/569ec47d3f9a603a2cf5f89c8d33996a219f0e923460328f93f4292a06ee204a.png" src="../_images/569ec47d3f9a603a2cf5f89c8d33996a219f0e923460328f93f4292a06ee204a.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_highest_degree_nodes</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns the nodes with the highest degree in the graph.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The NetworkX graph.</span>
<span class="sd">    top_n (int): The number of top nodes to return.</span>

<span class="sd">    Returns:</span>
<span class="sd">    List of tuples: Each tuple contains a node and its degree.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">degrees</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">graph</span><span class="o">.</span><span class="n">degree</span><span class="p">())</span>
    <span class="n">sorted_degrees</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">degrees</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sorted_degrees</span><span class="p">[:</span><span class="n">top_n</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">fetch_gene_info</span><span class="p">(</span><span class="n">gene_list</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetches gene information from MyGene.info.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    gene_list (list): List of gene symbols or Ensembl IDs.</span>

<span class="sd">    Returns:</span>
<span class="sd">    list: List of dictionaries containing gene information.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mg</span> <span class="o">=</span> <span class="n">mygene</span><span class="o">.</span><span class="n">MyGeneInfo</span><span class="p">()</span>
    <span class="n">gene_info</span> <span class="o">=</span> <span class="n">mg</span><span class="o">.</span><span class="n">querymany</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">scopes</span><span class="o">=</span><span class="s1">&#39;symbol,ensembl.gene&#39;</span><span class="p">,</span> 
                             <span class="n">fields</span><span class="o">=</span><span class="s1">&#39;name,symbol,entrezgene,summary,disease,pathway&#39;</span><span class="p">,</span> 
                             <span class="n">species</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">gene_info</span>

<span class="k">def</span> <span class="nf">print_gene_info_with_degree</span><span class="p">(</span><span class="n">top_genes_with_degrees</span><span class="p">,</span> <span class="n">gene_info</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prints gene information including the degree.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    top_genes_with_degrees (list): List of tuples containing gene symbols and their degrees.</span>
<span class="sd">    gene_info (list): List of dictionaries containing gene information.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">degree</span> <span class="ow">in</span> <span class="n">top_genes_with_degrees</span><span class="p">:</span>
        <span class="n">info</span> <span class="o">=</span> <span class="nb">next</span><span class="p">((</span><span class="n">item</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">gene_info</span> <span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="s1">&#39;query&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">),</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">info</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene Symbol: </span><span class="si">{</span><span class="n">info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;symbol&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;N/A&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Degree: </span><span class="si">{</span><span class="n">degree</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene Name: </span><span class="si">{</span><span class="n">info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;N/A&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Entrez ID: </span><span class="si">{</span><span class="n">info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;entrezgene&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;N/A&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Summary: </span><span class="si">{</span><span class="n">info</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;summary&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;N/A&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s1">&#39;disease&#39;</span> <span class="ow">in</span> <span class="n">info</span><span class="p">:</span>
                <span class="n">diseases</span> <span class="o">=</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">d</span><span class="p">[</span><span class="s1">&#39;term&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">info</span><span class="p">[</span><span class="s1">&#39;disease&#39;</span><span class="p">]])</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Diseases: </span><span class="si">{</span><span class="n">diseases</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Diseases: N/A&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s1">&#39;pathway&#39;</span> <span class="ow">in</span> <span class="n">info</span><span class="p">:</span>
                <span class="n">pathways</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">info</span><span class="p">[</span><span class="s1">&#39;pathway&#39;</span><span class="p">],</span> <span class="nb">dict</span><span class="p">):</span>
                    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">info</span><span class="p">[</span><span class="s1">&#39;pathway&#39;</span><span class="p">]:</span>
                        <span class="n">pathway_data</span> <span class="o">=</span> <span class="n">info</span><span class="p">[</span><span class="s1">&#39;pathway&#39;</span><span class="p">][</span><span class="n">key</span><span class="p">]</span>
                        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pathway_data</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
                            <span class="n">pathways</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">p</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">pathway_data</span> <span class="k">if</span> <span class="s1">&#39;name&#39;</span> <span class="ow">in</span> <span class="n">p</span><span class="p">])</span>
                        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pathway_data</span><span class="p">,</span> <span class="nb">dict</span><span class="p">)</span> <span class="ow">and</span> <span class="s1">&#39;name&#39;</span> <span class="ow">in</span> <span class="n">pathway_data</span><span class="p">:</span>
                            <span class="n">pathways</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pathway_data</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">])</span>
                        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pathway_data</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                            <span class="n">pathways</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pathway_data</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Pathways: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pathways</span><span class="p">)</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">pathways</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;N/A&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Pathways: N/A&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="mi">40</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene not found: </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Degree: </span><span class="si">{</span><span class="n">degree</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="mi">40</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">top_genes_with_degrees</span> <span class="o">=</span> <span class="n">get_highest_degree_nodes</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">gene_symbols</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">degree</span> <span class="ow">in</span> <span class="n">top_genes_with_degrees</span><span class="p">]</span>

<span class="c1"># Fetch gene information</span>
<span class="n">gene_info</span> <span class="o">=</span> <span class="n">fetch_gene_info</span><span class="p">(</span><span class="n">gene_symbols</span><span class="p">)</span>

<span class="c1"># Print gene information including degree</span>
<span class="n">print_gene_info_with_degree</span><span class="p">(</span><span class="n">top_genes_with_degrees</span><span class="p">,</span> <span class="n">gene_info</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>4 input query terms found dup hits:	[(&#39;IGHJ4&#39;, 2), (&#39;IGKV3-20&#39;, 2), (&#39;IGKV3-11&#39;, 2), (&#39;IGKV1-39&#39;, 2)]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Gene Symbol: IGKJ1
Degree: 118
Gene Name: immunoglobulin kappa joining 1
Entrez ID: 28950
Summary: Predicted to be involved in adaptive immune response. Predicted to be located in extracellular region and plasma membrane. Predicted to be part of immunoglobulin complex. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: MZB1
Degree: 109
Gene Name: marginal zone B and B1 cell specific protein
Entrez ID: 51237
Summary: Involved in positive regulation of cell population proliferation. Located in cytoplasm and extracellular region. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: FCRL5
Degree: 108
Gene Name: Fc receptor like 5
Entrez ID: 83416
Summary: This gene encodes a member of the immunoglobulin receptor superfamily and the Fc-receptor like family. This gene and several other Fc receptor-like gene members are clustered on the long arm of chromosome 1. The encoded protein is a single-pass type I membrane protein and contains 8 immunoglobulin-like C2-type domains. This gene is implicated in B cell development and lymphomagenesis. Alternatively spliced transcript variants encoding different isoforms have been identified. [provided by RefSeq, Sep 2010].
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGKC
Degree: 106
Gene Name: immunoglobulin kappa constant
Entrez ID: 3514
Summary: Predicted to enable antigen binding activity and immunoglobulin receptor binding activity. Involved in retina homeostasis. Located in blood microparticle and extracellular exosome. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGHJ4
Degree: 105
Gene Name: immunoglobulin heavy joining 4
Entrez ID: 28477
Summary: N/A
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGLL5
Degree: 99
Gene Name: immunoglobulin lambda like polypeptide 5
Entrez ID: 100423062
Summary: This gene encodes one of the immunoglobulin lambda-like polypeptides. It is located within the immunoglobulin lambda locus but it does not require somatic rearrangement for expression. The first exon of this gene is unrelated to immunoglobulin variable genes; the second and third exons are the immunoglobulin lambda joining 1 and the immunoglobulin lambda constant 1 gene segments. Alternative splicing results in multiple transcript variants. [provided by RefSeq, May 2010].
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGKV3-20
Degree: 97
Gene Name: immunoglobulin kappa variable 3-20
Entrez ID: 28912
Summary: Involved in antibacterial humoral response and glomerular filtration. Located in extracellular space. Part of monomeric IgA immunoglobulin complex; pentameric IgM immunoglobulin complex; and secretory IgA immunoglobulin complex. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGKV3-11
Degree: 96
Gene Name: immunoglobulin kappa variable 3-11
Entrez ID: 28914
Summary: Predicted to be involved in immune response. Located in blood microparticle and extracellular exosome. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGKV3D-11
Degree: 95
Gene Name: immunoglobulin kappa variable 3D-11
Entrez ID: 28876
Summary: Predicted to be involved in immune response. Located in blood microparticle and extracellular exosome. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
Gene Symbol: IGKV1-39
Degree: 95
Gene Name: immunoglobulin kappa variable 1-39
Entrez ID: 28930
Summary: Predicted to be involved in immune response. Located in blood microparticle and extracellular exosome. [provided by Alliance of Genome Resources, Apr 2022]
Diseases: N/A
Pathways: N/A
----------------------------------------
</pre></div>
</div>
</div>
</div>
</section>
<section id="interactive-visualization">
<h1>Interactive Visualization<a class="headerlink" href="#interactive-visualization" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">visualize_interactive_network</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;network.html&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualizes a NetworkX graph interactively using PyVis.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    graph (nx.Graph): The NetworkX graph to visualize.</span>
<span class="sd">    output_file (str): The filename for the output HTML file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">net</span> <span class="o">=</span> <span class="n">Network</span><span class="p">(</span><span class="n">notebook</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="s2">&quot;100%&quot;</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="s2">&quot;1000px&quot;</span><span class="p">)</span>
    <span class="n">net</span><span class="o">.</span><span class="n">from_nx</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>
    <span class="n">net</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the graph interactively</span>
<span class="n">visualize_interactive_network</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span> <span class="s1">&#39;pearson_network.html&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">save_interactive_network_plotly</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;network.html&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Interactive Network&quot;</span><span class="p">):</span>
    <span class="c1"># Extract node positions</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">spring_layout</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>

    <span class="c1"># Extract edges</span>
    <span class="n">edge_x</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">edge_y</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">graph</span><span class="o">.</span><span class="n">edges</span><span class="p">():</span>
        <span class="n">x0</span><span class="p">,</span> <span class="n">y0</span> <span class="o">=</span> <span class="n">pos</span><span class="p">[</span><span class="n">edge</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
        <span class="n">x1</span><span class="p">,</span> <span class="n">y1</span> <span class="o">=</span> <span class="n">pos</span><span class="p">[</span><span class="n">edge</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
        <span class="n">edge_x</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x0</span><span class="p">)</span>
        <span class="n">edge_x</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x1</span><span class="p">)</span>
        <span class="n">edge_x</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">edge_y</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">y0</span><span class="p">)</span>
        <span class="n">edge_y</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">y1</span><span class="p">)</span>
        <span class="n">edge_y</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

    <span class="n">edge_trace</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">edge_x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">edge_y</span><span class="p">,</span>
        <span class="n">line</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#888&#39;</span><span class="p">),</span>
        <span class="n">hoverinfo</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span>
        <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines&#39;</span><span class="p">)</span>

    <span class="c1"># Extract nodes</span>
    <span class="n">node_x</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">node_y</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">node_text</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">graph</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">pos</span><span class="p">[</span><span class="n">node</span><span class="p">]</span>
        <span class="n">node_x</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="n">node_y</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
        <span class="n">node_text</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">node</span><span class="p">))</span>

    <span class="n">node_trace</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Scatter</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">node_x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">node_y</span><span class="p">,</span>
        <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;markers+text&#39;</span><span class="p">,</span>
        <span class="n">text</span><span class="o">=</span><span class="n">node_text</span><span class="p">,</span>
        <span class="n">hoverinfo</span><span class="o">=</span><span class="s1">&#39;text&#39;</span><span class="p">,</span>
        <span class="n">marker</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
            <span class="n">showscale</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">colorscale</span><span class="o">=</span><span class="s1">&#39;YlGnBu&#39;</span><span class="p">,</span>
            <span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
            <span class="n">colorbar</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
                <span class="n">thickness</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
                <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Node Connections&#39;</span><span class="p">,</span>
                <span class="n">xanchor</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span>
                <span class="n">titleside</span><span class="o">=</span><span class="s1">&#39;right&#39;</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">Figure</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">[</span><span class="n">edge_trace</span><span class="p">,</span> <span class="n">node_trace</span><span class="p">],</span>
                    <span class="n">layout</span><span class="o">=</span><span class="n">go</span><span class="o">.</span><span class="n">Layout</span><span class="p">(</span>
                        <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span>
                        <span class="n">titlefont_size</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
                        <span class="n">showlegend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                        <span class="n">hovermode</span><span class="o">=</span><span class="s1">&#39;closest&#39;</span><span class="p">,</span>
                        <span class="n">margin</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">b</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span><span class="n">l</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span><span class="n">r</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span><span class="n">t</span><span class="o">=</span><span class="mi">40</span><span class="p">),</span>
                        <span class="n">annotations</span><span class="o">=</span><span class="p">[</span> <span class="nb">dict</span><span class="p">(</span>
                            <span class="n">text</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                            <span class="n">showarrow</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                            <span class="n">xref</span><span class="o">=</span><span class="s2">&quot;paper&quot;</span><span class="p">,</span> <span class="n">yref</span><span class="o">=</span><span class="s2">&quot;paper&quot;</span>
                        <span class="p">)],</span>
                        <span class="n">xaxis</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">showgrid</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">zeroline</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
                        <span class="n">yaxis</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">showgrid</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">zeroline</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
                    <span class="p">)</span>

    <span class="c1"># Save the figure to an HTML file</span>
    <span class="n">pio</span><span class="o">.</span><span class="n">write_html</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">output_file</span><span class="p">,</span> <span class="n">auto_open</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="todo-add-hover-info-to-plotly-plot-gene-info">
<h1>TODO: add hover info to plotly plot (gene info)<a class="headerlink" href="#todo-add-hover-info-to-plotly-plot-gene-info" title="Link to this heading">#</a></h1>
</section>
<section id="todo-introduce-clustering-kmeans-etc">
<h1>TODO: introduce clustering kmeans. etc<a class="headerlink" href="#todo-introduce-clustering-kmeans-etc" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">save_interactive_network_plotly</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">output_file</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s1">&#39;genecoexp_plotly_network.html&#39;</span><span class="p">),</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Gene Coexpression Network&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="storing-generated-networks">
<h1>Storing generated networks<a class="headerlink" href="#storing-generated-networks" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#example</span>
<span class="n">nx</span><span class="o">.</span><span class="n">write_gml</span><span class="p">(</span><span class="n">pearson_graph_pruned</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s1">&#39;gene_coexpression_network.gml&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="patient-similarity-network">
<h1>Patient similarity network<a class="headerlink" href="#patient-similarity-network" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Based on the same expression matrix we can create a patient similarity network.</span>
<span class="c1"># By transposing the matrix will switch the rows and columns,</span>
<span class="c1"># meaning that patients will become the rows and genes will become the columns.</span>
<span class="c1"># By doing this, you can compute the correlation (or similarity) between patients based on their gene expression profiles,</span>
<span class="c1"># and then create a network where nodes represent patients and edges represent similarities.</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Transpose the matrix</span>
<span class="n">patient_gene_matrix</span> <span class="o">=</span> <span class="n">df_renamed</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">patient_gene_matrix</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>TCGA-38-7271</th>
      <th>TCGA-55-7914</th>
      <th>TCGA-95-7043</th>
      <th>TCGA-73-4658</th>
      <th>TCGA-86-8076</th>
      <th>TCGA-55-7726</th>
      <th>TCGA-44-6147</th>
      <th>TCGA-50-5932</th>
      <th>TCGA-44-2661</th>
      <th>TCGA-86-7954</th>
      <th>...</th>
      <th>TCGA-97-A4M7</th>
      <th>TCGA-62-A46R</th>
      <th>TCGA-50-5055</th>
      <th>TCGA-38-4628</th>
      <th>TCGA-86-7713</th>
      <th>TCGA-86-8073</th>
      <th>TCGA-MN-A4N4</th>
      <th>TCGA-53-7626</th>
      <th>TCGA-44-A47G</th>
      <th>TCGA-55-6969</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A2M</th>
      <td>17.7492</td>
      <td>14.8513</td>
      <td>14.1691</td>
      <td>16.7238</td>
      <td>15.6783</td>
      <td>14.7566</td>
      <td>16.4368</td>
      <td>15.5476</td>
      <td>15.5478</td>
      <td>15.1337</td>
      <td>...</td>
      <td>15.9553</td>
      <td>13.9511</td>
      <td>16.3097</td>
      <td>14.3934</td>
      <td>15.8254</td>
      <td>16.3773</td>
      <td>14.9411</td>
      <td>16.7343</td>
      <td>15.6622</td>
      <td>14.8136</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>4.4411</td>
      <td>4.4530</td>
      <td>4.5026</td>
      <td>3.1704</td>
      <td>4.7422</td>
      <td>6.0918</td>
      <td>4.0602</td>
      <td>3.0827</td>
      <td>3.4608</td>
      <td>3.6367</td>
      <td>...</td>
      <td>4.7076</td>
      <td>2.4821</td>
      <td>8.0810</td>
      <td>3.3415</td>
      <td>5.3379</td>
      <td>4.0817</td>
      <td>6.7210</td>
      <td>5.2575</td>
      <td>3.7345</td>
      <td>4.3058</td>
    </tr>
    <tr>
      <th>A4GALT</th>
      <td>10.1862</td>
      <td>8.9312</td>
      <td>9.0834</td>
      <td>9.1443</td>
      <td>9.5150</td>
      <td>11.4452</td>
      <td>9.1462</td>
      <td>7.3597</td>
      <td>9.1602</td>
      <td>8.9344</td>
      <td>...</td>
      <td>9.1019</td>
      <td>8.6653</td>
      <td>10.8837</td>
      <td>8.6304</td>
      <td>9.3827</td>
      <td>9.4234</td>
      <td>11.9107</td>
      <td>9.5379</td>
      <td>9.8660</td>
      <td>9.7702</td>
    </tr>
    <tr>
      <th>AACSP1</th>
      <td>3.5845</td>
      <td>3.6762</td>
      <td>3.8623</td>
      <td>2.4821</td>
      <td>3.2271</td>
      <td>2.4821</td>
      <td>5.6614</td>
      <td>3.0827</td>
      <td>3.9428</td>
      <td>4.7089</td>
      <td>...</td>
      <td>5.4947</td>
      <td>3.0956</td>
      <td>2.4821</td>
      <td>3.7710</td>
      <td>2.9397</td>
      <td>6.6519</td>
      <td>8.8283</td>
      <td>2.4821</td>
      <td>4.1045</td>
      <td>6.8425</td>
    </tr>
    <tr>
      <th>AADAC</th>
      <td>9.6415</td>
      <td>6.5685</td>
      <td>5.5634</td>
      <td>4.3206</td>
      <td>7.2321</td>
      <td>6.8821</td>
      <td>8.1176</td>
      <td>4.7935</td>
      <td>11.7750</td>
      <td>7.0644</td>
      <td>...</td>
      <td>9.0856</td>
      <td>3.8162</td>
      <td>5.6471</td>
      <td>4.6785</td>
      <td>4.6578</td>
      <td>6.7641</td>
      <td>3.4947</td>
      <td>5.7120</td>
      <td>6.2214</td>
      <td>5.2016</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>ZSCAN23</th>
      <td>5.7385</td>
      <td>4.3796</td>
      <td>5.6728</td>
      <td>3.6533</td>
      <td>5.0478</td>
      <td>6.0489</td>
      <td>5.8058</td>
      <td>3.7897</td>
      <td>4.3451</td>
      <td>5.5259</td>
      <td>...</td>
      <td>4.7076</td>
      <td>2.4821</td>
      <td>4.6683</td>
      <td>3.4697</td>
      <td>4.3516</td>
      <td>7.2074</td>
      <td>6.9167</td>
      <td>6.0927</td>
      <td>5.3854</td>
      <td>3.6792</td>
    </tr>
    <tr>
      <th>ZSCAN31</th>
      <td>9.0817</td>
      <td>10.9565</td>
      <td>10.8095</td>
      <td>8.9874</td>
      <td>9.5810</td>
      <td>9.0185</td>
      <td>10.2085</td>
      <td>9.8803</td>
      <td>11.0327</td>
      <td>10.2707</td>
      <td>...</td>
      <td>9.2033</td>
      <td>10.5020</td>
      <td>8.8073</td>
      <td>9.8672</td>
      <td>10.6419</td>
      <td>10.2410</td>
      <td>9.4672</td>
      <td>10.0507</td>
      <td>7.4843</td>
      <td>8.5268</td>
    </tr>
    <tr>
      <th>ZSCAN4</th>
      <td>3.2709</td>
      <td>3.0918</td>
      <td>2.4821</td>
      <td>3.4468</td>
      <td>3.0118</td>
      <td>5.0894</td>
      <td>5.5341</td>
      <td>3.5082</td>
      <td>5.4987</td>
      <td>3.3645</td>
      <td>...</td>
      <td>4.4088</td>
      <td>7.4937</td>
      <td>4.3167</td>
      <td>5.0907</td>
      <td>6.7676</td>
      <td>5.1126</td>
      <td>3.3143</td>
      <td>4.2816</td>
      <td>4.4650</td>
      <td>2.4821</td>
    </tr>
    <tr>
      <th>ZWINT</th>
      <td>9.2615</td>
      <td>10.1334</td>
      <td>10.9149</td>
      <td>9.7828</td>
      <td>8.6640</td>
      <td>11.1212</td>
      <td>9.0718</td>
      <td>10.7305</td>
      <td>9.0013</td>
      <td>10.4268</td>
      <td>...</td>
      <td>9.7614</td>
      <td>10.7593</td>
      <td>9.7087</td>
      <td>10.4729</td>
      <td>10.9794</td>
      <td>10.1390</td>
      <td>10.0583</td>
      <td>9.4627</td>
      <td>9.5817</td>
      <td>10.8228</td>
    </tr>
    <tr>
      <th>ZYG11A</th>
      <td>4.0073</td>
      <td>8.3219</td>
      <td>6.7593</td>
      <td>3.4468</td>
      <td>4.7844</td>
      <td>5.0894</td>
      <td>9.4014</td>
      <td>4.7935</td>
      <td>4.2748</td>
      <td>7.1709</td>
      <td>...</td>
      <td>6.5412</td>
      <td>7.1933</td>
      <td>5.8778</td>
      <td>7.3822</td>
      <td>8.7596</td>
      <td>8.2478</td>
      <td>8.4566</td>
      <td>7.4014</td>
      <td>4.6804</td>
      <td>8.6068</td>
    </tr>
  </tbody>
</table>
<p>5338 rows × 498 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dictionary to store different correlation matrices</span>
<span class="n">patient_correlation_matrices</span> <span class="o">=</span> <span class="p">{}</span>

<span class="c1"># Pearson correlation</span>
<span class="n">patient_correlation_matrices</span><span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">patient_gene_matrix</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">)</span>

<span class="n">patient_correlation_matrices</span><span class="p">[</span><span class="s1">&#39;biweight_midcorrelation&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">abs_bicorr</span><span class="p">(</span><span class="n">patient_gene_matrix</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">patient_pearson_graph</span> <span class="o">=</span> <span class="n">create_graph_from_correlation</span><span class="p">(</span><span class="n">patient_correlation_matrices</span><span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">],</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">verify_edge_weights</span><span class="p">(</span><span class="n">patient_pearson_graph</span><span class="p">,</span> <span class="n">num_samples</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sample of 10 edges with weights:
Edge (TCGA-38-7271, TCGA-73-4658) - weight: 0.8696511707884843
Edge (TCGA-38-7271, TCGA-86-8076) - weight: 0.8173779911639248
Edge (TCGA-38-7271, TCGA-44-2661) - weight: 0.8652686243802332
Edge (TCGA-38-7271, TCGA-73-4662) - weight: 0.8446135021161085
Edge (TCGA-38-7271, TCGA-55-6986) - weight: 0.833876175940164
Edge (TCGA-38-7271, TCGA-49-6744) - weight: 0.8919809060862212
Edge (TCGA-38-7271, TCGA-69-7763) - weight: 0.8374050473870502
Edge (TCGA-38-7271, TCGA-44-6774) - weight: 0.802120255633156
Edge (TCGA-38-7271, TCGA-67-3774) - weight: 0.8179727328277505
Edge (TCGA-38-7271, TCGA-97-A4M2) - weight: 0.8045412817214057
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">patient_pearson_graph</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Correlation Network (Threshold = 0.8)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/90b6aebdf03722303177be3085bc89efefcd022280def576fc88a6f3a49eccbb.png" src="../_images/90b6aebdf03722303177be3085bc89efefcd022280def576fc88a6f3a49eccbb.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clean the graph by removing unconnected nodes</span>
<span class="n">patient_pearson_graph_pruned</span> <span class="o">=</span> <span class="n">clean_graph</span><span class="p">(</span><span class="n">patient_pearson_graph</span><span class="p">,</span>
                                    <span class="n">degree_threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                    <span class="n">keep_largest_component</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Patient Correlation Network (Threshold = 0.8)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/8b8baf20e79f55c9d3f221008e57b9b97ca2c3af4562bf397f225f6442d2b394.png" src="../_images/8b8baf20e79f55c9d3f221008e57b9b97ca2c3af4562bf397f225f6442d2b394.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">patient_pearson_graph_pruned_knn</span> <span class="o">=</span> <span class="n">knn_sparsification</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_graph_info</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 434
Number of edges: 10373
Sample nodes: [&#39;TCGA-38-7271&#39;, &#39;TCGA-55-7914&#39;, &#39;TCGA-73-4658&#39;, &#39;TCGA-86-8076&#39;, &#39;TCGA-55-7726&#39;, &#39;TCGA-44-6147&#39;, &#39;TCGA-50-5932&#39;, &#39;TCGA-44-2661&#39;, &#39;TCGA-86-7954&#39;, &#39;TCGA-73-4662&#39;]
Sample edges: [(&#39;TCGA-38-7271&#39;, &#39;TCGA-73-4658&#39;, {&#39;weight&#39;: 0.8696511707884843}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-86-8076&#39;, {&#39;weight&#39;: 0.8173779911639248}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-44-2661&#39;, {&#39;weight&#39;: 0.8652686243802332}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-73-4662&#39;, {&#39;weight&#39;: 0.8446135021161085}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-55-6986&#39;, {&#39;weight&#39;: 0.833876175940164}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-49-6744&#39;, {&#39;weight&#39;: 0.8919809060862212}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-69-7763&#39;, {&#39;weight&#39;: 0.8374050473870502}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-44-6774&#39;, {&#39;weight&#39;: 0.802120255633156}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-67-3774&#39;, {&#39;weight&#39;: 0.8179727328277505}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-97-A4M2&#39;, {&#39;weight&#39;: 0.8045412817214057})]
No self-loops in the graph.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">print_graph_info</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned_knn</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes: 434
Number of edges: 2956
Sample nodes: [&#39;TCGA-38-7271&#39;, &#39;TCGA-55-7914&#39;, &#39;TCGA-73-4658&#39;, &#39;TCGA-86-8076&#39;, &#39;TCGA-55-7726&#39;, &#39;TCGA-44-6147&#39;, &#39;TCGA-50-5932&#39;, &#39;TCGA-44-2661&#39;, &#39;TCGA-86-7954&#39;, &#39;TCGA-73-4662&#39;]
Sample edges: [(&#39;TCGA-38-7271&#39;, &#39;TCGA-53-7626&#39;, {&#39;weight&#39;: 0.893251013553719}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-49-6744&#39;, {&#39;weight&#39;: 0.8919809060862212}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-78-8648&#39;, {&#39;weight&#39;: 0.885760147175422}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-86-8671&#39;, {&#39;weight&#39;: 0.8825367028721114}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-50-5941&#39;, {&#39;weight&#39;: 0.8814255405956034}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-55-7574&#39;, {&#39;weight&#39;: 0.8799541588637929}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-97-7553&#39;, {&#39;weight&#39;: 0.8775024097687647}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-50-5055&#39;, {&#39;weight&#39;: 0.8726819735362156}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-50-5045&#39;, {&#39;weight&#39;: 0.8723464616125998}), (&#39;TCGA-38-7271&#39;, &#39;TCGA-73-4658&#39;, {&#39;weight&#39;: 0.8696511707884843})]
No self-loops in the graph.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned_knn</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;K-Nearest Neighbors (k=10) Patient Correlation Network&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/42b48b25c1803c7d80587fadc4d888d7db4d96ae07f5c170196cb7089c275b12.png" src="../_images/42b48b25c1803c7d80587fadc4d888d7db4d96ae07f5c170196cb7089c275b12.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">nx</span><span class="o">.</span><span class="n">write_gml</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned_knn</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s1">&#39;patient_coexpression_network_knn.gml&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="dnameth-pn">
<h1>DNAmeth PN<a class="headerlink" href="#dnameth-pn" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the data using pickle</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s2">&quot;ISMB_TCGA_DNAm.pkl&quot;</span><span class="p">)</span> <span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span> <span class="p">:</span> 
    <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_pickle</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>

<span class="n">meth_data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;datExpr&quot;</span><span class="p">]</span>
<span class="n">meth_data</span>

<span class="n">smoking_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s2">&quot;smoking.tsv&quot;</span><span class="p">)</span> <span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="c1">#Tab seperated document from EWAS</span>
<span class="n">common_annotated_cpgs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">smoking_df</span><span class="p">[</span><span class="s1">&#39;cpg&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()[</span><span class="n">smoking_df</span><span class="p">[</span><span class="s1">&#39;cpg&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> 
<span class="n">cpgs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">common_annotated_cpgs</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">meth_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>


<span class="n">common_patients</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">patient_pearson_graph_pruned_knn</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">meth_data</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>


<span class="n">meth_data_filt</span> <span class="o">=</span> <span class="n">meth_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span> <span class="nb">list</span><span class="p">(</span><span class="n">common_patients</span><span class="p">)</span> <span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">cpgs</span><span class="p">)]</span>

<span class="n">patient_meth_matrix</span> <span class="o">=</span> <span class="n">meth_data_filt</span><span class="o">.</span><span class="n">T</span>


<span class="c1"># Keep only patients who are prevalent in the patient gene network</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">meth_data_filt</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">patient_meth_matrix</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dictionary to store different correlation matrices</span>
<span class="n">p_meth_correlation_matrices</span> <span class="o">=</span> <span class="p">{}</span>

<span class="c1"># Pearson correlation</span>
<span class="n">p_meth_correlation_matrices</span><span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">patient_meth_matrix</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">)</span>
<span class="n">p_meth_pearson_graph</span> <span class="o">=</span> <span class="n">create_graph_from_correlation</span><span class="p">(</span><span class="n">p_meth_correlation_matrices</span><span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">],</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
<span class="c1"># Clean the graph by removing unconnected nodes</span>
<span class="n">p_meth_pearson_graph_pruned</span> <span class="o">=</span> <span class="n">clean_graph</span><span class="p">(</span><span class="n">p_meth_pearson_graph</span><span class="p">,</span>
                                    <span class="n">degree_threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                    <span class="n">keep_largest_component</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">visualize_graph</span><span class="p">(</span><span class="n">p_meth_pearson_graph_pruned</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Correlation Network (Threshold = 0.8)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p_meth_pearson_graph_pruned_knn</span> <span class="o">=</span> <span class="n">knn_sparsification</span><span class="p">(</span><span class="n">p_meth_pearson_graph_pruned</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">print_graph_info</span><span class="p">(</span><span class="n">p_meth_pearson_graph_pruned_knn</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_graph</span><span class="p">(</span><span class="n">p_meth_pearson_graph_pruned_knn</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Pearson Correlation Network (Threshold = 0.8)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">nx</span><span class="o">.</span><span class="n">write_gml</span><span class="p">(</span><span class="n">p_meth_pearson_graph_pruned_knn</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span><span class="s1">&#39;patient_meth_network_knn.gml&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="setting-up-pip-network-using-string">
<h1>Setting up PIP network using String<a class="headerlink" href="#setting-up-pip-network-using-string" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># create a concatenated list of gene names as strings</span>
<span class="c1"># note we are taking gene names from the &#39;gene&#39; column of the dataframe we obtained from the EWAS catalog and filtered</span>
<span class="c1"># to common CpG and Gene sites above. The join function then concatenates them using the &#39;%0D&#39; string</span>
<span class="c1"># to stitch them all together. This string will be used to help us build the API query URL.</span>
<span class="n">external_gene_names</span> <span class="o">=</span> <span class="s1">&#39;%0D&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="n">df_filt</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">300</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()))</span>

<span class="c1"># pass the list of Gene Names to the String-DB API return the String-IDs</span>
<span class="c1"># we first form the query url using the &#39;get_string_ids&#39; API function which takes a list of identifiers and</span>
<span class="c1"># converts them into the internal String-DB accession IDs. This massively speeds up the search and allows us to</span>
<span class="c1"># search for more than 10 at once which is an API restriction for other API functions if String-DB internal accessions </span>
<span class="c1"># aren&#39;t used.</span>
<span class="n">query_url</span> <span class="o">=</span> <span class="s1">&#39;https://string-db.org/api/tsv-no-header/get_string_ids?identifiers=&#39;</span><span class="o">+</span><span class="n">external_gene_names</span><span class="o">+</span><span class="s1">&#39;&amp;species=9606&amp;format=only-ids&#39;</span>

<span class="c1"># use the urllib library to retrieve the String-DB internal IDs</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">ul</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="n">query_url</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span>

<span class="c1"># now we want to query String-DB to retrieve interactions from this list of String-DB IDs</span>
<span class="c1"># we create a concatenated list of stringdbIDs in much the same way as above for the Entrez Gene IDs</span>
<span class="n">stringdbIDs</span> <span class="o">=</span> <span class="s1">&#39;%0D&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">splitlines</span><span class="p">())</span>

<span class="c1"># again we build the query for interactions using the String-DB IDs</span>
<span class="n">query_url</span> <span class="o">=</span> <span class="s1">&#39;https://string-db.org/api/tsv/network?identifiers=&#39;</span><span class="o">+</span><span class="n">stringdbIDs</span><span class="o">+</span><span class="s1">&#39;&amp;species=9606&#39;</span>

<span class="c1"># again using urllib to retrieve the interactions these are returned in a standard tab delimied text format</span>
<span class="n">interactions</span> <span class="o">=</span> <span class="n">ul</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="n">query_url</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>

<span class="c1"># we need to split the result by these &#39;tabs&#39; (\t - is used to identfy them)</span>
<span class="n">int_test</span> <span class="o">=</span> <span class="p">[</span><span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">interaction</span> <span class="ow">in</span> <span class="n">interactions</span><span class="p">]</span>

<span class="c1"># we extract the field names from the first row</span>
<span class="n">column_names</span> <span class="o">=</span> <span class="n">int_test</span><span class="p">[:</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

<span class="c1"># create a Pandas dataframe of the interaction data we have just retrieved from String-DB</span>
<span class="n">interactions_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">int_test</span><span class="p">,</span><span class="n">columns</span><span class="o">=</span><span class="n">column_names</span><span class="p">)</span>

<span class="c1"># delete the first row that held the fieldnames but we no longer need</span>
<span class="n">interactions_df</span> <span class="o">=</span> <span class="n">interactions_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># remove any duplicate rows</span>
<span class="n">final_interactions</span> <span class="o">=</span> <span class="n">interactions_df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

<span class="c1"># show the top of the protein-protein interaction table</span>
<span class="n">final_interactions</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./Section1"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Section 1</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#structuring-biomedical-data-using-networks-workshop-network-generation">Structuring Biomedical Data Using Networks  workshop - Network Generation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#table-of-contents">Table of Contents</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#libraries">Libraries</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data">Data</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tcga-luad-project">TCGA-LUAD Project</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata-analysis">Metadata analysis</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#expression-data-analysis">Expression data analysis</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data-filtering-and-preprocessing">Data Filtering and Preprocessing</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#thresholding-based-on-gene-expression-levels">Thresholding Based on Gene Expression Levels</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-retention-at-various-thresholds">Gene Retention at Various Thresholds</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#thresholding-based-on-gene-expression-variance">Thresholding Based on Gene Expression Variance</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#instead-of-gene-expression">instead of Gene Expression</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-retention-at-different-variance-levels">Gene Retention at Different Variance Levels</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-symbol-conversion-ensg-to-hgcn">Gene Symbol Conversion: ENSG to HGCN</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#checking-for-missing-data-and-outliers">Checking for Missing Data and Outliers</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-heterogeneity">Data heterogeneity</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#correlation-matrix-calculation">Correlation Matrix Calculation</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#consider-loading-in-precomputed-correlation-matrices-and-skipping-the-next-cell">Consider loading in precomputed correlation matrices and skipping the next cell</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-check-grouping-ordering">TODO: check grouping, ordering ?</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#graph-construction-from-correlation-matrices">Graph Construction from Correlation Matrices</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-list-comprehension-for-graph-generation">TODO: list comprehension for Graph generation ??</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#show-densities-what-treshold-is-useful-to-get-rid-of-the-fully-connected-network">show densities, what treshold is useful to get rid of the fully connected network</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#network-cleaning">Network Cleaning</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#sparsification-methods">Sparsification methods</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-show-connected-components-of-the-plot-above">TODO: show connected components of the plot above</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#interactive-visualization">Interactive Visualization</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-add-hover-info-to-plotly-plot-gene-info">TODO: add hover info to plotly plot (gene info)</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#todo-introduce-clustering-kmeans-etc">TODO: introduce clustering kmeans. etc</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#storing-generated-networks">Storing generated networks</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#patient-similarity-network">Patient similarity network</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#dnameth-pn">DNAmeth PN</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#setting-up-pip-network-using-string">Setting up PIP network using String</a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By https://biomedicalinformaticsgroup.github.io/
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>