
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Part 1: Over-representation and Enrichment Analysis &#8212; ISMB Tutorial Book</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=384b581d" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=efea14e4"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=36754332"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Section2/Part1';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Part 2: Similarity Network Fusion" href="Part2.html" />
    <link rel="prev" title="Section 2: Network Analysis" href="../Section2_intro.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="ISMB Tutorial Book - Home"/>
    <script>document.write(`<img src="../_static/logo.png" class="logo__image only-dark" alt="ISMB Tutorial Book - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    An Introduction to Multi Modal Analysis using Networks
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../GettingStarted.html">Getting Started</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../Section1_intro.html">Section 1 - Network Basics</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../Section1/Section1.html">Section 1</a></li>




















</ul>
</li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../Section2_intro.html">Section 2: Network Analysis</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Part 1: Over-representation and Enrichment Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="Part2.html">Part 2: Similarity Network Fusion</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../Section3_intro.html">Section 3 - SNF + GNNâ€™s</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part1.html"><font color="darkblue"> Part 1 : Message Passing </font></a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part2.html"><font color="darkblue"> Part 2 : Problem Formulation / Patient Similarity Network Generation </font></a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part3.html"><font color="darkblue"> Part 3 : Smoking Network Classification using Graph Convolutional Networks</font></a></li>
<li class="toctree-l2"><a class="reference internal" href="../Section3/Part4.html"><font color="darkblue"> Part 4 : Similarity Network Fusion with Graph Convolutional Networks</font></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../ConceptGlossary.html">Concept Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../functions_definitions.html">Custom Function Definitions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/Barry8197/ISMB_Test" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/Barry8197/ISMB_Test/issues/new?title=Issue%20on%20page%20%2FSection2/Part1.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/Section2/Part1.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Part 1: Over-representation and Enrichment Analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#load-networks-built-in-section-1">0. Load networks built in Section 1</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-network-analysis-with-g-exp">1. Gene Network Analysis with <code class="docutils literal notranslate"><span class="pre">G_exp</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#over-representation-analysis-ora">1.1 Over Representation Analysis (ORA)</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ora-on-gene-correlation-network">1.1.1 ORA on gene correlation network</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ora-on-gene-clusters">1.1.1 ORA on gene clusters</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualising-the-communities">Visualising the Communities</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#perform-ora-on-each-community">Perform ORA on Each Community</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-set-enrichment-analysis-gsea">1.2 Gene Set Enrichment Analysis (GSEA)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#assign-classes-based-on-phenotypic-attributes-in-metadata-e-g-smoking-status">Assign classes based on phenotypic attributes in metadata (e.g., smoking status)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualising-gsea-results">Visualising GSEA Results</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#barcode-enrichment-plot">Barcode Enrichment Plot</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#heatmap-visualisation">Heatmap Visualisation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clustermap-visualisation">Clustermap Visualisation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dot-plot-visualisation">Dot Plot Visualisation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gsea-on-clusters">GSEA on Clusters</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="part-1-over-representation-and-enrichment-analysis">
<h1>Part 1: Over-representation and Enrichment Analysis<a class="headerlink" href="#part-1-over-representation-and-enrichment-analysis" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">urllib</span>
<span class="kn">import</span> <span class="nn">urllib</span> <span class="k">as</span> <span class="nn">ul</span>
<span class="kn">import</span> <span class="nn">urllib.request</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
<span class="kn">import</span> <span class="nn">ast</span>
<span class="kn">from</span> <span class="nn">prettytable</span> <span class="kn">import</span> <span class="n">PrettyTable</span>
<span class="kn">import</span> <span class="nn">gseapy</span> <span class="k">as</span> <span class="nn">gp</span>
<span class="kn">from</span> <span class="nn">palettable</span> <span class="kn">import</span> <span class="n">wesanderson</span>

<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_colwidth&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_rows&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/chaeeunlee/Documents/VSC_workspaces/ISMB_Test/Section2
</pre></div>
</div>
</div>
</div>
<section id="load-networks-built-in-section-1">
<h2>0. Load networks built in Section 1<a class="headerlink" href="#load-networks-built-in-section-1" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>Gene correlation network</p></li>
<li><p>Patient network from TCGA gene expression data</p></li>
<li><p>Patient network from TCGA DNA methylation data</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define paths to .gml network files from Section 1</span>
<span class="c1"># These paths point to the Gene correlation network, Patient network from (1) TCGA gene expression data and (2) TCGA DNA methylation data.</span>
<span class="n">G_patient_gxp_path</span> <span class="o">=</span> <span class="s1">&#39;section2_data/section2_networks_v2/patient_coexpression_network_knn.gml&#39;</span>
<span class="n">G_patient_dnam_path</span> <span class="o">=</span> <span class="s1">&#39;section2_data/section2_networks_v2/patient_meth_network_knn.gml&#39;</span>
<span class="n">G_gxp_path</span> <span class="o">=</span> <span class="s1">&#39;section2_data/section2_networks_v1/gene_coexpression_network.gml&#39;</span>

<span class="c1"># Load the GML graphs into NetworkX graph objects</span>
<span class="c1"># nx.read_gml() function reads a graph from a GML file</span>
<span class="n">G_gxp</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">read_gml</span><span class="p">(</span><span class="n">G_gxp_path</span><span class="p">)</span>  <span class="c1"># Gene correlation network</span>
<span class="n">G_patient_gxp</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">read_gml</span><span class="p">(</span><span class="n">G_patient_gxp_path</span><span class="p">)</span>  <span class="c1"># Patient gene expression network</span>
<span class="n">G_patient_dnam</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">read_gml</span><span class="p">(</span><span class="n">G_patient_dnam_path</span><span class="p">)</span>  <span class="c1"># Patient methylation network</span>

<span class="c1"># Get all nodes in each graph</span>
<span class="c1"># The nodes represent genes or patients depending on the network</span>
<span class="n">G_gxp_nodes_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">G_gxp</span><span class="o">.</span><span class="n">nodes</span><span class="p">())</span>  <span class="c1"># Nodes in the gene correlation network</span>
<span class="n">G_patient_gxp_nodes_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">G_patient_gxp</span><span class="o">.</span><span class="n">nodes</span><span class="p">())</span>  <span class="c1"># Patient nodes</span>
<span class="n">G_patient_dnam_nodes_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">G_patient_dnam</span><span class="o">.</span><span class="n">nodes</span><span class="p">())</span>  <span class="c1"># Patient nodes</span>

<span class="c1"># Define paths to the raw TCGA datasets</span>
<span class="n">tcga_dnam_path</span> <span class="o">=</span> <span class="s1">&#39;section2_data/ISMB_TCGA_DNAm.pkl&#39;</span>  <span class="c1"># TCGA DNA methylation data</span>
<span class="n">tcga_gxp_path</span> <span class="o">=</span> <span class="s1">&#39;section2_data/ISMB_TCGA_GE.pkl&#39;</span>  <span class="c1"># TCGA Gene expression data</span>

<span class="c1"># Load the gene expression dataset</span>
<span class="c1"># pd.read_pickle() function loads a pickled pandas DataFrame or Series</span>
<span class="n">tcga_gxp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_pickle</span><span class="p">(</span><span class="n">tcga_gxp_path</span><span class="p">)</span>

<span class="c1"># For this example, we&#39;ll use a CSV file that includes gene symbols</span>
<span class="c1"># pd.read_csv() function loads a CSV file into a pandas DataFrame</span>
<span class="n">tcga_gxp_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;section2_data/tcga_ge_df_symbols_t.csv&#39;</span><span class="p">)</span>  <span class="c1"># Dataset with gene symbols</span>
<span class="c1"># Set &#39;GENES&#39; column as the index for easy access to gene-specific data</span>
<span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;GENES&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Extract metadata from the gene expression dataset</span>
<span class="c1"># Metadata might include information such as patient IDs, sample conditions, etc.</span>
<span class="n">tcga_gxp_meta</span> <span class="o">=</span> <span class="n">tcga_gxp</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">]</span>

<span class="c1"># Print the number of nodes in each network</span>
<span class="c1"># This provides a quick overview of the size of each network</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of nodes in gene correlation network: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">G_gxp_nodes_list</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of nodes in patient gene expression network: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">G_patient_gxp_nodes_list</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of nodes in patient DNA methylation network: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">G_patient_dnam_nodes_list</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of nodes in gene correlation network: 300
Number of nodes in patient gene expression network: 434
Number of nodes in patient DNA methylation network: 381
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
<span class="kn">from</span> <span class="nn">matplotlib.patches</span> <span class="kn">import</span> <span class="n">Patch</span>
<span class="kn">from</span> <span class="nn">matplotlib.lines</span> <span class="kn">import</span> <span class="n">Line2D</span>

<span class="k">def</span> <span class="nf">draw_network_with_node_attrs</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">node_attributes</span><span class="p">,</span> <span class="n">communities</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Network Visualization&#39;</span><span class="p">,</span> <span class="n">color_attr</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">shape_attr</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span> <span class="n">layout</span><span class="o">=</span><span class="s1">&#39;spring&#39;</span><span class="p">,</span> <span class="n">cmap_name</span><span class="o">=</span><span class="s1">&#39;tab20&#39;</span><span class="p">,</span> <span class="n">with_labels</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Draws a graph with nodes colored by their community memberships and includes a legend indicating the community index.</span>
<span class="sd">    Nodes can also be colored and shaped based on additional attributes.</span>

<span class="sd">    Args:</span>
<span class="sd">        G (networkx.Graph): The graph to be drawn.</span>
<span class="sd">        communities (list of lists): A list where each sublist contains the nodes belonging to a community.</span>
<span class="sd">        node_attributes (dict): A dictionary where keys are node names and values are dictionaries of attributes.</span>
<span class="sd">        color_attr (str, optional): Node attribute to color by.</span>
<span class="sd">        shape_attr (str, optional): Node attribute to shape by.</span>
<span class="sd">        figsize (tuple, optional): The size of the figure. Default is (10, 10).</span>
<span class="sd">        layout (str, optional): The layout algorithm for positioning nodes (e.g., &#39;spring&#39;, &#39;circular&#39;). Default is &#39;spring&#39;.</span>
<span class="sd">        cmap_name (str, optional): The name of the colormap to use for coloring communities. Default is &#39;tab20&#39;.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the graph G is empty or not defined.</span>
<span class="sd">        ValueError: If the communities list is empty or not defined.</span>
<span class="sd">        ValueError: If node_attributes is empty or not defined.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">G</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The graph is empty or not defined.&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">node_attributes</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Node attributes are empty or not defined.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">communities</span><span class="p">:</span>
        <span class="n">community_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">node</span><span class="p">:</span> <span class="n">idx</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">community</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">communities</span><span class="p">)</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">community</span><span class="p">}</span>
        <span class="n">node_attributes</span><span class="p">[</span><span class="s1">&#39;community&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">community_dict</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;node_attributes: </span><span class="si">{</span><span class="n">node_attributes</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="n">shapes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="s1">&#39;^&#39;</span><span class="p">,</span> <span class="s1">&#39;s&#39;</span><span class="p">,</span> <span class="s1">&#39;p&#39;</span><span class="p">,</span> <span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">,</span> <span class="s1">&#39;8&#39;</span><span class="p">,</span> <span class="s1">&#39;d&#39;</span><span class="p">,</span> <span class="s1">&#39;D&#39;</span><span class="p">,</span> <span class="s1">&#39;v&#39;</span><span class="p">,</span> <span class="s1">&#39;&lt;&#39;</span><span class="p">,</span> <span class="s1">&#39;&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;*&#39;</span><span class="p">,</span> <span class="s1">&#39;X&#39;</span><span class="p">]</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="n">cmap_name</span><span class="p">)</span>
    
    <span class="n">unique_attr_vals_color</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">node_attributes</span><span class="p">[</span><span class="n">color_attr</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span> <span class="k">if</span> <span class="n">color_attr</span> <span class="k">else</span> <span class="p">[]</span>
    <span class="n">color_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">val</span><span class="p">:</span> <span class="n">cmap</span><span class="p">(</span><span class="n">i</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_attr_vals_color</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_attr_vals_color</span><span class="p">)}</span>
    
    <span class="n">node_colors_from_attribute</span> <span class="o">=</span> <span class="p">[</span><span class="n">color_map</span><span class="p">[</span><span class="n">node_attributes</span><span class="p">[</span><span class="n">color_attr</span><span class="p">][</span><span class="n">node</span><span class="p">]]</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">()]</span> <span class="k">if</span> <span class="n">color_attr</span> <span class="k">else</span> <span class="p">[</span><span class="s1">&#39;blue&#39;</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">()))</span>
    
    <span class="n">unique_attr_vals_shape</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">node_attributes</span><span class="p">[</span><span class="n">shape_attr</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span> <span class="k">if</span> <span class="n">shape_attr</span> <span class="k">else</span> <span class="p">[]</span>
    <span class="n">shape_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">val</span><span class="p">:</span> <span class="n">shapes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">unique_attr_vals_shape</span><span class="p">)}</span>
    <span class="n">node_shapes_from_attribute</span> <span class="o">=</span> <span class="p">[</span><span class="n">shape_map</span><span class="p">[</span><span class="n">node_attributes</span><span class="p">[</span><span class="n">shape_attr</span><span class="p">][</span><span class="n">node</span><span class="p">]]</span> <span class="k">if</span> <span class="n">shape_attr</span> <span class="k">else</span> <span class="s1">&#39;o&#39;</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">()]</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">nx</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">layout</span><span class="si">}</span><span class="s1">_layout&#39;</span><span class="p">)(</span><span class="n">G</span><span class="p">)</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">nx</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">layout</span><span class="si">}</span><span class="s1">_layout&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="n">nx</span><span class="o">.</span><span class="n">spring_layout</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
    
    <span class="n">node_list_by_shape</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">node_idx_list_by_shape</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">shape_marker</span> <span class="ow">in</span> <span class="n">shapes</span><span class="p">[:</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_attr_vals_shape</span><span class="p">)]:</span>
        <span class="n">node_list_by_shape</span><span class="p">[</span><span class="n">shape_marker</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">node</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">())</span> <span class="k">if</span> <span class="n">shape_map</span><span class="p">[</span><span class="n">node_attributes</span><span class="p">[</span><span class="n">shape_attr</span><span class="p">][</span><span class="n">node</span><span class="p">]]</span> <span class="o">==</span> <span class="n">shape_marker</span><span class="p">]</span>
        <span class="n">node_idx_list_by_shape</span><span class="p">[</span><span class="n">shape_marker</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">node_idx</span> <span class="k">for</span> <span class="n">node_idx</span><span class="p">,</span> <span class="n">node</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">()))</span> <span class="k">if</span> <span class="n">shape_map</span><span class="p">[</span><span class="n">node_attributes</span><span class="p">[</span><span class="n">shape_attr</span><span class="p">][</span><span class="n">node</span><span class="p">]]</span> <span class="o">==</span> <span class="n">shape_marker</span><span class="p">]</span>
    
    <span class="k">for</span> <span class="n">shape_marker</span><span class="p">,</span> <span class="n">node_list</span> <span class="ow">in</span> <span class="n">node_list_by_shape</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_nodes</span><span class="p">(</span>
            <span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span>
            <span class="n">nodelist</span><span class="o">=</span><span class="n">node_list_by_shape</span><span class="p">[</span><span class="n">shape_marker</span><span class="p">],</span>
            <span class="n">node_color</span><span class="o">=</span><span class="p">[</span><span class="n">node_colors_from_attribute</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">node_idx_list_by_shape</span><span class="p">[</span><span class="n">shape_marker</span><span class="p">]],</span>
            <span class="n">node_shape</span><span class="o">=</span><span class="n">shape_marker</span><span class="p">,</span>
            <span class="n">node_size</span><span class="o">=</span><span class="mi">400</span><span class="p">,</span>
            <span class="n">edgecolors</span><span class="o">=</span><span class="s1">&#39;yellow&#39;</span>
        <span class="p">)</span>

    <span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_edges</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">with_labels</span><span class="p">:</span>
        <span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_labels</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">font_size</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>

    <span class="n">legend_fontsize</span> <span class="o">=</span> <span class="mi">14</span>
    <span class="c1"># leg1=None</span>
    <span class="k">if</span> <span class="n">shape_attr</span><span class="p">:</span>
        <span class="n">shape_legend_handles</span> <span class="o">=</span> <span class="p">[</span><span class="n">Line2D</span><span class="p">([</span><span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="n">shape_map</span><span class="p">[</span><span class="n">val</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">val</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">markerfacecolor</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span> <span class="k">for</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">unique_attr_vals_shape</span><span class="p">]</span>
        <span class="n">leg1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">shape_legend_handles</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">shape_attr</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s2"> (Shape)&quot;</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">legend_fontsize</span><span class="p">,</span> <span class="n">title_fontsize</span><span class="o">=</span><span class="n">legend_fontsize</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">color_attr</span><span class="p">:</span>
        <span class="n">color_legend_handles</span> <span class="o">=</span> <span class="p">[</span><span class="n">Patch</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="n">color_map</span><span class="p">[</span><span class="n">val</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">color_attr</span><span class="si">}</span><span class="s1"> </span><span class="si">{</span><span class="n">val</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">unique_attr_vals_color</span><span class="p">]</span>
        <span class="n">leg2</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">color_legend_handles</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">color_attr</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s2"> (Color)&quot;</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">legend_fontsize</span><span class="p">,</span> <span class="n">title_fontsize</span><span class="o">=</span><span class="n">legend_fontsize</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">shape_attr</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">add_artist</span><span class="p">(</span><span class="n">leg1</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">title</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
    
    
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Display the shape of the DataFrame</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Shape of the DataFrame (rows, columns):&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

<span class="c1"># List the columns in the DataFrame</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">List of columns in the DataFrame:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Shape of the DataFrame (rows, columns):
(20244, 498)

List of columns in the DataFrame:
Index([&#39;TCGA-38-7271&#39;, &#39;TCGA-55-7914&#39;, &#39;TCGA-95-7043&#39;, &#39;TCGA-73-4658&#39;,
       &#39;TCGA-86-8076&#39;, &#39;TCGA-55-7726&#39;, &#39;TCGA-44-6147&#39;, &#39;TCGA-50-5932&#39;,
       &#39;TCGA-44-2661&#39;, &#39;TCGA-86-7954&#39;,
       ...
       &#39;TCGA-97-A4M7&#39;, &#39;TCGA-62-A46R&#39;, &#39;TCGA-50-5055&#39;, &#39;TCGA-38-4628&#39;,
       &#39;TCGA-86-7713&#39;, &#39;TCGA-86-8073&#39;, &#39;TCGA-MN-A4N4&#39;, &#39;TCGA-53-7626&#39;,
       &#39;TCGA-44-A47G&#39;, &#39;TCGA-55-6969&#39;],
      dtype=&#39;object&#39;, length=498)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cancer_genes_path</span> <span class="o">=</span> <span class="s1">&#39;section2_data/Census_allFri Jul  5 14_32_40 2024.csv&#39;</span>
<span class="n">cancer_genes_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cancer_genes_path</span><span class="p">)</span>

<span class="n">cancer_genes</span> <span class="o">=</span> <span class="n">cancer_genes_df</span><span class="p">[</span><span class="s1">&#39;Gene Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="c1"># Create the dictionary with node labels as keys and boolean as values</span>
<span class="n">found_in_cancer_genes</span> <span class="o">=</span> <span class="p">{</span><span class="n">node</span><span class="p">:</span> <span class="nb">bool</span><span class="p">(</span><span class="n">node</span> <span class="ow">in</span> <span class="n">cancer_genes</span><span class="p">)</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">G_gxp</span><span class="o">.</span><span class="n">nodes</span><span class="p">())}</span>
<span class="c1"># true_nodes = [node for node, is_present in found_in_cancer_genes.items() if is_present]</span>

<span class="c1"># # Print the resulting dictionary</span>
<span class="c1"># print(found_in_cancer_genes)</span>
<span class="c1"># print(true_nodes)</span>
<span class="c1"># print(len(true_nodes))</span>
<span class="c1"># print(len(list(G_gxp.nodes())))</span>
<span class="n">node_attributes</span><span class="o">=</span><span class="p">{}</span>
<span class="n">node_attributes</span><span class="p">[</span><span class="s1">&#39;cancer_gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">found_in_cancer_genes</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">node_attributes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;cancer_gene&#39;: {&#39;A2M&#39;: False,
  &#39;ABI3BP&#39;: False,
  &#39;ACAP1&#39;: False,
  &#39;ACP5&#39;: False,
  &#39;ACTA2&#39;: False,
  &#39;ACTG2&#39;: False,
  &#39;ACVRL1&#39;: False,
  &#39;ADAM12&#39;: False,
  &#39;ADAMTS12&#39;: False,
  &#39;ADAMTS2&#39;: False,
  &#39;ADGRA2&#39;: False,
  &#39;ADH1B&#39;: False,
  &#39;AEBP1&#39;: False,
  &#39;AIF1&#39;: False,
  &#39;ALOX5&#39;: False,
  &#39;ALOX5AP&#39;: False,
  &#39;ANGPTL2&#39;: False,
  &#39;ANTXR1&#39;: False,
  &#39;AOC3&#39;: False,
  &#39;APBB1IP&#39;: False,
  &#39;APOC1&#39;: False,
  &#39;APOL3&#39;: False,
  &#39;ARHGAP30&#39;: False,
  &#39;ARHGAP9&#39;: False,
  &#39;ARHGEF6&#39;: False,
  &#39;BCL6B&#39;: False,
  &#39;BGN&#39;: False,
  &#39;BIN2&#39;: False,
  &#39;BTK&#39;: True,
  &#39;C1QA&#39;: False,
  &#39;C1QB&#39;: False,
  &#39;C1QC&#39;: False,
  &#39;C1orf162&#39;: False,
  &#39;C3AR1&#39;: False,
  &#39;C5AR1&#39;: False,
  &#39;C7&#39;: False,
  &#39;CALCRL&#39;: False,
  &#39;CASP1&#39;: False,
  &#39;CCDC80&#39;: False,
  &#39;CCL5&#39;: False,
  &#39;CCN4&#39;: False,
  &#39;CCR1&#39;: False,
  &#39;CCR5&#39;: False,
  &#39;CD14&#39;: False,
  &#39;CD163&#39;: False,
  &#39;CD2&#39;: False,
  &#39;CD300A&#39;: False,
  &#39;CD34&#39;: False,
  &#39;CD37&#39;: False,
  &#39;CD3E&#39;: False,
  &#39;CD4&#39;: False,
  &#39;CD52&#39;: False,
  &#39;CD53&#39;: False,
  &#39;CD6&#39;: False,
  &#39;CD74&#39;: True,
  &#39;CD84&#39;: False,
  &#39;CD86&#39;: False,
  &#39;CD8A&#39;: False,
  &#39;CD93&#39;: False,
  &#39;CDH5&#39;: False,
  &#39;CHRDL1&#39;: False,
  &#39;CLEC14A&#39;: False,
  &#39;CMKLR1&#39;: False,
  &#39;COL10A1&#39;: False,
  &#39;COL11A1&#39;: False,
  &#39;COL15A1&#39;: False,
  &#39;COL1A1&#39;: True,
  &#39;COL1A2&#39;: False,
  &#39;COL3A1&#39;: True,
  &#39;COL4A1&#39;: False,
  &#39;COL4A2&#39;: False,
  &#39;COL5A1&#39;: False,
  &#39;COL5A2&#39;: False,
  &#39;COL6A3&#39;: False,
  &#39;CORO1A&#39;: False,
  &#39;CSF1R&#39;: True,
  &#39;CSF2RB&#39;: False,
  &#39;CTHRC1&#39;: False,
  &#39;CXCL10&#39;: False,
  &#39;CXCL9&#39;: False,
  &#39;CYBB&#39;: False,
  &#39;CYTH4&#39;: False,
  &#39;CYTIP&#39;: False,
  &#39;CYYR1&#39;: False,
  &#39;DAB2&#39;: False,
  &#39;DCHS1&#39;: False,
  &#39;DCN&#39;: False,
  &#39;DIPK2B&#39;: False,
  &#39;DOCK2&#39;: False,
  &#39;DOCK8&#39;: False,
  &#39;DOK2&#39;: False,
  &#39;DOK3&#39;: False,
  &#39;EDNRB&#39;: False,
  &#39;EMILIN1&#39;: False,
  &#39;EMILIN2&#39;: False,
  &#39;EPB41L2&#39;: False,
  &#39;ERG&#39;: True,
  &#39;EVI2B&#39;: False,
  &#39;F13A1&#39;: False,
  &#39;FAM78A&#39;: False,
  &#39;FAP&#39;: False,
  &#39;FBN1&#39;: False,
  &#39;FCER1G&#39;: False,
  &#39;FCGR2A&#39;: False,
  &#39;FCGR3A&#39;: False,
  &#39;FERMT3&#39;: False,
  &#39;FGD2&#39;: False,
  &#39;FGD5&#39;: False,
  &#39;FGL2&#39;: False,
  &#39;FGR&#39;: False,
  &#39;FHL1&#39;: False,
  &#39;FLI1&#39;: True,
  &#39;FLT4&#39;: True,
  &#39;FMO2&#39;: False,
  &#39;FNDC1&#39;: False,
  &#39;FOLR2&#39;: False,
  &#39;FPR1&#39;: False,
  &#39;FPR3&#39;: False,
  &#39;FSTL1&#39;: False,
  &#39;FYB1&#39;: False,
  &#39;GAS7&#39;: True,
  &#39;GBP1&#39;: False,
  &#39;GBP4&#39;: False,
  &#39;GBP5&#39;: False,
  &#39;GIMAP4&#39;: False,
  &#39;GIMAP6&#39;: False,
  &#39;GIMAP7&#39;: False,
  &#39;GIMAP8&#39;: False,
  &#39;GLIPR2&#39;: False,
  &#39;GMFG&#39;: False,
  &#39;GREM1&#39;: False,
  &#39;GYPC&#39;: False,
  &#39;HAVCR2&#39;: False,
  &#39;HCK&#39;: False,
  &#39;HCP5&#39;: False,
  &#39;HEG1&#39;: False,
  &#39;HK3&#39;: False,
  &#39;HLA-A&#39;: True,
  &#39;HLA-B&#39;: False,
  &#39;HLA-C&#39;: False,
  &#39;HLA-DMA&#39;: False,
  &#39;HLA-DMB&#39;: False,
  &#39;HLA-DOA&#39;: False,
  &#39;HLA-DPA1&#39;: False,
  &#39;HLA-DPB1&#39;: False,
  &#39;HLA-DQA1&#39;: False,
  &#39;HLA-DQB1&#39;: False,
  &#39;HLA-DRA&#39;: False,
  &#39;HLA-DRB1&#39;: False,
  &#39;HLA-DRB5&#39;: False,
  &#39;HLA-F&#39;: False,
  &#39;HTRA3&#39;: False,
  &#39;IGSF6&#39;: False,
  &#39;IKZF1&#39;: True,
  &#39;IL10RA&#39;: False,
  &#39;IL16&#39;: False,
  &#39;IL2RB&#39;: False,
  &#39;IL2RG&#39;: False,
  &#39;IL33&#39;: False,
  &#39;IL7R&#39;: True,
  &#39;INMT&#39;: False,
  &#39;IRAG1&#39;: False,
  &#39;IRF8&#39;: False,
  &#39;ISLR&#39;: False,
  &#39;ITGA11&#39;: False,
  &#39;ITGA4&#39;: False,
  &#39;ITGA8&#39;: False,
  &#39;ITGAL&#39;: False,
  &#39;ITGAM&#39;: False,
  &#39;ITGAX&#39;: False,
  &#39;ITGB2&#39;: False,
  &#39;JAML&#39;: False,
  &#39;KCNAB2&#39;: False,
  &#39;KLHL6&#39;: False,
  &#39;LAIR1&#39;: False,
  &#39;LAMA2&#39;: False,
  &#39;LAMA4&#39;: False,
  &#39;LAPTM5&#39;: False,
  &#39;LCK&#39;: True,
  &#39;LCP1&#39;: True,
  &#39;LCP2&#39;: False,
  &#39;LDB2&#39;: False,
  &#39;LHFPL6&#39;: True,
  &#39;LILRB2&#39;: False,
  &#39;LILRB4&#39;: False,
  &#39;LIPA&#39;: False,
  &#39;LMCD1&#39;: False,
  &#39;LMOD1&#39;: False,
  &#39;LRRC15&#39;: False,
  &#39;LRRC32&#39;: False,
  &#39;LSP1&#39;: False,
  &#39;LST1&#39;: False,
  &#39;LTBP2&#39;: False,
  &#39;LUM&#39;: False,
  &#39;MARCO&#39;: False,
  &#39;MFAP4&#39;: False,
  &#39;MMP11&#39;: False,
  &#39;MMP2&#39;: False,
  &#39;MMRN2&#39;: False,
  &#39;MNDA&#39;: False,
  &#39;MPEG1&#39;: False,
  &#39;MRC1&#39;: False,
  &#39;MS4A4A&#39;: False,
  &#39;MS4A6A&#39;: False,
  &#39;MS4A7&#39;: False,
  &#39;MSR1&#39;: False,
  &#39;MSRB3&#39;: False,
  &#39;MXRA5&#39;: False,
  &#39;MYH11&#39;: True,
  &#39;MYLK&#39;: False,
  &#39;MYO1F&#39;: False,
  &#39;NCF2&#39;: False,
  &#39;NCF4&#39;: False,
  &#39;NCKAP1L&#39;: False,
  &#39;NFAM1&#39;: False,
  &#39;NOTCH4&#39;: False,
  &#39;OLFML2B&#39;: False,
  &#39;PARVG&#39;: False,
  &#39;PCDH12&#39;: False,
  &#39;PCDH17&#39;: False,
  &#39;PDGFRA&#39;: True,
  &#39;PDGFRB&#39;: True,
  &#39;PECAM1&#39;: False,
  &#39;PIK3AP1&#39;: False,
  &#39;PIK3CG&#39;: False,
  &#39;PIK3R5&#39;: False,
  &#39;PLEK&#39;: False,
  &#39;PLXDC2&#39;: False,
  &#39;PLXNC1&#39;: False,
  &#39;PODN&#39;: False,
  &#39;POSTN&#39;: False,
  &#39;PPP1R16B&#39;: False,
  &#39;PRELP&#39;: False,
  &#39;PREX1&#39;: False,
  &#39;PRKCB&#39;: True,
  &#39;PRRX1&#39;: True,
  &#39;PSMB9&#39;: False,
  &#39;PTAFR&#39;: False,
  &#39;PTGER4&#39;: False,
  &#39;PTPN7&#39;: False,
  &#39;PTPRB&#39;: True,
  &#39;PTPRC&#39;: True,
  &#39;RAMP2&#39;: False,
  &#39;RAMP3&#39;: False,
  &#39;RASAL3&#39;: False,
  &#39;RASSF2&#39;: False,
  &#39;RCSD1&#39;: False,
  &#39;RNASE6&#39;: False,
  &#39;ROBO4&#39;: False,
  &#39;S1PR1&#39;: False,
  &#39;SAMHD1&#39;: False,
  &#39;SAMSN1&#39;: False,
  &#39;SASH3&#39;: False,
  &#39;SCN7A&#39;: False,
  &#39;SELPLG&#39;: False,
  &#39;SH2D3C&#39;: False,
  &#39;SHANK3&#39;: False,
  &#39;SIGLEC1&#39;: False,
  &#39;SIRPA&#39;: True,
  &#39;SLA&#39;: False,
  &#39;SLC15A3&#39;: False,
  &#39;SLC1A3&#39;: False,
  &#39;SLC7A7&#39;: False,
  &#39;SLCO2B1&#39;: False,
  &#39;SLIT2&#39;: False,
  &#39;SLIT3&#39;: False,
  &#39;SPARC&#39;: False,
  &#39;SPARCL1&#39;: False,
  &#39;SPI1&#39;: False,
  &#39;SPN&#39;: False,
  &#39;SPON1&#39;: False,
  &#39;SSC5D&#39;: False,
  &#39;STARD8&#39;: False,
  &#39;STAT1&#39;: False,
  &#39;SULF1&#39;: False,
  &#39;SVEP1&#39;: False,
  &#39;SYNPO2&#39;: False,
  &#39;TAGLN&#39;: False,
  &#39;TAP1&#39;: False,
  &#39;TBC1D10C&#39;: False,
  &#39;TCF4&#39;: False,
  &#39;THBS2&#39;: False,
  &#39;THY1&#39;: False,
  &#39;TIE1&#39;: False,
  &#39;TLR4&#39;: False,
  &#39;TNFSF13B&#39;: False,
  &#39;TRAC&#39;: False,
  &#39;TRBC1&#39;: False,
  &#39;TRBC2&#39;: False,
  &#39;TREM2&#39;: False,
  &#39;TRPV2&#39;: False,
  &#39;TYROBP&#39;: False,
  &#39;UBE2L6&#39;: False,
  &#39;UNC5B&#39;: False,
  &#39;VCAN&#39;: False,
  &#39;VSIG4&#39;: False,
  &#39;VWF&#39;: False,
  &#39;WAS&#39;: True,
  &#39;ZEB1&#39;: True,
  &#39;ZEB2&#39;: False}}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check for overlapping nodes between the patient networks</span>
<span class="c1"># In patient networks, the nodes represent patients. We want to ensure that there are sufficient number of nodes commonly found in the networks of different modalities before performing Similarity Network Fusion (Part 2), so that they represent different aspects of the same patients.</span>
<span class="c1"># Convert the node lists to sets and find the intersection to identify overlapping nodes</span>

<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">overlap</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">G_patient_gxp_nodes_list</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">G_patient_dnam_nodes_list</span><span class="p">))</span>

<span class="c1"># Get the number of overlapping nodes by taking the length of the intersection set</span>
<span class="n">num_overlap</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">overlap</span><span class="p">)</span>

<span class="c1"># Print the number of overlapping nodes</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">num_overlap</span><span class="si">}</span><span class="s2"> overlapping nodes between patient networks&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>381 overlapping nodes between patient networks
</pre></div>
</div>
</div>
</div>
</section>
<section id="gene-network-analysis-with-g-exp">
<h2>1. Gene Network Analysis with <code class="docutils literal notranslate"><span class="pre">G_exp</span></code><a class="headerlink" href="#gene-network-analysis-with-g-exp" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># We are using &#39;KEGG_2021_Human&#39;, which refers to the KEGG (Kyoto Encyclopedia of Genes and Genomes) pathway database.</span>
<span class="c1"># KEGG is a collection of manually curated databases dealing with genomes, biological pathways, diseases, drugs, and chemical substances.</span>
<span class="c1"># Alternative gene sets like &#39;MSigDB_Hallmark_2020&#39; can also be used.</span>
<span class="n">gene_sets</span> <span class="o">=</span> <span class="s1">&#39;KEGG_2021_Human&#39;</span>  <span class="c1"># &#39;MSigDB_Hallmark_2020&#39;</span>
</pre></div>
</div>
</div>
</div>
<section id="over-representation-analysis-ora">
<h3>1.1 Over Representation Analysis (ORA)<a class="headerlink" href="#over-representation-analysis-ora" title="Link to this heading">#</a></h3>
<p>Over-representation analysis (ORA) is a method used to identify which predefined gene sets are disproportionately represented in a given set of genes compared to what would be expected by random chance (Huang et al., 2009). (https://pnnl-comp-mass-spec.github.io/proteomics-data-analysis-tutorial/ora.html) We recommend using Over-representation analysis (ORA) only when Gene Set Enrichment Analysis (GSEA) is not suitable. Although we are using the <code class="docutils literal notranslate"><span class="pre">gseapy</span></code> library for ORA in this the tutorial, itâ€™s important to note that ORA and GSEA are distinct methods.</p>
<section id="ora-on-gene-correlation-network">
<h4>1.1.1 ORA on gene correlation network<a class="headerlink" href="#ora-on-gene-correlation-network" title="Link to this heading">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Perform ORA on all nodes in the gene correlation network</span>
<span class="n">enr_all_nodes</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">enrichr</span><span class="p">(</span><span class="n">gene_list</span><span class="o">=</span><span class="n">G_gxp_nodes_list</span><span class="p">,</span> <span class="n">gene_sets</span><span class="o">=</span><span class="p">[</span><span class="n">gene_sets</span><span class="p">],</span> <span class="n">organism</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

<span class="c1"># Display the top 10 enrichment results</span>

<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">enr_all_nodes</span><span class="o">.</span><span class="n">results</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene_set</th>
      <th>Term</th>
      <th>Overlap</th>
      <th>P-value</th>
      <th>Adjusted P-value</th>
      <th>Old P-value</th>
      <th>Old Adjusted P-value</th>
      <th>Odds Ratio</th>
      <th>Combined Score</th>
      <th>Genes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>KEGG_2021_Human</td>
      <td>Cell adhesion molecules</td>
      <td>32/148</td>
      <td>4.963642e-28</td>
      <td>9.778376e-26</td>
      <td>0</td>
      <td>0</td>
      <td>20.158518</td>
      <td>1267.370905</td>
      <td>CD86;ITGAM;SELPLG;ITGB2;ITGAL;SPN;CDH5;HLA-DMA;HLA-DMB;HLA-DOA;CD34;HLA-DQA1;HLA-DPA1;HLA-DRB5;ITGA4;HLA-B;HLA-C;HLA-A;HLA-F;CD2;VCAN;CD4;PTPRC;CD6;CD8A;ITGA8;PECAM1;HLA-DPB1;HLA-DRA;SIGLEC1;HLA-DRB1;HLA-DQB1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>KEGG_2021_Human</td>
      <td>Phagosome</td>
      <td>29/152</td>
      <td>7.718499e-24</td>
      <td>7.602722e-22</td>
      <td>0</td>
      <td>0</td>
      <td>17.032160</td>
      <td>906.424701</td>
      <td>ITGAM;NCF2;NCF4;ITGB2;THBS2;CORO1A;FCGR3A;HLA-DMA;HLA-DMB;MRC1;CD14;HLA-DOA;HLA-DQA1;HLA-DPA1;MSR1;HLA-DRB5;HLA-B;CYBB;HLA-C;TAP1;HLA-A;HLA-F;MARCO;FCGR2A;HLA-DPB1;HLA-DRA;TLR4;HLA-DRB1;HLA-DQB1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>KEGG_2021_Human</td>
      <td>Staphylococcus aureus infection</td>
      <td>24/95</td>
      <td>5.021667e-23</td>
      <td>3.297562e-21</td>
      <td>0</td>
      <td>0</td>
      <td>24.040416</td>
      <td>1234.371891</td>
      <td>C1QB;C1QA;HLA-DRB5;ITGAM;SELPLG;ITGB2;C5AR1;PTAFR;FPR1;FPR3;ITGAL;FCGR3A;HLA-DMA;HLA-DMB;FCGR2A;C3AR1;HLA-DPB1;HLA-DRA;HLA-DOA;HLA-DQA1;HLA-DRB1;HLA-DPA1;HLA-DQB1;C1QC</td>
    </tr>
    <tr>
      <th>3</th>
      <td>KEGG_2021_Human</td>
      <td>Leishmaniasis</td>
      <td>21/77</td>
      <td>4.634501e-21</td>
      <td>2.282492e-19</td>
      <td>0</td>
      <td>0</td>
      <td>26.403226</td>
      <td>1236.219058</td>
      <td>HLA-DRB5;ITGAM;ITGA4;NCF2;PRKCB;STAT1;ITGB2;NCF4;CYBB;FCGR3A;HLA-DMA;HLA-DMB;FCGR2A;HLA-DPB1;HLA-DRA;HLA-DOA;TLR4;HLA-DQA1;HLA-DRB1;HLA-DPA1;HLA-DQB1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>KEGG_2021_Human</td>
      <td>Viral myocarditis</td>
      <td>18/60</td>
      <td>4.647925e-19</td>
      <td>1.831283e-17</td>
      <td>0</td>
      <td>0</td>
      <td>29.875380</td>
      <td>1261.120325</td>
      <td>CD86;HLA-DRB5;LAMA2;ITGB2;HLA-B;HLA-C;HLA-A;ITGAL;HLA-F;HLA-DMA;HLA-DMB;HLA-DPB1;HLA-DRA;HLA-DOA;HLA-DQA1;HLA-DRB1;HLA-DPA1;HLA-DQB1</td>
    </tr>
    <tr>
      <th>5</th>
      <td>KEGG_2021_Human</td>
      <td>Hematopoietic cell lineage</td>
      <td>21/99</td>
      <td>1.392113e-18</td>
      <td>4.570771e-17</td>
      <td>0</td>
      <td>0</td>
      <td>18.934932</td>
      <td>778.523171</td>
      <td>CSF1R;HLA-DRB5;ITGAM;ITGA4;CD3E;CD2;HLA-DMA;CD4;HLA-DMB;CD8A;HLA-DPB1;HLA-DRA;CD37;CD14;HLA-DOA;IL7R;CD34;HLA-DQA1;HLA-DRB1;HLA-DPA1;HLA-DQB1</td>
    </tr>
    <tr>
      <th>6</th>
      <td>KEGG_2021_Human</td>
      <td>Allograft rejection</td>
      <td>15/38</td>
      <td>3.498581e-18</td>
      <td>9.846007e-17</td>
      <td>0</td>
      <td>0</td>
      <td>45.027460</td>
      <td>1809.841566</td>
      <td>CD86;HLA-DRB5;HLA-B;HLA-C;HLA-A;HLA-F;HLA-DMA;HLA-DMB;HLA-DPB1;HLA-DRA;HLA-DOA;HLA-DQA1;HLA-DRB1;HLA-DPA1;HLA-DQB1</td>
    </tr>
    <tr>
      <th>7</th>
      <td>KEGG_2021_Human</td>
      <td>Graft-versus-host disease</td>
      <td>15/42</td>
      <td>2.114525e-17</td>
      <td>5.207019e-16</td>
      <td>0</td>
      <td>0</td>
      <td>38.348928</td>
      <td>1472.411543</td>
      <td>CD86;HLA-DRB5;HLA-B;HLA-C;HLA-A;HLA-F;HLA-DMA;HLA-DMB;HLA-DPB1;HLA-DRA;HLA-DOA;HLA-DQA1;HLA-DRB1;HLA-DPA1;HLA-DQB1</td>
    </tr>
    <tr>
      <th>8</th>
      <td>KEGG_2021_Human</td>
      <td>Type I diabetes mellitus</td>
      <td>15/43</td>
      <td>3.203949e-17</td>
      <td>7.013088e-16</td>
      <td>0</td>
      <td>0</td>
      <td>36.977444</td>
      <td>1404.387131</td>
      <td>CD86;HLA-DRB5;HLA-B;HLA-C;HLA-A;HLA-F;HLA-DMA;HLA-DMB;HLA-DPB1;HLA-DRA;HLA-DOA;HLA-DQA1;HLA-DRB1;HLA-DPA1;HLA-DQB1</td>
    </tr>
    <tr>
      <th>9</th>
      <td>KEGG_2021_Human</td>
      <td>Antigen processing and presentation</td>
      <td>18/78</td>
      <td>8.383314e-17</td>
      <td>1.651513e-15</td>
      <td>0</td>
      <td>0</td>
      <td>20.893617</td>
      <td>773.433715</td>
      <td>CD74;HLA-DRB5;HLA-B;HLA-C;TAP1;HLA-A;HLA-F;HLA-DMA;CD4;HLA-DMB;CD8A;HLA-DPB1;HLA-DRA;HLA-DOA;HLA-DQA1;HLA-DRB1;HLA-DPA1;HLA-DQB1</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="ora-on-gene-clusters">
<h4>1.1.1 ORA on gene clusters<a class="headerlink" href="#ora-on-gene-clusters" title="Link to this heading">#</a></h4>
<p>Clustering on Gene Correlation Network</p>
<ul class="simple">
<li><p>Use community detection algorithm to identify communities in the network: The <code class="docutils literal notranslate"><span class="pre">greedy_modularity_communities</span></code> function in NetworkX implements a community detection algorithm that optimises modularity using a greedy approach. It iteratively merges pairs of nodes or communities that result in the largest increase in modularity until no further improvement is possible. Modularity measures the density of links inside communities compared to links between communities, aiming to maximize this value to identify densely connected groups within the network.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Cluster the gene correlation network using the greedy modularity communities algorithm</span>
<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">communities</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">algorithms</span><span class="o">.</span><span class="n">community</span><span class="o">.</span><span class="n">modularity_max</span><span class="o">.</span><span class="n">greedy_modularity_communities</span><span class="p">(</span><span class="n">G_gxp</span><span class="p">)</span>

<span class="c1"># Print the number of detected communities</span>
<span class="c1">### YOUR CODE HERE ###</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The network has </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">communities</span><span class="p">)</span><span class="si">}</span><span class="s1"> communities.</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="c1"># Create subgraphs for each community</span>
<span class="n">subgraphs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">community</span> <span class="ow">in</span> <span class="n">communities</span><span class="p">:</span>
    <span class="c1">### YOUR CODE HERE ###</span>
    <span class="n">subgraphs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">G_gxp</span><span class="o">.</span><span class="n">subgraph</span><span class="p">(</span><span class="n">community</span><span class="p">))</span>

<span class="c1"># Print the number of nodes in each community</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">subgraph</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">subgraphs</span><span class="p">):</span>
    <span class="c1">### YOUR CODE HERE ###</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Community </span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1"> has </span><span class="si">{</span><span class="n">subgraph</span><span class="o">.</span><span class="n">number_of_nodes</span><span class="p">()</span><span class="si">}</span><span class="s1"> nodes.&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The network has 6 communities.

Community 1 has 101 nodes.
Community 2 has 67 nodes.
Community 3 has 64 nodes.
Community 4 has 51 nodes.
Community 5 has 15 nodes.
Community 6 has 2 nodes.
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="visualising-the-communities">
<h3>Visualising the Communities<a class="headerlink" href="#visualising-the-communities" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the gene correlation network with community memberships</span>
<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">draw_network_with_node_attrs</span><span class="p">(</span><span class="n">G_gxp</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Gene Network&#39;</span><span class="p">,</span> <span class="n">node_attributes</span><span class="o">=</span><span class="n">node_attributes</span><span class="p">,</span> <span class="n">communities</span><span class="o">=</span><span class="n">communities</span><span class="p">,</span> <span class="n">color_attr</span><span class="o">=</span><span class="s1">&#39;community&#39;</span><span class="p">,</span> <span class="n">shape_attr</span><span class="o">=</span><span class="s1">&#39;cancer_gene&#39;</span><span class="p">,</span> <span class="n">with_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>node_attributes: {&#39;cancer_gene&#39;: {&#39;A2M&#39;: False, &#39;ABI3BP&#39;: False, &#39;ACAP1&#39;: False, &#39;ACP5&#39;: False, &#39;ACTA2&#39;: False, &#39;ACTG2&#39;: False, &#39;ACVRL1&#39;: False, &#39;ADAM12&#39;: False, &#39;ADAMTS12&#39;: False, &#39;ADAMTS2&#39;: False, &#39;ADGRA2&#39;: False, &#39;ADH1B&#39;: False, &#39;AEBP1&#39;: False, &#39;AIF1&#39;: False, &#39;ALOX5&#39;: False, &#39;ALOX5AP&#39;: False, &#39;ANGPTL2&#39;: False, &#39;ANTXR1&#39;: False, &#39;AOC3&#39;: False, &#39;APBB1IP&#39;: False, &#39;APOC1&#39;: False, &#39;APOL3&#39;: False, &#39;ARHGAP30&#39;: False, &#39;ARHGAP9&#39;: False, &#39;ARHGEF6&#39;: False, &#39;BCL6B&#39;: False, &#39;BGN&#39;: False, &#39;BIN2&#39;: False, &#39;BTK&#39;: True, &#39;C1QA&#39;: False, &#39;C1QB&#39;: False, &#39;C1QC&#39;: False, &#39;C1orf162&#39;: False, &#39;C3AR1&#39;: False, &#39;C5AR1&#39;: False, &#39;C7&#39;: False, &#39;CALCRL&#39;: False, &#39;CASP1&#39;: False, &#39;CCDC80&#39;: False, &#39;CCL5&#39;: False, &#39;CCN4&#39;: False, &#39;CCR1&#39;: False, &#39;CCR5&#39;: False, &#39;CD14&#39;: False, &#39;CD163&#39;: False, &#39;CD2&#39;: False, &#39;CD300A&#39;: False, &#39;CD34&#39;: False, &#39;CD37&#39;: False, &#39;CD3E&#39;: False, &#39;CD4&#39;: False, &#39;CD52&#39;: False, &#39;CD53&#39;: False, &#39;CD6&#39;: False, &#39;CD74&#39;: True, &#39;CD84&#39;: False, &#39;CD86&#39;: False, &#39;CD8A&#39;: False, &#39;CD93&#39;: False, &#39;CDH5&#39;: False, &#39;CHRDL1&#39;: False, &#39;CLEC14A&#39;: False, &#39;CMKLR1&#39;: False, &#39;COL10A1&#39;: False, &#39;COL11A1&#39;: False, &#39;COL15A1&#39;: False, &#39;COL1A1&#39;: True, &#39;COL1A2&#39;: False, &#39;COL3A1&#39;: True, &#39;COL4A1&#39;: False, &#39;COL4A2&#39;: False, &#39;COL5A1&#39;: False, &#39;COL5A2&#39;: False, &#39;COL6A3&#39;: False, &#39;CORO1A&#39;: False, &#39;CSF1R&#39;: True, &#39;CSF2RB&#39;: False, &#39;CTHRC1&#39;: False, &#39;CXCL10&#39;: False, &#39;CXCL9&#39;: False, &#39;CYBB&#39;: False, &#39;CYTH4&#39;: False, &#39;CYTIP&#39;: False, &#39;CYYR1&#39;: False, &#39;DAB2&#39;: False, &#39;DCHS1&#39;: False, &#39;DCN&#39;: False, &#39;DIPK2B&#39;: False, &#39;DOCK2&#39;: False, &#39;DOCK8&#39;: False, &#39;DOK2&#39;: False, &#39;DOK3&#39;: False, &#39;EDNRB&#39;: False, &#39;EMILIN1&#39;: False, &#39;EMILIN2&#39;: False, &#39;EPB41L2&#39;: False, &#39;ERG&#39;: True, &#39;EVI2B&#39;: False, &#39;F13A1&#39;: False, &#39;FAM78A&#39;: False, &#39;FAP&#39;: False, &#39;FBN1&#39;: False, &#39;FCER1G&#39;: False, &#39;FCGR2A&#39;: False, &#39;FCGR3A&#39;: False, &#39;FERMT3&#39;: False, &#39;FGD2&#39;: False, &#39;FGD5&#39;: False, &#39;FGL2&#39;: False, &#39;FGR&#39;: False, &#39;FHL1&#39;: False, &#39;FLI1&#39;: True, &#39;FLT4&#39;: True, &#39;FMO2&#39;: False, &#39;FNDC1&#39;: False, &#39;FOLR2&#39;: False, &#39;FPR1&#39;: False, &#39;FPR3&#39;: False, &#39;FSTL1&#39;: False, &#39;FYB1&#39;: False, &#39;GAS7&#39;: True, &#39;GBP1&#39;: False, &#39;GBP4&#39;: False, &#39;GBP5&#39;: False, &#39;GIMAP4&#39;: False, &#39;GIMAP6&#39;: False, &#39;GIMAP7&#39;: False, &#39;GIMAP8&#39;: False, &#39;GLIPR2&#39;: False, &#39;GMFG&#39;: False, &#39;GREM1&#39;: False, &#39;GYPC&#39;: False, &#39;HAVCR2&#39;: False, &#39;HCK&#39;: False, &#39;HCP5&#39;: False, &#39;HEG1&#39;: False, &#39;HK3&#39;: False, &#39;HLA-A&#39;: True, &#39;HLA-B&#39;: False, &#39;HLA-C&#39;: False, &#39;HLA-DMA&#39;: False, &#39;HLA-DMB&#39;: False, &#39;HLA-DOA&#39;: False, &#39;HLA-DPA1&#39;: False, &#39;HLA-DPB1&#39;: False, &#39;HLA-DQA1&#39;: False, &#39;HLA-DQB1&#39;: False, &#39;HLA-DRA&#39;: False, &#39;HLA-DRB1&#39;: False, &#39;HLA-DRB5&#39;: False, &#39;HLA-F&#39;: False, &#39;HTRA3&#39;: False, &#39;IGSF6&#39;: False, &#39;IKZF1&#39;: True, &#39;IL10RA&#39;: False, &#39;IL16&#39;: False, &#39;IL2RB&#39;: False, &#39;IL2RG&#39;: False, &#39;IL33&#39;: False, &#39;IL7R&#39;: True, &#39;INMT&#39;: False, &#39;IRAG1&#39;: False, &#39;IRF8&#39;: False, &#39;ISLR&#39;: False, &#39;ITGA11&#39;: False, &#39;ITGA4&#39;: False, &#39;ITGA8&#39;: False, &#39;ITGAL&#39;: False, &#39;ITGAM&#39;: False, &#39;ITGAX&#39;: False, &#39;ITGB2&#39;: False, &#39;JAML&#39;: False, &#39;KCNAB2&#39;: False, &#39;KLHL6&#39;: False, &#39;LAIR1&#39;: False, &#39;LAMA2&#39;: False, &#39;LAMA4&#39;: False, &#39;LAPTM5&#39;: False, &#39;LCK&#39;: True, &#39;LCP1&#39;: True, &#39;LCP2&#39;: False, &#39;LDB2&#39;: False, &#39;LHFPL6&#39;: True, &#39;LILRB2&#39;: False, &#39;LILRB4&#39;: False, &#39;LIPA&#39;: False, &#39;LMCD1&#39;: False, &#39;LMOD1&#39;: False, &#39;LRRC15&#39;: False, &#39;LRRC32&#39;: False, &#39;LSP1&#39;: False, &#39;LST1&#39;: False, &#39;LTBP2&#39;: False, &#39;LUM&#39;: False, &#39;MARCO&#39;: False, &#39;MFAP4&#39;: False, &#39;MMP11&#39;: False, &#39;MMP2&#39;: False, &#39;MMRN2&#39;: False, &#39;MNDA&#39;: False, &#39;MPEG1&#39;: False, &#39;MRC1&#39;: False, &#39;MS4A4A&#39;: False, &#39;MS4A6A&#39;: False, &#39;MS4A7&#39;: False, &#39;MSR1&#39;: False, &#39;MSRB3&#39;: False, &#39;MXRA5&#39;: False, &#39;MYH11&#39;: True, &#39;MYLK&#39;: False, &#39;MYO1F&#39;: False, &#39;NCF2&#39;: False, &#39;NCF4&#39;: False, &#39;NCKAP1L&#39;: False, &#39;NFAM1&#39;: False, &#39;NOTCH4&#39;: False, &#39;OLFML2B&#39;: False, &#39;PARVG&#39;: False, &#39;PCDH12&#39;: False, &#39;PCDH17&#39;: False, &#39;PDGFRA&#39;: True, &#39;PDGFRB&#39;: True, &#39;PECAM1&#39;: False, &#39;PIK3AP1&#39;: False, &#39;PIK3CG&#39;: False, &#39;PIK3R5&#39;: False, &#39;PLEK&#39;: False, &#39;PLXDC2&#39;: False, &#39;PLXNC1&#39;: False, &#39;PODN&#39;: False, &#39;POSTN&#39;: False, &#39;PPP1R16B&#39;: False, &#39;PRELP&#39;: False, &#39;PREX1&#39;: False, &#39;PRKCB&#39;: True, &#39;PRRX1&#39;: True, &#39;PSMB9&#39;: False, &#39;PTAFR&#39;: False, &#39;PTGER4&#39;: False, &#39;PTPN7&#39;: False, &#39;PTPRB&#39;: True, &#39;PTPRC&#39;: True, &#39;RAMP2&#39;: False, &#39;RAMP3&#39;: False, &#39;RASAL3&#39;: False, &#39;RASSF2&#39;: False, &#39;RCSD1&#39;: False, &#39;RNASE6&#39;: False, &#39;ROBO4&#39;: False, &#39;S1PR1&#39;: False, &#39;SAMHD1&#39;: False, &#39;SAMSN1&#39;: False, &#39;SASH3&#39;: False, &#39;SCN7A&#39;: False, &#39;SELPLG&#39;: False, &#39;SH2D3C&#39;: False, &#39;SHANK3&#39;: False, &#39;SIGLEC1&#39;: False, &#39;SIRPA&#39;: True, &#39;SLA&#39;: False, &#39;SLC15A3&#39;: False, &#39;SLC1A3&#39;: False, &#39;SLC7A7&#39;: False, &#39;SLCO2B1&#39;: False, &#39;SLIT2&#39;: False, &#39;SLIT3&#39;: False, &#39;SPARC&#39;: False, &#39;SPARCL1&#39;: False, &#39;SPI1&#39;: False, &#39;SPN&#39;: False, &#39;SPON1&#39;: False, &#39;SSC5D&#39;: False, &#39;STARD8&#39;: False, &#39;STAT1&#39;: False, &#39;SULF1&#39;: False, &#39;SVEP1&#39;: False, &#39;SYNPO2&#39;: False, &#39;TAGLN&#39;: False, &#39;TAP1&#39;: False, &#39;TBC1D10C&#39;: False, &#39;TCF4&#39;: False, &#39;THBS2&#39;: False, &#39;THY1&#39;: False, &#39;TIE1&#39;: False, &#39;TLR4&#39;: False, &#39;TNFSF13B&#39;: False, &#39;TRAC&#39;: False, &#39;TRBC1&#39;: False, &#39;TRBC2&#39;: False, &#39;TREM2&#39;: False, &#39;TRPV2&#39;: False, &#39;TYROBP&#39;: False, &#39;UBE2L6&#39;: False, &#39;UNC5B&#39;: False, &#39;VCAN&#39;: False, &#39;VSIG4&#39;: False, &#39;VWF&#39;: False, &#39;WAS&#39;: True, &#39;ZEB1&#39;: True, &#39;ZEB2&#39;: False}, &#39;community&#39;: {&#39;LST1&#39;: 0, &#39;FPR1&#39;: 0, &#39;PTAFR&#39;: 0, &#39;LIPA&#39;: 0, &#39;HLA-DQB1&#39;: 0, &#39;TRPV2&#39;: 0, &#39;HLA-DPB1&#39;: 0, &#39;C5AR1&#39;: 0, &#39;FCGR3A&#39;: 0, &#39;TNFSF13B&#39;: 0, &#39;LILRB2&#39;: 0, &#39;NCKAP1L&#39;: 0, &#39;HAVCR2&#39;: 0, &#39;CASP1&#39;: 0, &#39;CD300A&#39;: 0, &#39;MS4A4A&#39;: 0, &#39;GLIPR2&#39;: 0, &#39;CD86&#39;: 0, &#39;PIK3AP1&#39;: 0, &#39;HLA-DRB5&#39;: 0, &#39;C3AR1&#39;: 0, &#39;HLA-DMA&#39;: 0, &#39;NCF4&#39;: 0, &#39;LAIR1&#39;: 0, &#39;RNASE6&#39;: 0, &#39;APBB1IP&#39;: 0, &#39;LCP1&#39;: 0, &#39;ITGB2&#39;: 0, &#39;CSF1R&#39;: 0, &#39;FGR&#39;: 0, &#39;TYROBP&#39;: 0, &#39;GYPC&#39;: 0, &#39;NFAM1&#39;: 0, &#39;ACP5&#39;: 0, &#39;MSR1&#39;: 0, &#39;CD52&#39;: 0, &#39;APOC1&#39;: 0, &#39;FCER1G&#39;: 0, &#39;SLC7A7&#39;: 0, &#39;VSIG4&#39;: 0, &#39;LAPTM5&#39;: 0, &#39;LCP2&#39;: 0, &#39;SAMSN1&#39;: 0, &#39;TLR4&#39;: 0, &#39;ITGAM&#39;: 0, &#39;SLC1A3&#39;: 0, &#39;JAML&#39;: 0, &#39;C1QC&#39;: 0, &#39;HLA-DRB1&#39;: 0, &#39;MARCO&#39;: 0, &#39;ALOX5&#39;: 0, &#39;CD163&#39;: 0, &#39;GMFG&#39;: 0, &#39;ALOX5AP&#39;: 0, &#39;MYO1F&#39;: 0, &#39;CD74&#39;: 0, &#39;HLA-DMB&#39;: 0, &#39;EMILIN2&#39;: 0, &#39;ITGAX&#39;: 0, &#39;TREM2&#39;: 0, &#39;CD4&#39;: 0, &#39;CCR1&#39;: 0, &#39;C1orf162&#39;: 0, &#39;DAB2&#39;: 0, &#39;MS4A7&#39;: 0, &#39;HCK&#39;: 0, &#39;SPI1&#39;: 0, &#39;PLEK&#39;: 0, &#39;SLCO2B1&#39;: 0, &#39;PLXNC1&#39;: 0, &#39;SIRPA&#39;: 0, &#39;SIGLEC1&#39;: 0, &#39;FERMT3&#39;: 0, &#39;HLA-DOA&#39;: 0, &#39;SLC15A3&#39;: 0, &#39;DOK2&#39;: 0, &#39;CD53&#39;: 0, &#39;FPR3&#39;: 0, &#39;LILRB4&#39;: 0, &#39;MPEG1&#39;: 0, &#39;MNDA&#39;: 0, &#39;C1QA&#39;: 0, &#39;FGL2&#39;: 0, &#39;C1QB&#39;: 0, &#39;HLA-DRA&#39;: 0, &#39;CD14&#39;: 0, &#39;SAMHD1&#39;: 0, &#39;HK3&#39;: 0, &#39;NCF2&#39;: 0, &#39;MRC1&#39;: 0, &#39;CYBB&#39;: 0, &#39;IGSF6&#39;: 0, &#39;HLA-DQA1&#39;: 0, &#39;CD84&#39;: 0, &#39;AIF1&#39;: 0, &#39;SELPLG&#39;: 0, &#39;FCGR2A&#39;: 0, &#39;MS4A6A&#39;: 0, &#39;CMKLR1&#39;: 0, &#39;BTK&#39;: 0, &#39;HLA-DPA1&#39;: 0, &#39;MSRB3&#39;: 1, &#39;THY1&#39;: 1, &#39;MYLK&#39;: 1, &#39;LTBP2&#39;: 1, &#39;ADAMTS12&#39;: 1, &#39;PRELP&#39;: 1, &#39;UNC5B&#39;: 1, &#39;ADAMTS2&#39;: 1, &#39;SSC5D&#39;: 1, &#39;TAGLN&#39;: 1, &#39;POSTN&#39;: 1, &#39;AEBP1&#39;: 1, &#39;GREM1&#39;: 1, &#39;COL4A2&#39;: 1, &#39;COL15A1&#39;: 1, &#39;ISLR&#39;: 1, &#39;PRRX1&#39;: 1, &#39;SPARC&#39;: 1, &#39;COL5A2&#39;: 1, &#39;CCDC80&#39;: 1, &#39;VCAN&#39;: 1, &#39;FBN1&#39;: 1, &#39;IRAG1&#39;: 1, &#39;FNDC1&#39;: 1, &#39;SYNPO2&#39;: 1, &#39;LAMA4&#39;: 1, &#39;CCN4&#39;: 1, &#39;ANTXR1&#39;: 1, &#39;COL11A1&#39;: 1, &#39;FSTL1&#39;: 1, &#39;FAP&#39;: 1, &#39;MXRA5&#39;: 1, &#39;LUM&#39;: 1, &#39;COL5A1&#39;: 1, &#39;ACTG2&#39;: 1, &#39;PDGFRA&#39;: 1, &#39;MMP11&#39;: 1, &#39;BGN&#39;: 1, &#39;LMOD1&#39;: 1, &#39;MYH11&#39;: 1, &#39;TCF4&#39;: 1, &#39;OLFML2B&#39;: 1, &#39;DCHS1&#39;: 1, &#39;COL10A1&#39;: 1, &#39;COL3A1&#39;: 1, &#39;ANGPTL2&#39;: 1, &#39;ACTA2&#39;: 1, &#39;ITGA11&#39;: 1, &#39;PODN&#39;: 1, &#39;LRRC15&#39;: 1, &#39;EMILIN1&#39;: 1, &#39;COL6A3&#39;: 1, &#39;SULF1&#39;: 1, &#39;SPON1&#39;: 1, &#39;COL1A2&#39;: 1, &#39;ZEB1&#39;: 1, &#39;PDGFRB&#39;: 1, &#39;HTRA3&#39;: 1, &#39;COL1A1&#39;: 1, &#39;ADGRA2&#39;: 1, &#39;LMCD1&#39;: 1, &#39;COL4A1&#39;: 1, &#39;ADAM12&#39;: 1, &#39;DCN&#39;: 1, &#39;CTHRC1&#39;: 1, &#39;MMP2&#39;: 1, &#39;THBS2&#39;: 1, &#39;RAMP3&#39;: 2, &#39;PTPRB&#39;: 2, &#39;STARD8&#39;: 2, &#39;A2M&#39;: 2, &#39;RAMP2&#39;: 2, &#39;CLEC14A&#39;: 2, &#39;BCL6B&#39;: 2, &#39;SH2D3C&#39;: 2, &#39;SCN7A&#39;: 2, &#39;GAS7&#39;: 2, &#39;FHL1&#39;: 2, &#39;CD93&#39;: 2, &#39;GIMAP4&#39;: 2, &#39;LDB2&#39;: 2, &#39;LAMA2&#39;: 2, &#39;AOC3&#39;: 2, &#39;GIMAP6&#39;: 2, &#39;SLIT2&#39;: 2, &#39;TIE1&#39;: 2, &#39;GIMAP8&#39;: 2, &#39;FMO2&#39;: 2, &#39;MFAP4&#39;: 2, &#39;VWF&#39;: 2, &#39;CYYR1&#39;: 2, &#39;ARHGEF6&#39;: 2, &#39;ADH1B&#39;: 2, &#39;EPB41L2&#39;: 2, &#39;FGD5&#39;: 2, &#39;SPARCL1&#39;: 2, &#39;RASSF2&#39;: 2, &#39;HEG1&#39;: 2, &#39;LHFPL6&#39;: 2, &#39;DIPK2B&#39;: 2, &#39;LRRC32&#39;: 2, &#39;PCDH12&#39;: 2, &#39;SLIT3&#39;: 2, &#39;SHANK3&#39;: 2, &#39;PECAM1&#39;: 2, &#39;FLI1&#39;: 2, &#39;PLXDC2&#39;: 2, &#39;ERG&#39;: 2, &#39;EDNRB&#39;: 2, &#39;ABI3BP&#39;: 2, &#39;PREX1&#39;: 2, &#39;ACVRL1&#39;: 2, &#39;ROBO4&#39;: 2, &#39;CALCRL&#39;: 2, &#39;RCSD1&#39;: 2, &#39;ITGA8&#39;: 2, &#39;FLT4&#39;: 2, &#39;CHRDL1&#39;: 2, &#39;SVEP1&#39;: 2, &#39;MMRN2&#39;: 2, &#39;CDH5&#39;: 2, &#39;ZEB2&#39;: 2, &#39;PCDH17&#39;: 2, &#39;C7&#39;: 2, &#39;PTGER4&#39;: 2, &#39;S1PR1&#39;: 2, &#39;INMT&#39;: 2, &#39;CD34&#39;: 2, &#39;GIMAP7&#39;: 2, &#39;IL33&#39;: 2, &#39;NOTCH4&#39;: 2, &#39;ACAP1&#39;: 3, &#39;PTPRC&#39;: 3, &#39;SLA&#39;: 3, &#39;CCR5&#39;: 3, &#39;WAS&#39;: 3, &#39;LCK&#39;: 3, &#39;IRF8&#39;: 3, &#39;LSP1&#39;: 3, &#39;IL16&#39;: 3, &#39;TRAC&#39;: 3, &#39;KCNAB2&#39;: 3, &#39;ITGA4&#39;: 3, &#39;PARVG&#39;: 3, &#39;CCL5&#39;: 3, &#39;CYTIP&#39;: 3, &#39;RASAL3&#39;: 3, &#39;DOK3&#39;: 3, &#39;PPP1R16B&#39;: 3, &#39;CD8A&#39;: 3, &#39;SPN&#39;: 3, &#39;PIK3CG&#39;: 3, &#39;CORO1A&#39;: 3, &#39;SASH3&#39;: 3, &#39;IL2RB&#39;: 3, &#39;TRBC2&#39;: 3, &#39;PRKCB&#39;: 3, &#39;DOCK2&#39;: 3, &#39;CYTH4&#39;: 3, &#39;ARHGAP30&#39;: 3, &#39;CD37&#39;: 3, &#39;KLHL6&#39;: 3, &#39;CSF2RB&#39;: 3, &#39;TRBC1&#39;: 3, &#39;BIN2&#39;: 3, &#39;FGD2&#39;: 3, &#39;PTPN7&#39;: 3, &#39;IL2RG&#39;: 3, &#39;DOCK8&#39;: 3, &#39;CD6&#39;: 3, &#39;PIK3R5&#39;: 3, &#39;IL7R&#39;: 3, &#39;ITGAL&#39;: 3, &#39;CD2&#39;: 3, &#39;CD3E&#39;: 3, &#39;IKZF1&#39;: 3, &#39;ARHGAP9&#39;: 3, &#39;EVI2B&#39;: 3, &#39;TBC1D10C&#39;: 3, &#39;FYB1&#39;: 3, &#39;IL10RA&#39;: 3, &#39;FAM78A&#39;: 3, &#39;STAT1&#39;: 4, &#39;CXCL10&#39;: 4, &#39;HLA-F&#39;: 4, &#39;GBP1&#39;: 4, &#39;GBP5&#39;: 4, &#39;APOL3&#39;: 4, &#39;HCP5&#39;: 4, &#39;UBE2L6&#39;: 4, &#39;TAP1&#39;: 4, &#39;PSMB9&#39;: 4, &#39;GBP4&#39;: 4, &#39;CXCL9&#39;: 4, &#39;HLA-A&#39;: 4, &#39;HLA-B&#39;: 4, &#39;HLA-C&#39;: 4, &#39;F13A1&#39;: 5, &#39;FOLR2&#39;: 5}}
</pre></div>
</div>
<img alt="../_images/d0b9ad3e172a273e8878c3f36131f0c1f3eb62f055ec56285e834b39c0f36753.png" src="../_images/d0b9ad3e172a273e8878c3f36131f0c1f3eb62f055ec56285e834b39c0f36753.png" />
</div>
</div>
</section>
<section id="perform-ora-on-each-community">
<h3>Perform ORA on Each Community<a class="headerlink" href="#perform-ora-on-each-community" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Perform ORA on Each Community: Performing ORA on individual clusters can help in understanding distinct biological significance of each cluster, revealing how certain pathways or functions are associated with specific subsets of genes and potentially uncovering biological themes that may be missed.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">communities</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[frozenset({&#39;ACP5&#39;,
            &#39;AIF1&#39;,
            &#39;ALOX5&#39;,
            &#39;ALOX5AP&#39;,
            &#39;APBB1IP&#39;,
            &#39;APOC1&#39;,
            &#39;BTK&#39;,
            &#39;C1QA&#39;,
            &#39;C1QB&#39;,
            &#39;C1QC&#39;,
            &#39;C1orf162&#39;,
            &#39;C3AR1&#39;,
            &#39;C5AR1&#39;,
            &#39;CASP1&#39;,
            &#39;CCR1&#39;,
            &#39;CD14&#39;,
            &#39;CD163&#39;,
            &#39;CD300A&#39;,
            &#39;CD4&#39;,
            &#39;CD52&#39;,
            &#39;CD53&#39;,
            &#39;CD74&#39;,
            &#39;CD84&#39;,
            &#39;CD86&#39;,
            &#39;CMKLR1&#39;,
            &#39;CSF1R&#39;,
            &#39;CYBB&#39;,
            &#39;DAB2&#39;,
            &#39;DOK2&#39;,
            &#39;EMILIN2&#39;,
            &#39;FCER1G&#39;,
            &#39;FCGR2A&#39;,
            &#39;FCGR3A&#39;,
            &#39;FERMT3&#39;,
            &#39;FGL2&#39;,
            &#39;FGR&#39;,
            &#39;FPR1&#39;,
            &#39;FPR3&#39;,
            &#39;GLIPR2&#39;,
            &#39;GMFG&#39;,
            &#39;GYPC&#39;,
            &#39;HAVCR2&#39;,
            &#39;HCK&#39;,
            &#39;HK3&#39;,
            &#39;HLA-DMA&#39;,
            &#39;HLA-DMB&#39;,
            &#39;HLA-DOA&#39;,
            &#39;HLA-DPA1&#39;,
            &#39;HLA-DPB1&#39;,
            &#39;HLA-DQA1&#39;,
            &#39;HLA-DQB1&#39;,
            &#39;HLA-DRA&#39;,
            &#39;HLA-DRB1&#39;,
            &#39;HLA-DRB5&#39;,
            &#39;IGSF6&#39;,
            &#39;ITGAM&#39;,
            &#39;ITGAX&#39;,
            &#39;ITGB2&#39;,
            &#39;JAML&#39;,
            &#39;LAIR1&#39;,
            &#39;LAPTM5&#39;,
            &#39;LCP1&#39;,
            &#39;LCP2&#39;,
            &#39;LILRB2&#39;,
            &#39;LILRB4&#39;,
            &#39;LIPA&#39;,
            &#39;LST1&#39;,
            &#39;MARCO&#39;,
            &#39;MNDA&#39;,
            &#39;MPEG1&#39;,
            &#39;MRC1&#39;,
            &#39;MS4A4A&#39;,
            &#39;MS4A6A&#39;,
            &#39;MS4A7&#39;,
            &#39;MSR1&#39;,
            &#39;MYO1F&#39;,
            &#39;NCF2&#39;,
            &#39;NCF4&#39;,
            &#39;NCKAP1L&#39;,
            &#39;NFAM1&#39;,
            &#39;PIK3AP1&#39;,
            &#39;PLEK&#39;,
            &#39;PLXNC1&#39;,
            &#39;PTAFR&#39;,
            &#39;RNASE6&#39;,
            &#39;SAMHD1&#39;,
            &#39;SAMSN1&#39;,
            &#39;SELPLG&#39;,
            &#39;SIGLEC1&#39;,
            &#39;SIRPA&#39;,
            &#39;SLC15A3&#39;,
            &#39;SLC1A3&#39;,
            &#39;SLC7A7&#39;,
            &#39;SLCO2B1&#39;,
            &#39;SPI1&#39;,
            &#39;TLR4&#39;,
            &#39;TNFSF13B&#39;,
            &#39;TREM2&#39;,
            &#39;TRPV2&#39;,
            &#39;TYROBP&#39;,
            &#39;VSIG4&#39;}),
 frozenset({&#39;ACTA2&#39;,
            &#39;ACTG2&#39;,
            &#39;ADAM12&#39;,
            &#39;ADAMTS12&#39;,
            &#39;ADAMTS2&#39;,
            &#39;ADGRA2&#39;,
            &#39;AEBP1&#39;,
            &#39;ANGPTL2&#39;,
            &#39;ANTXR1&#39;,
            &#39;BGN&#39;,
            &#39;CCDC80&#39;,
            &#39;CCN4&#39;,
            &#39;COL10A1&#39;,
            &#39;COL11A1&#39;,
            &#39;COL15A1&#39;,
            &#39;COL1A1&#39;,
            &#39;COL1A2&#39;,
            &#39;COL3A1&#39;,
            &#39;COL4A1&#39;,
            &#39;COL4A2&#39;,
            &#39;COL5A1&#39;,
            &#39;COL5A2&#39;,
            &#39;COL6A3&#39;,
            &#39;CTHRC1&#39;,
            &#39;DCHS1&#39;,
            &#39;DCN&#39;,
            &#39;EMILIN1&#39;,
            &#39;FAP&#39;,
            &#39;FBN1&#39;,
            &#39;FNDC1&#39;,
            &#39;FSTL1&#39;,
            &#39;GREM1&#39;,
            &#39;HTRA3&#39;,
            &#39;IRAG1&#39;,
            &#39;ISLR&#39;,
            &#39;ITGA11&#39;,
            &#39;LAMA4&#39;,
            &#39;LMCD1&#39;,
            &#39;LMOD1&#39;,
            &#39;LRRC15&#39;,
            &#39;LTBP2&#39;,
            &#39;LUM&#39;,
            &#39;MMP11&#39;,
            &#39;MMP2&#39;,
            &#39;MSRB3&#39;,
            &#39;MXRA5&#39;,
            &#39;MYH11&#39;,
            &#39;MYLK&#39;,
            &#39;OLFML2B&#39;,
            &#39;PDGFRA&#39;,
            &#39;PDGFRB&#39;,
            &#39;PODN&#39;,
            &#39;POSTN&#39;,
            &#39;PRELP&#39;,
            &#39;PRRX1&#39;,
            &#39;SPARC&#39;,
            &#39;SPON1&#39;,
            &#39;SSC5D&#39;,
            &#39;SULF1&#39;,
            &#39;SYNPO2&#39;,
            &#39;TAGLN&#39;,
            &#39;TCF4&#39;,
            &#39;THBS2&#39;,
            &#39;THY1&#39;,
            &#39;UNC5B&#39;,
            &#39;VCAN&#39;,
            &#39;ZEB1&#39;}),
 frozenset({&#39;A2M&#39;,
            &#39;ABI3BP&#39;,
            &#39;ACVRL1&#39;,
            &#39;ADH1B&#39;,
            &#39;AOC3&#39;,
            &#39;ARHGEF6&#39;,
            &#39;BCL6B&#39;,
            &#39;C7&#39;,
            &#39;CALCRL&#39;,
            &#39;CD34&#39;,
            &#39;CD93&#39;,
            &#39;CDH5&#39;,
            &#39;CHRDL1&#39;,
            &#39;CLEC14A&#39;,
            &#39;CYYR1&#39;,
            &#39;DIPK2B&#39;,
            &#39;EDNRB&#39;,
            &#39;EPB41L2&#39;,
            &#39;ERG&#39;,
            &#39;FGD5&#39;,
            &#39;FHL1&#39;,
            &#39;FLI1&#39;,
            &#39;FLT4&#39;,
            &#39;FMO2&#39;,
            &#39;GAS7&#39;,
            &#39;GIMAP4&#39;,
            &#39;GIMAP6&#39;,
            &#39;GIMAP7&#39;,
            &#39;GIMAP8&#39;,
            &#39;HEG1&#39;,
            &#39;IL33&#39;,
            &#39;INMT&#39;,
            &#39;ITGA8&#39;,
            &#39;LAMA2&#39;,
            &#39;LDB2&#39;,
            &#39;LHFPL6&#39;,
            &#39;LRRC32&#39;,
            &#39;MFAP4&#39;,
            &#39;MMRN2&#39;,
            &#39;NOTCH4&#39;,
            &#39;PCDH12&#39;,
            &#39;PCDH17&#39;,
            &#39;PECAM1&#39;,
            &#39;PLXDC2&#39;,
            &#39;PREX1&#39;,
            &#39;PTGER4&#39;,
            &#39;PTPRB&#39;,
            &#39;RAMP2&#39;,
            &#39;RAMP3&#39;,
            &#39;RASSF2&#39;,
            &#39;RCSD1&#39;,
            &#39;ROBO4&#39;,
            &#39;S1PR1&#39;,
            &#39;SCN7A&#39;,
            &#39;SH2D3C&#39;,
            &#39;SHANK3&#39;,
            &#39;SLIT2&#39;,
            &#39;SLIT3&#39;,
            &#39;SPARCL1&#39;,
            &#39;STARD8&#39;,
            &#39;SVEP1&#39;,
            &#39;TIE1&#39;,
            &#39;VWF&#39;,
            &#39;ZEB2&#39;}),
 frozenset({&#39;ACAP1&#39;,
            &#39;ARHGAP30&#39;,
            &#39;ARHGAP9&#39;,
            &#39;BIN2&#39;,
            &#39;CCL5&#39;,
            &#39;CCR5&#39;,
            &#39;CD2&#39;,
            &#39;CD37&#39;,
            &#39;CD3E&#39;,
            &#39;CD6&#39;,
            &#39;CD8A&#39;,
            &#39;CORO1A&#39;,
            &#39;CSF2RB&#39;,
            &#39;CYTH4&#39;,
            &#39;CYTIP&#39;,
            &#39;DOCK2&#39;,
            &#39;DOCK8&#39;,
            &#39;DOK3&#39;,
            &#39;EVI2B&#39;,
            &#39;FAM78A&#39;,
            &#39;FGD2&#39;,
            &#39;FYB1&#39;,
            &#39;IKZF1&#39;,
            &#39;IL10RA&#39;,
            &#39;IL16&#39;,
            &#39;IL2RB&#39;,
            &#39;IL2RG&#39;,
            &#39;IL7R&#39;,
            &#39;IRF8&#39;,
            &#39;ITGA4&#39;,
            &#39;ITGAL&#39;,
            &#39;KCNAB2&#39;,
            &#39;KLHL6&#39;,
            &#39;LCK&#39;,
            &#39;LSP1&#39;,
            &#39;PARVG&#39;,
            &#39;PIK3CG&#39;,
            &#39;PIK3R5&#39;,
            &#39;PPP1R16B&#39;,
            &#39;PRKCB&#39;,
            &#39;PTPN7&#39;,
            &#39;PTPRC&#39;,
            &#39;RASAL3&#39;,
            &#39;SASH3&#39;,
            &#39;SLA&#39;,
            &#39;SPN&#39;,
            &#39;TBC1D10C&#39;,
            &#39;TRAC&#39;,
            &#39;TRBC1&#39;,
            &#39;TRBC2&#39;,
            &#39;WAS&#39;}),
 frozenset({&#39;APOL3&#39;,
            &#39;CXCL10&#39;,
            &#39;CXCL9&#39;,
            &#39;GBP1&#39;,
            &#39;GBP4&#39;,
            &#39;GBP5&#39;,
            &#39;HCP5&#39;,
            &#39;HLA-A&#39;,
            &#39;HLA-B&#39;,
            &#39;HLA-C&#39;,
            &#39;HLA-F&#39;,
            &#39;PSMB9&#39;,
            &#39;STAT1&#39;,
            &#39;TAP1&#39;,
            &#39;UBE2L6&#39;}),
 frozenset({&#39;F13A1&#39;, &#39;FOLR2&#39;})]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to perform ORA on a given list of genes</span>
<span class="k">def</span> <span class="nf">communityORA</span><span class="p">(</span><span class="n">genes</span><span class="p">):</span>
    <span class="n">enr</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">enrichr</span><span class="p">(</span><span class="n">gene_list</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span> <span class="n">gene_sets</span><span class="o">=</span><span class="p">[</span><span class="n">gene_sets</span><span class="p">],</span> <span class="n">organism</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">enr</span>

<span class="c1"># Convert communities to lists for ORA</span>
<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">communities</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">community</span><span class="p">)</span> <span class="k">for</span> <span class="n">community</span> <span class="ow">in</span> <span class="n">communities</span><span class="p">]</span>
<span class="n">communities</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">communities</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="nb">len</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Perform ORA for three selected communities</span>
<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">community1_enr</span> <span class="o">=</span> <span class="n">communityORA</span><span class="p">(</span><span class="n">communities</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">community2_enr</span> <span class="o">=</span> <span class="n">communityORA</span><span class="p">(</span><span class="n">communities</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">community3_enr</span> <span class="o">=</span> <span class="n">communityORA</span><span class="p">(</span><span class="n">communities</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>

<span class="c1"># Print the top 10 results for each community</span>
<span class="n">x</span> <span class="o">=</span> <span class="n">PrettyTable</span><span class="p">()</span>
<span class="n">x</span><span class="o">.</span><span class="n">field_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Community 1&quot;</span><span class="p">,</span> <span class="s2">&quot;Community 2&quot;</span><span class="p">,</span> <span class="s2">&quot;Community 3&quot;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
    <span class="n">x</span><span class="o">.</span><span class="n">add_row</span><span class="p">([</span><span class="n">community1_enr</span><span class="o">.</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;Term&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">community2_enr</span><span class="o">.</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;Term&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">community3_enr</span><span class="o">.</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;Term&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>+----------------------------------------------+------------------------------------------------------+-------------------------------------------------+
|                 Community 1                  |                     Community 2                      |                   Community 3                   |
+----------------------------------------------+------------------------------------------------------+-------------------------------------------------+
|       Staphylococcus aureus infection        |           Protein digestion and absorption           |             Cell adhesion molecules             |
|                  Phagosome                   |                    Focal adhesion                    |       Complement and coagulation cascades       |
|                Leishmaniasis                 |               ECM-receptor interaction               |             ECM-receptor interaction            |
|                 Tuberculosis                 |              Relaxin signaling pathway               |                  Focal adhesion                 |
|             Rheumatoid arthritis             |              PI3K-Akt signaling pathway              |          Human papillomavirus infection         |
|                    Asthma                    | AGE-RAGE signaling pathway in diabetic complications |               Tyrosine metabolism               |
|         Systemic lupus erythematosus         |                      Amoebiasis                      |        Vascular smooth muscle contraction       |
| Intestinal immune network for IgA production |            Human papillomavirus infection            |     Neuroactive ligand-receptor interaction     |
|           Cell adhesion molecules            |          Vascular smooth muscle contraction          | Arrhythmogenic right ventricular cardiomyopathy |
|             Allograft rejection              |               Proteoglycans in cancer                |            PI3K-Akt signaling pathway           |
+----------------------------------------------+------------------------------------------------------+-------------------------------------------------+
</pre></div>
</div>
</div>
</div>
</section>
<section id="gene-set-enrichment-analysis-gsea">
<h3>1.2 Gene Set Enrichment Analysis (GSEA)<a class="headerlink" href="#gene-set-enrichment-analysis-gsea" title="Link to this heading">#</a></h3>
<p>Gene Set Enrichment Analysis (GSEA) is a genome-wide expression analysis method designed to interpret expression profiles focusing on pre-defined gene sets. These gene sets are curated based on prior biological knowledge, such as published information about biochemical pathways or patterns of coexpression observed in previous experiment studies. The genes can be ordered in a ranked list, according to their differential expression between the classes. The primary objective of GSEA is to assess whether the genes within a given gene set tend to occur toward the top (or bottom) of the ranked list. This ranking is based on the correlation between gene expression and a particular phenotypic class distinction. By evaluating the distribution of gene set members within the ranked list, GSEA identifies whether the set is correlated with the phenotypic class, thus providing insights into underlying biological mechanisms. This method contrasts with traditional single-gene analysis by focusing on the collective behavior of gene sets, thereby uncovering biologically significant patterns that might be overlooked when examining individual genes in isolation. We use <code class="docutils literal notranslate"><span class="pre">gseapy</span></code> to perform GSEA on <code class="docutils literal notranslate"><span class="pre">KEGG_2021_Human</span></code> assigned classes based on a phenotypic attribute (e.g., smoking status). (ref)</p>
<!-- (https://pnnl-comp-mass-spec.github.io/proteomics-data-analysis-tutorial/gsea.html) for referencing.  -->
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Display the first five rows of the dataframe &quot;tcga_gxp_df&quot;</span>
<span class="c1">### YOUR CODE HERE ###</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>        TCGA-38-7271  TCGA-55-7914  TCGA-95-7043  TCGA-73-4658  TCGA-86-8076  \
GENES                                                                          
TSPAN6       11.3668       11.5434       11.4110       12.2149       11.2882   
DPM1         10.3673       10.5282       11.2018       10.3249       10.2095   
SCYL3         9.7884        9.7292        9.4449        9.2340        9.8186   
FIRRM         8.2552        7.9951        8.3546        7.7537        7.9844   
FGR          10.6173        8.4858        7.3211       11.0656       10.5213   

        TCGA-55-7726  TCGA-44-6147  TCGA-50-5932  TCGA-44-2661  TCGA-86-7954  \
GENES                                                                          
TSPAN6       11.2084       11.1717       12.1281       11.9908       12.2601   
DPM1         11.6034       10.5684       10.5895       10.8699       10.8979   
SCYL3         9.0884       10.2428       10.1989        9.8516        9.7973   
FIRRM         8.7984        8.8721        8.3550        8.1877        9.2347   
FGR           9.3386        9.3440        8.7403       10.6243        9.8945   

        TCGA-73-4662  TCGA-44-7671  TCGA-78-8660  TCGA-62-A46P  TCGA-55-6978  \
GENES                                                                          
TSPAN6       11.6963       11.4364       10.8184       11.5513       11.2016   
DPM1         10.2597       10.2811       10.5428       11.4632       10.4191   
SCYL3        10.4597        9.5387        9.1616        9.6713        9.1914   
FIRRM         9.8346        7.8180        9.0867        7.4249        7.2293   
FGR          10.1067        8.2663       10.5393        9.3087       11.7055   

        TCGA-50-6592  TCGA-38-4625  TCGA-80-5611  TCGA-86-8054  TCGA-55-6986  \
GENES                                                                          
TSPAN6       11.5201       12.3420       11.9378       12.6268       12.3495   
DPM1         11.1089       11.7516       11.0717       11.3598       10.5437   
SCYL3         9.1989        9.0392        9.7859        9.5704        9.3099   
FIRRM         9.1192       10.0768        9.4085       10.0984        7.4506   
FGR          10.1633        9.9218       10.7935        8.6680        9.3646   

        TCGA-L9-A5IP  TCGA-69-7764  TCGA-49-6744  TCGA-75-5125  TCGA-38-4626  \
GENES                                                                          
TSPAN6       10.0858       10.1934       11.4793       11.8151       10.9083   
DPM1         10.1693       10.4140       10.5354       10.7697       11.1321   
SCYL3         9.9071       10.2033        9.4620        8.8882        8.9993   
FIRRM         9.3028        8.9049        8.0116        8.5089        7.1778   
FGR           8.4917        8.3683       10.5502       10.4143       13.7653   

        TCGA-69-7763  TCGA-86-8279  TCGA-93-8067  TCGA-97-8179  TCGA-55-A48Y  \
GENES                                                                          
TSPAN6       10.9789       12.3310       11.0959       12.2844       12.2692   
DPM1         10.9468       11.0926       10.7382       10.8152       10.3957   
SCYL3         9.2914       10.1648        9.6722        9.8902        9.5831   
FIRRM         7.2466        9.0061        8.9549        9.0305        8.8959   
FGR           9.8602        8.9363        9.0198        9.2149        9.1690   

        TCGA-86-8055  TCGA-91-6835  TCGA-55-6982  TCGA-55-A4DF  TCGA-44-6774  \
GENES                                                                          
TSPAN6       12.4793       11.6572       10.4046       11.7654       11.5185   
DPM1         10.4709       10.1559        9.9805       11.4481       10.4410   
SCYL3         9.6185        9.3662        9.4164        9.3445        9.0884   
FIRRM         8.5437        7.0665        7.0667        9.0817        8.4102   
FGR           9.9654       10.4230       11.6101       10.4511        9.6184   

        TCGA-50-5066  TCGA-05-5423  TCGA-67-3774  TCGA-97-A4M2  TCGA-95-7567  \
GENES                                                                          
TSPAN6       12.6804       12.6905       11.6912       10.3112       10.6976   
DPM1         12.3449       12.0215        9.8339        9.8155       11.1057   
SCYL3         9.9466       10.5563        9.5650        9.5591        9.8518   
FIRRM         9.6505        9.6165        8.0169        7.4297        9.2275   
FGR           9.9466       10.1096       10.0554       12.5322        9.0607   

        TCGA-49-AAR0  TCGA-44-2656  TCGA-53-7813  TCGA-O1-A52J  TCGA-35-4122  \
GENES                                                                          
TSPAN6       10.3748       11.7572       11.9412       11.4387       13.5286   
DPM1         10.6229       10.3072       10.9156       10.6567       11.9269   
SCYL3         9.2797        9.9954       10.1827        9.6889        9.2378   
FIRRM         8.1630        8.6609        8.6540        8.5377       10.2647   
FGR          11.0522       10.5198        8.5024       11.5016       10.8670   

        TCGA-55-8092  TCGA-49-6761  TCGA-49-4507  TCGA-55-7816  TCGA-78-7145  \
GENES                                                                          
TSPAN6       11.3330       10.4583       10.6897        9.8078       12.0396   
DPM1         11.2991       10.1423       10.3601       10.1592       10.7380   
SCYL3        10.1650        9.2111        8.3509        9.3441        9.2187   
FIRRM         9.1622        6.8973        7.7967        7.6247        9.1262   
FGR           8.8416       12.2780        8.8244       11.6315        9.7211   

        TCGA-55-6983  TCGA-53-7624  TCGA-97-A4M3  TCGA-50-5068  TCGA-78-8648  \
GENES                                                                          
TSPAN6       10.7139       10.9416       10.6148       12.3594       10.2641   
DPM1         10.4286       10.4356       10.3590       12.4895       10.1547   
SCYL3         9.2483        9.7711        9.4156       10.4581        8.9691   
FIRRM         6.8227        9.6407        7.7675        8.8511        7.9248   
FGR          11.8538        8.3346        8.5788        9.9279       12.0813   

        TCGA-44-6778  TCGA-80-5608  TCGA-86-8281  TCGA-MP-A4T6  TCGA-55-8085  \
GENES                                                                          
TSPAN6       10.3341       12.1122       12.6116       11.6780       11.3810   
DPM1         10.5626       10.0477       10.5382        9.9913       10.2894   
SCYL3         9.2670        9.5723       10.6370       10.3945       10.0023   
FIRRM         7.1463        8.5977        8.7141        8.4354        9.3176   
FGR          12.3180        9.6880        8.6507       10.0116        9.9325   

        TCGA-62-8399  TCGA-97-A4M5  TCGA-97-7547  TCGA-05-5429  TCGA-55-7994  \
GENES                                                                          
TSPAN6       11.2908       11.6346       12.7436       12.0022       11.3310   
DPM1         10.8054       10.1605       10.0190       11.5081       10.0984   
SCYL3         9.4692        9.3656       10.1750       10.6553        9.5655   
FIRRM         9.1287        7.7488        7.4630        9.1531        9.0223   
FGR           9.8269       10.7108        9.0030        7.6588       10.6705   

        TCGA-55-8094  TCGA-05-4425  TCGA-44-4112  TCGA-49-6767  TCGA-49-4490  \
GENES                                                                          
TSPAN6       12.1321       10.9049       10.8718       13.6990       11.2008   
DPM1         11.3601       10.6430       10.6521       11.1387       11.3382   
SCYL3         9.6904        9.1352       10.5748        9.0598        9.3388   
FIRRM         8.8943        8.0016        8.9260        9.3683        7.3502   
FGR           7.2463       10.0586        8.2823        9.8099       11.0707   

        TCGA-MP-A4T9  TCGA-50-5942  TCGA-MP-A4SV  TCGA-49-AAR4  TCGA-05-4397  \
GENES                                                                          
TSPAN6       12.5260       11.3536       10.9310       11.1425       11.6935   
DPM1         10.2255       10.0836       10.4917       10.0090       11.9545   
SCYL3         9.0138       10.1941        9.7808        9.2933        9.5845   
FIRRM         7.9927        7.5116        9.1125        8.6168       10.1938   
FGR          10.4863        8.3067        9.7500       10.6279        9.5733   

        TCGA-44-A47A  TCGA-86-8359  TCGA-78-7539  TCGA-MP-A4T8  TCGA-99-8032  \
GENES                                                                          
TSPAN6       10.7492       11.5179       11.4193       12.0642       10.9989   
DPM1         10.0906       11.3592        9.9883       10.3982       10.5526   
SCYL3        10.2226        9.7462        9.5927        9.3558        9.2934   
FIRRM         8.4295        8.3625        8.2439        8.6173        8.4754   
FGR          10.6769        9.8680       10.4574        8.1190        8.9752   

        TCGA-50-6595  TCGA-55-6968  TCGA-44-8120  TCGA-55-8302  TCGA-99-8025  \
GENES                                                                          
TSPAN6       10.1955        9.8547       12.1975       11.8144       11.3360   
DPM1         10.2111       10.2224       10.3080       10.4408       10.1729   
SCYL3         9.5384        9.4034        9.1338        9.0750        9.5706   
FIRRM         6.9050        7.3625        7.4183        8.6270        8.2785   
FGR          11.6027       12.6021        9.4494       10.2858        8.6901   

        TCGA-64-1679  TCGA-95-8039  TCGA-44-A479  TCGA-44-6148  TCGA-NJ-A55O  \
GENES                                                                          
TSPAN6       11.8734       11.6592       10.9656       11.3940       10.0536   
DPM1         10.0785       10.3417       11.0045        9.9599       10.3394   
SCYL3         8.9542        9.3475        9.1997        9.3012       10.0962   
FIRRM         8.6444        8.0100        8.3573        7.2562        8.2185   
FGR           9.5814        9.9495        9.8711       11.3526       11.0880   

        TCGA-MP-A5C7  TCGA-64-5778  TCGA-55-6971  TCGA-49-AARN  TCGA-44-A47B  \
GENES                                                                          
TSPAN6       11.2797       12.5194       11.1115       11.3718       11.0992   
DPM1         10.2760       11.3405       10.1568       10.4120       10.2278   
SCYL3        10.6254        9.6891        9.6099        9.7300        9.3831   
FIRRM         8.4154        9.1849        7.8837        8.2931        8.8700   
FGR           7.6519        9.8892       10.6668        8.6482        9.6278   

        TCGA-55-5899  TCGA-49-AAQV  TCGA-99-8028  TCGA-75-6205  TCGA-97-8552  \
GENES                                                                          
TSPAN6       10.5568       12.4242       10.8483       10.2531       12.1970   
DPM1         10.8406       10.5518       10.6743       10.1342       10.6351   
SCYL3         8.8079        9.7910        8.8135        9.0254        9.3585   
FIRRM         8.8411        8.7850        7.2378        8.3485        7.1572   
FGR           9.3992        8.9880       11.3194       10.8624       10.9816   

        TCGA-50-8459  TCGA-05-5425  TCGA-78-7150  TCGA-86-A4P7  TCGA-49-4512  \
GENES                                                                          
TSPAN6       11.0435       11.5506       11.8431       10.0167       11.1656   
DPM1         10.0925       11.4544       10.6302       10.0922       10.3143   
SCYL3         8.9728        9.6625        9.4022        9.1939        8.8231   
FIRRM         6.9420        9.2182        9.6594        7.7147        7.1512   
FGR          11.6053       10.5979        7.9816       10.6734       10.0487   

        TCGA-55-8206  TCGA-55-8614  TCGA-64-5815  TCGA-L9-A50W  TCGA-73-4675  \
GENES                                                                          
TSPAN6       11.6305       12.1896       11.0415       11.2189       11.8492   
DPM1         10.0360       10.5034       10.0045       10.0630        9.8141   
SCYL3        10.0313        9.0316        8.7873        9.5787        9.4269   
FIRRM         7.6799        8.5778        7.9356        7.2182        7.3954   
FGR          11.1478        8.7977       10.6032        9.0852        9.8649   

        TCGA-55-7995  TCGA-05-4433  TCGA-55-7727  TCGA-44-2668  TCGA-44-A4SU  \
GENES                                                                          
TSPAN6       11.3662       10.4561       11.6218       10.8172       12.8664   
DPM1         10.3974       10.2130       10.9346       10.9549       10.2702   
SCYL3         9.5178       10.0869        9.9300        9.7401        9.8480   
FIRRM         8.9182        8.2565        8.4605        9.2185        8.1700   
FGR          10.7973       11.6060        8.7619       10.3262        8.5420   

        TCGA-55-7907  TCGA-69-7765  TCGA-49-4487  TCGA-44-2662  TCGA-67-6216  \
GENES                                                                          
TSPAN6       11.5225       11.3115       11.5578       10.2480       12.5978   
DPM1         10.4995       10.5450       11.4992       10.0573        9.6467   
SCYL3         9.3712        9.7986        9.4122        8.5703        9.3572   
FIRRM         8.8033        8.3172        9.7413        6.4640        8.4542   
FGR           9.4730        9.5663       10.6520       12.2258        9.6665   

        TCGA-55-7283  TCGA-86-8280  TCGA-91-6840  TCGA-78-7154  TCGA-49-4488  \
GENES                                                                          
TSPAN6       11.3813       12.2305       12.8613       12.2530       11.6121   
DPM1         10.1493       10.0933       11.0881       12.0558       11.0665   
SCYL3         9.4105        9.8800        9.4892        9.5534        9.9758   
FIRRM         8.4359        8.2728        9.2869        9.2661        8.2354   
FGR           9.4678       10.2449        9.3221        8.5334        9.9490   

        TCGA-93-7348  TCGA-62-A470  TCGA-78-7147  TCGA-50-5936  TCGA-55-6984  \
GENES                                                                          
TSPAN6       11.4228       10.9524       12.5578       10.0787       10.5769   
DPM1         10.3363       10.5444       10.3785       10.1670       10.3885   
SCYL3         9.7279        9.8356        9.5767        9.4541       10.2115   
FIRRM         8.4339        8.5297        8.7293        6.9337        8.1152   
FGR           8.9464       10.0200        8.6182       12.3758        9.1607   

        TCGA-50-5941  TCGA-69-7978  TCGA-78-7220  TCGA-55-8616  TCGA-44-A4SS  \
GENES                                                                          
TSPAN6       12.1344       11.0673       11.5503       11.5436       11.9898   
DPM1         10.8217       10.4592       11.4573       10.3285       10.6120   
SCYL3         9.6805        8.8872        9.8853        9.4723        9.6099   
FIRRM         8.3272        8.0495        9.6303        8.3736        9.0144   
FGR          10.6504       10.8448        8.2533        9.1937       10.5314   

        TCGA-55-7570  TCGA-78-7146  TCGA-44-3398  TCGA-05-5420  TCGA-50-5072  \
GENES                                                                          
TSPAN6       11.0150       11.7263       12.6511       12.7900       11.3529   
DPM1         11.0075       10.4944       11.1837       11.7792       10.1174   
SCYL3         9.4472        8.9678        9.5603       10.3938       10.0234   
FIRRM         9.8250        9.5564        8.4491        9.3827        9.7339   
FGR           8.0932        9.0344       10.3383       11.5452        9.3794   

        TCGA-05-4396  TCGA-05-4405  TCGA-50-5935  TCGA-38-4629  TCGA-55-8619  \
GENES                                                                          
TSPAN6       10.8869       11.7057       10.8965       12.3545       10.2583   
DPM1         10.3780        9.6809       10.1563       11.0345        9.9656   
SCYL3         9.7362        9.9193        9.1959        8.6483        9.2934   
FIRRM         7.6906        8.7026        6.5987        8.8552        7.3927   
FGR           7.9106        9.2890       11.5038       10.5689       11.5897   

        TCGA-05-4410  TCGA-73-4676  TCGA-97-8172  TCGA-44-7661  TCGA-05-4384  \
GENES                                                                          
TSPAN6       11.3722       10.6833       10.0881       10.4599       10.7081   
DPM1          9.6733       10.8477        9.6414       10.4871       10.1464   
SCYL3         9.9753        8.8689       10.2857        9.1127       10.0076   
FIRRM         8.3048        7.3380        7.7705        8.6787        7.9360   
FGR           9.9777       12.7109        9.8607       10.6532       10.3516   

        TCGA-44-2655  TCGA-80-5607  TCGA-67-3770  TCGA-91-6836  TCGA-95-7562  \
GENES                                                                          
TSPAN6       12.0397       11.7018       12.8565       10.2880       12.7978   
DPM1         10.5615       10.7224       10.3989       10.2802       10.9204   
SCYL3        10.2159        8.6584        9.8026        9.4232       10.0368   
FIRRM         8.1440        8.2584        7.6809        6.7938       10.2544   
FGR           9.3908        9.6290        9.9126       12.3441        9.3238   

        TCGA-55-8511  TCGA-44-6776  TCGA-95-7948  TCGA-91-7771  TCGA-50-5944  \
GENES                                                                          
TSPAN6       11.4706       10.8476       11.0805       11.9734       12.0851   
DPM1         10.2679       10.4497       10.7129       11.1628       10.4072   
SCYL3         9.4994        9.1809        9.6941        9.4753        9.7173   
FIRRM         8.6699        6.8795        7.6993        8.0260        7.6870   
FGR           9.7591       11.6668        7.4797       10.3716        9.5435   

        TCGA-MN-A4N5  TCGA-73-4677  TCGA-78-7540  TCGA-91-6829  TCGA-78-8640  \
GENES                                                                          
TSPAN6       12.1795       11.5371       10.6095       11.0963       10.3040   
DPM1         11.0103       10.5530       10.7263       10.4121       10.6011   
SCYL3         9.6417        9.9378        9.5271        9.2557       10.0794   
FIRRM         9.1015        7.9151        7.7997        8.4602        9.3955   
FGR           9.3356        9.5930       10.9491        8.8594       10.3561   

        TCGA-62-8398  TCGA-55-8512  TCGA-83-5908  TCGA-55-6987  TCGA-93-A4JP  \
GENES                                                                          
TSPAN6       11.8762       11.8054       11.2086       11.3926       11.8264   
DPM1         10.4475       10.4723       11.0430       10.3518       10.2918   
SCYL3         9.6618        9.4519        9.9135        9.8384        9.5287   
FIRRM         9.3654        7.1228       10.1851        8.9583        8.6005   
FGR           9.6110       10.3872        9.7308       11.2202       10.9032   

        TCGA-73-A9RS  TCGA-L4-A4E5  TCGA-86-8074  TCGA-86-8358  TCGA-78-7158  \
GENES                                                                          
TSPAN6       10.1947       10.4312       12.0766        9.8166       11.1579   
DPM1         11.4870       10.6665       10.6233       10.2224        9.8892   
SCYL3         9.9231        9.5156        9.7516        9.2277       11.8243   
FIRRM         9.1412        8.2863        9.0798        9.8016        8.4744   
FGR           8.9799        8.3602        9.2764        8.4090        8.3675   

        TCGA-91-8497  TCGA-49-AARO  TCGA-78-7159  TCGA-55-7227  TCGA-86-7714  \
GENES                                                                          
TSPAN6        9.9593       11.9509       12.1032       11.3955       12.1417   
DPM1          9.8752       10.2606       10.4278       10.1204       10.5046   
SCYL3         9.4455        9.2924        9.3730        9.2214        9.6726   
FIRRM         7.2335        8.1654        9.1363        7.8797        8.3459   
FGR          10.6002       10.9577        9.2030       10.3880        9.4056   

        TCGA-L9-A7SV  TCGA-78-7143  TCGA-91-8499  TCGA-49-AAR3  TCGA-55-8620  \
GENES                                                                          
TSPAN6       10.2608       11.8510       14.0084       10.9162       10.3611   
DPM1         10.5811        9.9467       11.3648       10.6513       10.7957   
SCYL3         9.1222        9.8190        9.1053        9.3210        7.9588   
FIRRM         8.1668        8.8832        9.6468        9.1314        8.0587   
FGR           8.3870        9.6441        9.7136       10.5115        9.5767   

        TCGA-69-8255  TCGA-75-6207  TCGA-62-A46Y  TCGA-NJ-A4YF  TCGA-91-6830  \
GENES                                                                          
TSPAN6       11.0810       11.3337       12.2993       10.7740       11.7357   
DPM1         11.3752       10.8785       10.6078       12.4100       10.6967   
SCYL3         9.3426        9.7862        9.6415        8.9893        9.3399   
FIRRM         8.6297        8.7160        7.9313        8.2261        8.4090   
FGR           9.7196        9.4578       10.1899        9.4236        9.9364   

        TCGA-62-8395  TCGA-49-4486  TCGA-44-6145  TCGA-86-A4P8  TCGA-78-7537  \
GENES                                                                          
TSPAN6       10.5070       11.8349       10.1911        9.9479       11.2712   
DPM1         10.1885       11.0354       10.5423       10.0410        9.7252   
SCYL3         9.3363       10.2430        8.8740        9.4316        9.7870   
FIRRM         7.5129        8.0114        6.7926        7.5435        7.5338   
FGR           8.9602        7.7962       12.3873       11.9105        8.7270   

        TCGA-44-3919  TCGA-35-4123  TCGA-62-8394  TCGA-69-7761  TCGA-62-A46U  \
GENES                                                                          
TSPAN6       12.0021       12.6501       12.4788       10.4014        9.9802   
DPM1         10.0059       11.3611       10.4094       10.7109       10.2990   
SCYL3         9.2739        9.8320        9.0703       10.4094        9.2132   
FIRRM         8.4134        9.6792        9.4323        9.0489        8.3288   
FGR           9.9545        9.8101        8.9173       10.3324       12.3838   

        TCGA-97-8547  TCGA-97-7554  TCGA-50-6673  TCGA-95-7039  TCGA-95-7944  \
GENES                                                                          
TSPAN6       11.2009       11.4114       11.7631       11.2826       10.2483   
DPM1          9.6637       10.2146       10.3319       10.9650       11.0105   
SCYL3         8.9828        9.2478        9.0566        9.0567        9.1049   
FIRRM         7.7048        8.5637        8.3277        8.5330        8.9807   
FGR           9.6721        9.5631        9.7512        8.7838       10.0118   

        TCGA-55-8301  TCGA-78-7152  TCGA-05-4390  TCGA-44-7659  TCGA-97-7941  \
GENES                                                                          
TSPAN6       10.8454       11.3710       11.8123       10.4946       12.8853   
DPM1         10.4496        9.5375       11.0195       10.4263       10.0014   
SCYL3        10.0076        9.5664        9.2488       10.0838        9.7055   
FIRRM         8.9977        8.2473        9.6170        7.6683        7.3489   
FGR          10.7761        9.7616        9.7293        9.5544        9.9237   

        TCGA-49-4514  TCGA-55-A490  TCGA-55-8508  TCGA-MP-A4TE  TCGA-97-A4M1  \
GENES                                                                          
TSPAN6       12.0419        9.7825       11.3508       12.2113       11.2109   
DPM1         11.7075        9.7232       10.4508       10.6472       10.4813   
SCYL3         8.8724        9.1715        9.8818        9.5381        9.6877   
FIRRM         7.9927        7.1645        8.9068        8.3126        7.8546   
FGR           9.0296       10.8014       10.3112        8.6518       10.3076   

        TCGA-75-6206  TCGA-86-8671  TCGA-78-7160  TCGA-64-1681  TCGA-49-4494  \
GENES                                                                          
TSPAN6       10.7897       10.9245       11.1760       12.6706       12.6263   
DPM1         10.2382        9.8675        9.7808       10.7801       10.9638   
SCYL3         9.9796        9.7347        9.7612        9.4238        9.0414   
FIRRM         7.8164        7.7755        9.0864        8.3691        7.6968   
FGR          10.6595       11.3323        9.8601        9.7748        9.4944   

        TCGA-50-5946  TCGA-55-7913  TCGA-44-6779  TCGA-49-AARE  TCGA-05-4403  \
GENES                                                                          
TSPAN6       11.1184       12.0847       12.0260       12.1872       11.1890   
DPM1         10.7165       10.8844        9.6931       10.1148       10.5937   
SCYL3         9.6469       10.0605        9.5776        9.2431        9.6940   
FIRRM         9.3912        9.6854        9.1414        8.3699        7.2528   
FGR           7.5205        8.9309        9.4078        9.6848       11.1939   

        TCGA-99-AA5R  TCGA-73-4659  TCGA-50-8457  TCGA-75-5147  TCGA-55-8507  \
GENES                                                                          
TSPAN6       10.5380       11.1885       10.6572       13.0727       11.4913   
DPM1         10.0443       10.1051        9.8038       10.9006       10.7497   
SCYL3         9.5104        8.7772        9.6834        9.7049        9.5464   
FIRRM         7.3840        8.3109        7.9273        8.9461        9.0077   
FGR          11.7875       10.2860        9.8802        9.7145        9.9715   

        TCGA-78-7536  TCGA-95-A4VK  TCGA-38-4627  TCGA-67-6215  TCGA-69-7973  \
GENES                                                                          
TSPAN6       11.3143       11.0925       13.0532       10.3230       11.2059   
DPM1         11.1005        9.6334       10.6470        9.6086       10.4662   
SCYL3        10.1800       10.1731        8.9434       10.1129        9.5120   
FIRRM         9.9350        8.1543        8.1164        8.5358        9.3662   
FGR           9.3146        9.1964        9.8973       10.0729        8.3242   

        TCGA-05-5715  TCGA-75-7030  TCGA-44-2666  TCGA-62-A472  TCGA-55-6985  \
GENES                                                                          
TSPAN6       12.4663       11.5834       10.9066       10.7345       12.4693   
DPM1          9.7965        9.6259       10.6728       10.4062       10.4734   
SCYL3         9.9568        9.3915        9.6556        9.6057        9.2214   
FIRRM         8.7857        7.2239        8.3636        7.7923        8.2715   
FGR          10.7904        9.7788        8.8759        9.5566        9.6591   

        TCGA-J2-A4AG  TCGA-97-A4LX  TCGA-55-6543  TCGA-97-7938  TCGA-35-3615  \
GENES                                                                          
TSPAN6       11.5575       10.9213       12.4820       11.2840       11.5141   
DPM1         10.2548       10.3992       10.7673       10.5584       10.6872   
SCYL3         9.3705        9.4333        9.4012        9.7564        9.8985   
FIRRM         8.4018        8.0629        7.7019        8.0887        8.0258   
FGR           9.9914       10.9376       10.5745        9.0358        8.9110   

        TCGA-44-2657  TCGA-55-8505  TCGA-97-8175  TCGA-49-4510  TCGA-MP-A4TK  \
GENES                                                                          
TSPAN6       10.0187       10.8694       12.1195       11.9295       11.0260   
DPM1         10.1128       10.5001       10.7337       10.6823       10.3774   
SCYL3         9.3117        9.2219        9.1313        9.1250        9.3155   
FIRRM         7.4817        8.4937        8.3959        7.6639        8.4993   
FGR          11.7532        8.7806       10.1075        9.3523       10.6877   

        TCGA-50-5044  TCGA-49-AAR2  TCGA-05-4250  TCGA-97-7937  TCGA-64-5774  \
GENES                                                                          
TSPAN6       10.9569       11.7140       11.9341       11.8186       12.2848   
DPM1         10.5507        9.5723       11.0563       11.3820       10.7984   
SCYL3         8.5723        8.9149        8.8896       10.1626        9.3895   
FIRRM         8.4815        7.8020        9.0679        9.0500        9.1413   
FGR           9.4762        9.5429       10.0884        8.0550        8.0269   

        TCGA-86-8674  TCGA-50-6590  TCGA-64-5779  TCGA-75-5146  TCGA-44-6144  \
GENES                                                                          
TSPAN6       11.6558        9.3353       11.0606       12.0642        9.8124   
DPM1         10.2953       10.4317       10.9132       10.7706       10.3170   
SCYL3        10.0364        8.5380        9.5676       10.0258        9.3249   
FIRRM         8.1932        8.8357        8.7674        8.1115        7.1208   
FGR           7.6216       10.2748        8.8910        9.2807       12.6989   

        TCGA-78-8662  TCGA-55-7911  TCGA-86-8585  TCGA-38-A44F  TCGA-64-1680  \
GENES                                                                          
TSPAN6       11.7632       10.2884       12.5910       11.3659       12.7780   
DPM1          9.8159       10.7112       10.5911       10.5216       11.0934   
SCYL3         9.4584        9.4167        9.1765        9.3333        9.9249   
FIRRM         8.8311        8.8040        8.6506        7.1148        7.9377   
FGR           8.5133        9.2839       11.4400       10.9657        8.1342   

        TCGA-97-A4M6  TCGA-75-6214  TCGA-55-6980  TCGA-97-7553  TCGA-55-A48X  \
GENES                                                                          
TSPAN6       12.6628       11.5601       10.2000       12.0653       11.7465   
DPM1         10.1872       11.4650       10.0845       10.6251       10.2238   
SCYL3        10.2258        9.3517        9.3738        9.4879        9.8991   
FIRRM         8.2413        8.8884        7.3040        8.5518        8.7430   
FGR          10.1091        8.2446       11.4480       11.8178        9.6027   

        TCGA-64-1677  TCGA-73-7498  TCGA-38-6178  TCGA-44-7670  TCGA-62-A471  \
GENES                                                                          
TSPAN6       12.3923       11.4238       11.8239       11.9585       12.2011   
DPM1         11.8974       10.0741       10.2667       11.4990       10.2143   
SCYL3         9.2620       10.0403        9.3082        8.6391        9.4582   
FIRRM         8.8249        7.8675        7.9440        8.9658        9.2448   
FGR           9.3934        9.5944        9.3491        8.3665        8.7138   

        TCGA-49-4505  TCGA-NJ-A7XG  TCGA-55-6981  TCGA-91-6848  TCGA-55-8090  \
GENES                                                                          
TSPAN6       12.1427       12.2740       10.5354       10.6766       11.2521   
DPM1         10.7719       10.4364        9.9946       10.6327        9.9117   
SCYL3         9.2066        9.6627        8.9497        8.4621        8.9667   
FIRRM         6.9924        7.9411        7.4517        8.9327        7.2081   
FGR          10.0874        7.3313       11.4358       10.6241        9.8658   

        TCGA-55-7725  TCGA-55-8207  TCGA-44-6146  TCGA-05-4434  TCGA-55-6979  \
GENES                                                                          
TSPAN6       10.8593       11.5535       10.6908       10.5088       12.1103   
DPM1         10.7585       10.6781       10.8452       10.2105       10.5266   
SCYL3        10.2945        9.5630       10.3852        9.7173        9.4638   
FIRRM         8.2532        8.2563        7.7443        8.7259        9.3832   
FGR          10.1871       10.5421        7.4162       10.8045       10.7947   

        TCGA-05-4427  TCGA-55-8615  TCGA-50-5939  TCGA-05-4418  TCGA-67-6217  \
GENES                                                                          
TSPAN6       10.6678       11.3580       11.2907       12.4034       11.2444   
DPM1         10.7068        9.5639       10.3362       10.1308       10.2051   
SCYL3         8.7810        9.3086        9.1673        9.4087        9.7618   
FIRRM         8.8253        8.0031        8.2955        8.2765        7.8229   
FGR          10.0003        9.5008       10.2772        9.7556        9.2886   

        TCGA-49-6745  TCGA-55-1595  TCGA-49-6742  TCGA-05-4402  TCGA-05-4382  \
GENES                                                                          
TSPAN6       11.5573       10.6814       11.1324       12.5396       10.8427   
DPM1         10.3487       10.4209       10.3560        9.9935       10.3846   
SCYL3         8.9834        9.8248        9.1231        9.4784        8.9424   
FIRRM         8.6316        8.5860        6.9472        8.3785        8.4556   
FGR          10.2222        9.3398       11.4583        9.4793       11.1001   

        TCGA-55-7576  TCGA-67-3773  TCGA-78-7633  TCGA-50-6597  TCGA-44-2659  \
GENES                                                                          
TSPAN6       11.0895       10.6759       11.6387       11.7932       10.2316   
DPM1         10.2926       10.5870       10.2013       10.1682       10.0883   
SCYL3        10.0797        9.7576       10.4031        9.3747        9.6217   
FIRRM         9.1862        7.5927        8.6970        5.8352        7.6451   
FGR           9.5743        9.4919        8.2160        9.1174        9.3648   

        TCGA-95-7947  TCGA-55-7724  TCGA-J2-A4AD  TCGA-55-8091  TCGA-55-1592  \
GENES                                                                          
TSPAN6       11.7782       11.6473       11.4396       11.4026       12.0780   
DPM1         11.1945       10.6828       10.6559       10.3926       10.7314   
SCYL3        10.1896        9.4505        9.9546        9.1939        9.7137   
FIRRM         9.2664        8.0690        8.7478        7.4938        8.7066   
FGR           9.3430       10.2621        8.1118        9.7819       10.5958   

        TCGA-73-4670  TCGA-55-1594  TCGA-55-8621  TCGA-50-5051  TCGA-49-4501  \
GENES                                                                          
TSPAN6       11.4128       13.0523       10.6577       11.9695       12.1888   
DPM1         11.3593       12.3765       10.0253        9.8573       10.7042   
SCYL3         9.3683        9.1750        9.1421        9.4496        9.7904   
FIRRM         9.4710        9.3642        7.7739        8.0841        8.1780   
FGR           9.1164        9.0904       11.5633        8.7876        9.4450   

        TCGA-J2-8194  TCGA-44-8119  TCGA-55-8203  TCGA-97-8177  TCGA-55-7573  \
GENES                                                                          
TSPAN6       12.0151       11.7835       11.2857       11.6833       11.8953   
DPM1          9.9889       10.8183       10.5255       10.8122        9.9752   
SCYL3         9.9654        9.4372        9.9617        9.2584        9.9010   
FIRRM         7.4742        8.6408        8.6898        7.7883        8.4558   
FGR           9.5219        9.4730        9.6956       10.5655        9.3805   

        TCGA-55-8089  TCGA-38-4630  TCGA-78-7166  TCGA-38-4631  TCGA-55-1596  \
GENES                                                                          
TSPAN6       11.2430       12.4815       11.4569       12.1892       12.9238   
DPM1         10.8340       10.7147       10.3240       11.6552       11.4148   
SCYL3         9.5597        9.9597        9.9692        9.3522        9.7765   
FIRRM         9.1530       10.1528        8.5159        9.8086        9.7207   
FGR          11.0442        8.2006        8.7170        9.3075        8.8891   

        TCGA-91-A4BD  TCGA-67-3771  TCGA-J2-8192  TCGA-55-A48Z  TCGA-97-8176  \
GENES                                                                          
TSPAN6       11.3341       10.4122       12.3289       12.5716       11.8607   
DPM1         11.1563       11.1943       10.2782       10.4931       10.3662   
SCYL3         8.7170        9.6048        9.5941        9.7829        9.4945   
FIRRM         7.6168        8.8016        7.8785        8.5502        8.8609   
FGR          10.7038       10.1909       10.0962        9.5766        9.6280   

        TCGA-86-6851  TCGA-50-5931  TCGA-NJ-A4YI  TCGA-97-7552  TCGA-MP-A4T4  \
GENES                                                                          
TSPAN6       10.6942        9.9505        9.8565       11.1532       11.0083   
DPM1         10.2788       10.6498       10.3249       10.2040       10.5631   
SCYL3         9.8494        8.8990        9.6106       10.0654        9.4878   
FIRRM         8.8188        6.1655        8.1194        7.5525        8.4967   
FGR          10.2669       12.1436       10.2015       10.4543       10.6049   

        TCGA-L4-A4E6  TCGA-44-5643  TCGA-MP-A4TD  TCGA-05-4244  TCGA-50-5930  \
GENES                                                                          
TSPAN6        9.7749       11.8172       11.5265       11.7539        9.6730   
DPM1         10.0752       11.5878       10.1247        9.9811       10.3253   
SCYL3         9.3299        8.8287        9.3280        9.8321        9.3744   
FIRRM         7.3293        8.9952        8.4818        9.1144        7.1372   
FGR          12.8714        9.2953        9.3366       10.4120       13.2448   

        TCGA-44-6777  TCGA-05-4430  TCGA-05-4426  TCGA-44-6775  TCGA-05-4420  \
GENES                                                                          
TSPAN6       11.0506       12.0793       12.3549       11.7789       12.2414   
DPM1          9.9141       10.6591       10.3958       10.5550       11.0409   
SCYL3         9.2514        9.0631        9.2725        9.5093        9.1426   
FIRRM         8.1151        8.3783        7.9145        8.7581        8.7234   
FGR          11.4904       10.0490        9.7272       10.4148        9.0093   

        TCGA-55-8506  TCGA-44-7672  TCGA-62-8402  TCGA-86-8056  TCGA-05-4422  \
GENES                                                                          
TSPAN6       11.0330       11.7827       12.6700       10.3898       11.4081   
DPM1         10.5295       10.5735       11.0035       10.3892        9.4086   
SCYL3         9.4783        9.1251        9.4141        8.7055       10.5948   
FIRRM         8.6154        8.4363        8.5036        6.9473        8.7350   
FGR           9.4597       11.0705       10.3704       10.0069        8.9187   

        TCGA-75-7027  TCGA-78-7535  TCGA-50-8460  TCGA-75-7031  TCGA-L9-A444  \
GENES                                                                          
TSPAN6       11.9690       11.8487       11.3940       11.7816       10.4786   
DPM1         10.6769       10.6982       10.5302       10.5096       10.2051   
SCYL3         9.4177        9.2547        9.1221        9.8962        9.6861   
FIRRM         9.5664        7.9311        7.4117        8.5619        7.7587   
FGR           8.5024       11.1808       10.2892       10.0072       10.2733   

        TCGA-73-4666  TCGA-86-7953  TCGA-86-A4D0  TCGA-NJ-A4YQ  TCGA-91-8496  \
GENES                                                                          
TSPAN6       11.3588       11.5056       11.5659       10.9077       12.8481   
DPM1         10.4566        9.7456       10.7550       10.3593       10.7888   
SCYL3         9.6418        9.2168        9.5201       10.1573       11.0186   
FIRRM         9.8614        9.5839        9.6539        8.6016        8.4839   
FGR          10.1902       10.1827        8.2901       10.3673       11.2238   

        TCGA-67-3772  TCGA-55-7281  TCGA-05-4424  TCGA-69-A59K  TCGA-75-7025  \
GENES                                                                          
TSPAN6       13.2597       12.1043       11.8132       10.1705       12.6113   
DPM1         11.0977       10.4838       10.1326       10.8935       10.1556   
SCYL3         9.9912        8.6901        9.4823        9.8922        9.7491   
FIRRM         8.4853        7.8965        8.7199        9.0767        7.6906   
FGR           9.7679       11.1662       10.1624       10.4986        9.4949   

        TCGA-55-8514  TCGA-95-8494  TCGA-05-4395  TCGA-93-A4JQ  TCGA-44-8117  \
GENES                                                                          
TSPAN6       12.7393       10.8212       11.0022        9.9679       10.4334   
DPM1         10.2567       10.6765       11.0303       10.0184       11.2173   
SCYL3         9.9393        8.9300        9.7732        9.4625       10.4722   
FIRRM         7.8826        8.7882        8.3477        8.9140        9.6005   
FGR           9.8046       10.5421        8.6035       10.0874        8.8794   

        TCGA-55-8204  TCGA-50-5933  TCGA-MN-A4N1  TCGA-55-7903  TCGA-86-8669  \
GENES                                                                          
TSPAN6       12.5928       10.5347       11.5644       11.3865       11.6430   
DPM1         11.0190       10.6367       10.4657       10.7996       10.3412   
SCYL3         9.2756        8.5539        9.4335        9.9991        9.6167   
FIRRM         8.8265        6.7687        8.6109        9.0711        8.1448   
FGR           9.3127       12.1308        8.7331        9.3541        9.8539   

        TCGA-55-6970  TCGA-50-6594  TCGA-86-8075  TCGA-MP-A4TF  TCGA-69-7760  \
GENES                                                                          
TSPAN6       11.5043       12.6001       12.6972       11.6720       13.9840   
DPM1         10.5155       10.9056       10.4717       11.0922       10.7262   
SCYL3         9.9887        9.3647        9.5085       10.1092        9.5667   
FIRRM         9.3025        9.5287        9.0400        9.1511        8.6888   
FGR           9.6341        9.2426        9.6600        9.0266        7.6209   

        TCGA-78-7161  TCGA-91-6849  TCGA-99-8033  TCGA-55-8205  TCGA-55-8510  \
GENES                                                                          
TSPAN6       12.3083       11.9275       10.5632       10.5139       11.6059   
DPM1         10.4186       10.4778       10.8818       10.6244        9.9601   
SCYL3         9.4444        9.7623       10.0350        8.7668        9.2513   
FIRRM         7.5787        7.6862        9.4371        8.6166        8.6654   
FGR           7.7846       11.4869        9.6619       11.0722       11.3274   

        TCGA-91-6828  TCGA-50-5049  TCGA-99-7458  TCGA-49-AARR  TCGA-MP-A4TC  \
GENES                                                                          
TSPAN6       11.0129       10.9546       11.0942       11.9468       11.2219   
DPM1         10.3842       10.8874        9.8919       10.2086       10.3251   
SCYL3         8.9995        9.6443        9.7327        9.2836        9.1662   
FIRRM         7.1258        8.8318        7.7291        7.0919        8.4898   
FGR          11.8255       11.0363        9.7664       10.6127        9.8380   

        TCGA-49-AARQ  TCGA-93-A4JN  TCGA-95-A4VP  TCGA-69-8453  TCGA-55-8513  \
GENES                                                                          
TSPAN6       11.4285       11.4685       11.9371       10.7517       11.0095   
DPM1         10.6865       10.7666       10.0151       10.1534        9.8252   
SCYL3         9.4191        9.7447        9.1672        9.4019        9.4430   
FIRRM         8.6201        9.6507        8.3449        7.9233        7.3756   
FGR           9.8804        9.4436       10.2768       12.4871       12.3841   

        TCGA-55-6975  TCGA-86-8278  TCGA-75-6203  TCGA-49-6743  TCGA-55-7574  \
GENES                                                                          
TSPAN6        9.9635       12.0445       11.4242       10.2734       10.5699   
DPM1         10.3318       10.4129       10.1555       10.6197       10.1391   
SCYL3         9.4113        9.4753        9.4045        9.4526        9.5222   
FIRRM         7.0956        8.0906        7.5837        9.4268        8.7405   
FGR          11.3146        9.2367       12.1103        9.2876       10.3038   

        TCGA-78-8655  TCGA-05-4389  TCGA-86-7701  TCGA-91-6831  TCGA-55-7728  \
GENES                                                                          
TSPAN6       12.1478       12.6972       11.1324        9.7751       10.5446   
DPM1         10.2610       10.6608       10.4759       10.8768        9.4212   
SCYL3        10.1365        9.6688        9.3138        9.0650        9.9093   
FIRRM         8.4197        9.1549        8.5749        6.7188        7.2544   
FGR          10.1554       10.1882        9.1459       11.8745       13.0418   

        TCGA-97-A4M0  TCGA-50-6593  TCGA-86-6562  TCGA-62-A46S  TCGA-86-A4JF  \
GENES                                                                          
TSPAN6        9.7483       12.8170       11.8853       12.4547       10.2181   
DPM1         10.1342        9.6859       10.7873        9.9960       10.3806   
SCYL3        10.1098        9.0354       10.1722        9.8626        9.6334   
FIRRM         8.5006        8.2515        9.3439        7.7788        9.3662   
FGR          10.0448       10.4839        9.4837        9.6713       10.2333   

        TCGA-J2-A4AE  TCGA-55-8087  TCGA-78-7542  TCGA-78-7148  TCGA-55-A493  \
GENES                                                                          
TSPAN6       11.4485       11.1449       10.6456       11.8871       10.8452   
DPM1         10.5536        9.9261       11.0626       10.3189       10.3556   
SCYL3         9.4093       10.5626        9.3962        9.6598        8.8236   
FIRRM         7.7287        7.9233        9.7096        8.2986        8.7221   
FGR          10.7622        9.1548        8.0891        8.7510       10.8129   

        TCGA-91-A4BC  TCGA-05-4432  TCGA-55-6712  TCGA-4B-A93V  TCGA-93-7347  \
GENES                                                                          
TSPAN6       10.8434       11.1721       11.5949       12.3783       10.1246   
DPM1         10.8261       10.7788       10.4143       10.6831        9.6615   
SCYL3         8.8226        9.9742        9.4260        9.4048        9.1882   
FIRRM         8.4881        9.5706        8.3269        9.1220        7.3841   
FGR           9.4356       10.0938        9.8184        9.2556       10.0784   

        TCGA-86-8668  TCGA-MP-A4TI  TCGA-69-8254  TCGA-64-1676  TCGA-62-A46V  \
GENES                                                                          
TSPAN6       11.0102       10.7886       12.4142       13.2311       11.3050   
DPM1          9.8619        9.6148       10.3166       12.1833       10.8128   
SCYL3         9.2809        9.3073       10.7647        9.6210        9.8062   
FIRRM         7.6842        8.9006        8.1361        8.7213        8.9814   
FGR           9.5976       11.5936        9.9069       10.2541        8.7086   

        TCGA-78-7167  TCGA-55-7284  TCGA-78-7162  TCGA-75-6212  TCGA-97-7546  \
GENES                                                                          
TSPAN6       11.1573       10.8217       11.2405       12.1907       11.6480   
DPM1         10.2940       10.3546       10.1513        9.9935       10.6634   
SCYL3        10.4229       10.1405        9.7423        9.3376        9.6765   
FIRRM         8.3672        7.7418        7.4675        8.0250        8.5535   
FGR           7.8255       11.3981        9.5923        9.6872        9.5713   

        TCGA-44-5644  TCGA-55-8299  TCGA-75-6211  TCGA-MP-A4SW  TCGA-78-7149  \
GENES                                                                          
TSPAN6       10.7273       10.8609       12.3249       11.5895       11.3989   
DPM1         10.6151       10.3034       11.0388        9.6084       10.1049   
SCYL3         9.5753        9.3374        9.5080        9.3468        9.6630   
FIRRM         8.8129        8.5775        9.0771        7.9622        7.5304   
FGR           8.1365       10.1498        7.9491       10.4613        8.3216   

        TCGA-S2-AA1A  TCGA-95-A4VN  TCGA-55-8096  TCGA-05-4398  TCGA-97-8171  \
GENES                                                                          
TSPAN6       12.5339       10.8605       12.1548       11.3018       11.1675   
DPM1         10.3442       10.0398       10.2584       10.6060        9.9462   
SCYL3         9.9180        9.6452        9.6021        9.1445       10.7054   
FIRRM         7.6512        8.9814        8.5059        9.0425        8.6382   
FGR           9.9512       10.6102       10.5580       10.5036        8.2208   

        TCGA-55-A492  TCGA-71-8520  TCGA-44-3396  TCGA-55-A57B  TCGA-L9-A443  \
GENES                                                                          
TSPAN6       11.8632       11.6934       11.9824       11.7301       11.1696   
DPM1         10.6935       10.1466       10.9851        9.7722       10.2135   
SCYL3        10.3428        9.5334        8.8020        9.4802        9.6960   
FIRRM         7.6665        9.1114        8.6390        7.8171        8.0749   
FGR           8.9027        8.7322       11.0152        9.7544        9.0746   

        TCGA-55-A4DG  TCGA-67-4679  TCGA-64-5781  TCGA-93-A4JO  TCGA-NJ-A4YP  \
GENES                                                                          
TSPAN6        9.6553       10.9355       11.4443       11.6764       10.8236   
DPM1         10.4614       10.0556       10.9664       10.8430       10.4458   
SCYL3        10.3752       10.3503        9.6493        9.6399        9.8303   
FIRRM         7.9138        7.9378        8.5424        8.6728        8.9180   
FGR           8.9877        9.9202        8.5830       10.6752        8.9076   

        TCGA-69-7974  TCGA-MP-A4T7  TCGA-55-6642  TCGA-49-AAR9  TCGA-69-8253  \
GENES                                                                          
TSPAN6       10.9707       10.3842       11.2515       11.6288       10.6570   
DPM1         10.4976       10.7279       10.0624       11.0496       11.2922   
SCYL3         9.2408        9.2355        9.4554        9.8303        9.4898   
FIRRM         8.5680        8.4723        8.2765        9.7289        7.1520   
FGR          10.7197       10.5941        8.6956        6.9987        9.4047   

        TCGA-05-4249  TCGA-44-7669  TCGA-71-6725  TCGA-73-7499  TCGA-MP-A4TH  \
GENES                                                                          
TSPAN6       11.5372       11.0214       11.6107       12.2799        9.9885   
DPM1         10.4160       10.8245       10.8518       11.4070        9.4562   
SCYL3        10.1210        9.4890        9.5164        9.9131       10.0144   
FIRRM         8.3952        9.6248        7.8581        9.5527        7.6923   
FGR           9.6932        9.7027        9.7226       10.1130       10.0054   

        TCGA-55-8208  TCGA-78-7156  TCGA-75-5122  TCGA-50-7109  TCGA-NJ-A55R  \
GENES                                                                          
TSPAN6       10.8204       11.5185       11.2179       11.4793       10.5852   
DPM1         10.7687       10.2546       10.4619       11.0633       10.1659   
SCYL3         9.3105       10.5826        8.8982        8.9016        9.7922   
FIRRM         9.2327        7.5736        8.4610        7.8178        7.9795   
FGR          11.5320        8.2472       11.6150        9.4476        8.2822   

        TCGA-53-A4EZ  TCGA-NJ-A4YG  TCGA-86-A456  TCGA-38-4632  TCGA-MP-A4TJ  \
GENES                                                                          
TSPAN6       11.8688       10.9679       11.1557       12.5560       11.0278   
DPM1         10.0678       10.2079       10.5422       11.4943       10.3799   
SCYL3         9.1497        9.8255        9.0385        9.5056        9.3626   
FIRRM         8.2111        8.0822        7.2275        9.7294        8.3773   
FGR           9.3303       10.4898       10.7294       10.2548       10.8896   

        TCGA-97-8174  TCGA-MP-A4SY  TCGA-62-8397  TCGA-L9-A8F4  TCGA-75-5126  \
GENES                                                                          
TSPAN6       11.2249       11.7138       12.0610       11.0386       11.7022   
DPM1          9.8810       10.8335       10.4836       10.4381       11.1644   
SCYL3        10.0143        9.2308        9.8938        9.1274        8.7230   
FIRRM         7.8637        8.7273        7.0051        8.8181        8.4091   
FGR           9.7065        9.3320        9.9604       10.3624       11.3919   

        TCGA-MP-A4TA  TCGA-86-7711  TCGA-50-5045  TCGA-05-4417  TCGA-44-7660  \
GENES                                                                          
TSPAN6       10.7661       11.5559       12.3662       11.6695       12.3629   
DPM1         10.8843       10.8769       10.9079        9.8726       10.8641   
SCYL3         9.3022        9.4158        9.0778        9.3721        9.6511   
FIRRM         9.6044        9.5028        7.9725        7.9351        8.9627   
FGR           9.2426       10.1716       11.0515        9.8812        9.2150   

        TCGA-69-7979  TCGA-55-A491  TCGA-L9-A743  TCGA-55-A494  TCGA-44-7662  \
GENES                                                                          
TSPAN6       10.0493       12.6781       11.1947       10.2288       11.1006   
DPM1         10.8005       10.4343       10.3455        9.4051       10.9179   
SCYL3         9.2463        9.7447        9.0294        9.7439        9.6689   
FIRRM         9.6060        9.2114        8.3101        8.3419       10.1380   
FGR           9.3057       10.0473       10.6724        7.2922        9.7134   

        TCGA-55-7910  TCGA-69-7980  TCGA-55-8097  TCGA-73-4668  TCGA-NJ-A55A  \
GENES                                                                          
TSPAN6       10.7090       11.5433       10.6578       12.2400       12.2886   
DPM1         11.6558       10.2703        9.5834       11.0609       10.2996   
SCYL3         9.4864        9.4727        9.5110        8.8101        9.7125   
FIRRM         9.0545        8.7019        7.3637        9.0045        7.8968   
FGR           9.0199        9.8077        9.4394        9.4059        9.3254   

        TCGA-86-8672  TCGA-86-8673  TCGA-78-7153  TCGA-97-A4M7  TCGA-62-A46R  \
GENES                                                                          
TSPAN6       11.0114       11.1331       12.7791       11.4336       11.4533   
DPM1         10.9338       10.2698       10.3567        9.8932       10.7229   
SCYL3         8.7629        8.2262        9.4804        9.2713        9.7087   
FIRRM         8.2508        8.2384        8.0894        8.1065        8.5633   
FGR          10.4796        9.4242        8.0340       10.4030       10.2213   

        TCGA-50-5055  TCGA-38-4628  TCGA-86-7713  TCGA-86-8073  TCGA-MN-A4N4  \
GENES                                                                          
TSPAN6       11.8079       11.3954       12.2414       10.7676       11.6560   
DPM1         10.7633       11.0195       10.5073       10.5503       11.0739   
SCYL3         9.3311        9.7850       10.4121        9.0277        9.5578   
FIRRM         7.7298        9.3048       10.0026        8.2811        8.7883   
FGR          10.3702        8.7169        8.1769        9.9471        8.9532   

        TCGA-53-7626  TCGA-44-A47G  TCGA-55-6969  
GENES                                             
TSPAN6       10.8912       11.3200       10.5764  
DPM1         10.1742        9.9137       10.9022  
SCYL3         9.6307        9.3312        9.3425  
FIRRM         8.0179        7.8703        8.8147  
FGR          10.6328       11.3029        9.6115  
</pre></div>
</div>
</div>
</div>
</section>
<section id="assign-classes-based-on-phenotypic-attributes-in-metadata-e-g-smoking-status">
<h3>Assign classes based on phenotypic attributes in metadata (e.g., smoking status)<a class="headerlink" href="#assign-classes-based-on-phenotypic-attributes-in-metadata-e-g-smoking-status" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a dictionary to store the class assignments for each sample</span>
<span class="n">classes</span> <span class="o">=</span> <span class="p">{}</span>

<span class="c1"># Iterate through the columns (samples)</span>
<span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">G_gxp_nodes_list</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span> <span class="c1"># &quot;.loc[G_gxp_nodes_list]&quot; grabs rows of specified gene symbols</span>
    <span class="c1"># Assign class labels based on the smoking status from the metadata</span>
    <span class="k">if</span> <span class="n">tcga_gxp_meta</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">sample</span><span class="p">,</span> <span class="s1">&#39;Smoked&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Smoker&#39;</span><span class="p">:</span> <span class="c1">### YOUR CODE HERE ###</span>
        <span class="n">classes</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;smoker&#39;</span> 
    <span class="k">elif</span> <span class="n">tcga_gxp_meta</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">sample</span><span class="p">,</span> <span class="s1">&#39;Smoked&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Never&#39;</span><span class="p">:</span>
        <span class="n">classes</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;control&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">pass</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Perform GSEA using the prepared data and class assignments</span>
<span class="n">gs_res</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">gsea</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">G_gxp_nodes_list</span><span class="p">],</span> <span class="n">gene_sets</span><span class="o">=</span><span class="n">gene_sets</span><span class="p">,</span> <span class="bp">cls</span><span class="o">=</span><span class="n">classes</span><span class="p">,</span> <span class="n">permutation_num</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;signal_to_noise&#39;</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>

<span class="c1"># Display the top results from the GSEA</span>
<span class="c1">### YOUR CODE HERE ###</span>
<span class="n">gs_res</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Name</th>
      <th>Term</th>
      <th>ES</th>
      <th>NES</th>
      <th>NOM p-val</th>
      <th>FDR q-val</th>
      <th>FWER p-val</th>
      <th>Tag %</th>
      <th>Gene %</th>
      <th>Lead_genes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>gsea</td>
      <td>Toxoplasmosis</td>
      <td>0.672196</td>
      <td>1.87385</td>
      <td>0.017544</td>
      <td>0.131839</td>
      <td>0.05</td>
      <td>11/19</td>
      <td>14.33%</td>
      <td>HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;ALOX5;HLA-DOA</td>
    </tr>
    <tr>
      <th>1</th>
      <td>gsea</td>
      <td>Influenza A</td>
      <td>0.708245</td>
      <td>1.7236</td>
      <td>0.117647</td>
      <td>0.477917</td>
      <td>0.11</td>
      <td>11/17</td>
      <td>14.33%</td>
      <td>HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;CASP1;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA</td>
    </tr>
    <tr>
      <th>2</th>
      <td>gsea</td>
      <td>Th1 and Th2 cell differentiation</td>
      <td>0.735164</td>
      <td>1.659187</td>
      <td>0.096154</td>
      <td>0.399637</td>
      <td>0.18</td>
      <td>10/16</td>
      <td>14.33%</td>
      <td>HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA</td>
    </tr>
    <tr>
      <th>3</th>
      <td>gsea</td>
      <td>Th17 cell differentiation</td>
      <td>0.735164</td>
      <td>1.659187</td>
      <td>0.096154</td>
      <td>0.399637</td>
      <td>0.18</td>
      <td>10/16</td>
      <td>14.33%</td>
      <td>HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA</td>
    </tr>
    <tr>
      <th>4</th>
      <td>gsea</td>
      <td>Human T-cell leukemia virus 1 infection</td>
      <td>0.645243</td>
      <td>1.649653</td>
      <td>0.083333</td>
      <td>0.323006</td>
      <td>0.18</td>
      <td>13/22</td>
      <td>18.67%</td>
      <td>HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;SPI1;HLA-DOA;ITGB2;HLA-F</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="visualising-gsea-results">
<h3>Visualising GSEA Results<a class="headerlink" href="#visualising-gsea-results" title="Link to this heading">#</a></h3>
<p>Once you have performed GSEA, the next step is to visualise the results. Visualisation helps in interpreting the biological significance of the enriched gene sets. Here, we visualise GSEA results with Barcode Enrichment Plot, Heatmap, Clustermap, and Dot Plot.</p>
</section>
</section>
<section id="barcode-enrichment-plot">
<h2>Barcode Enrichment Plot<a class="headerlink" href="#barcode-enrichment-plot" title="Link to this heading">#</a></h2>
<p>Barcode Enrichment Plot shows the positions of members of a given gene set in a ranked list of enrichment scores for the top enriched terms. The scores are ranked left to right from smallest to largest. The ranked scores are represented by a shaded bar, forming a pattern like a barcode.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Extract the enriched terms from the GSEA results. The terms represent pathways or functional categories that are significantly enriched in the dataset.</span>
<span class="n">terms</span> <span class="o">=</span> <span class="n">gs_res</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">Term</span>

<span class="c1"># Plot the top 5 enriched terms</span>
<span class="c1"># The plot function visualizes the enrichment results</span>
<span class="n">axs</span> <span class="o">=</span> <span class="n">gs_res</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">terms</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">show_ranking</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">legend_kws</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;loc&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">0</span><span class="p">)})</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/68cae591387adc9f09777ffc9132322c6c4e2ba2f0b9b12ea9dae42b888f8162.png" src="../_images/68cae591387adc9f09777ffc9132322c6c4e2ba2f0b9b12ea9dae42b888f8162.png" />
</div>
</div>
<p>You can view and extract leading-edge genes from GSEA results. Leading-edge genes are the subset of genes that contribute most to the enrichment score.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># View leading-edge genes from the GSEA results. </span>
<span class="n">gs_res</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">Lead_genes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0                                                          HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;ALOX5;HLA-DOA
1                                                          HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;CASP1;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA
2                                                                HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA
3                                                                HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA
4                                               HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;SPI1;HLA-DOA;ITGB2;HLA-F
5                                                HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;NCF4;NCF2;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA;ITGB2
6         HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;NCF4;NCF2;HLA-DQB1;HLA-DQA1;HLA-DRB5;MSR1;HLA-DOA;MARCO;ITGB2;HLA-F;MRC1;HLA-B;ITGAM;TLR4
7                                   HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA;ITGB2;HLA-F;LAMA2;HLA-B;CD86
8                                           HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA;ITGB2;ACP5;TLR4;CD86
9                                                      HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;BTK;HLA-DRB5;HLA-DOA;HLA-F
10                                                                       HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5
11                                     HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA;CD37;CSF1R;ITGAM;CD4;CD34
12        HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;SELPLG;HLA-DQA1;HLA-DRB5;HLA-DOA;ITGB2;HLA-F;SPN;ITGA8;HLA-B;ITGAM;CD86;CD4;CD34
13                                                    HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;CD74;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA;HLA-F
14                               HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;CD74;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA;ITGB2;MRC1;LSP1;ITGAM;TLR4
15                                                    HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;CD74;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA;HLA-F
16                                                         HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA;HLA-F
17                                                         HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA;HLA-F
18                                                         HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA;HLA-F
19                                                         HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5;HLA-DOA;HLA-F
20                                                                HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;SELPLG;HLA-DQA1;HLA-DRB5
21                                                                                                    NCF4;NCF2;BTK;SPI1;TYROBP;TREM2;ACP5;CSF1R;SIRPA
22                                                                                       CASP1;NCF4;NCF2;SELPLG;ITGB2;ITGAM;TLR4;FPR3;FPR1;CYBB;FCGR2A
23    IL2RB;COL4A1;COL1A1;COL1A2;COL4A2;THBS2;COL6A3;ITGA11;VWF;IL2RG;ITGA4;IL7R;FLT4;PDGFRB;LAMA4;PIK3R5;PIK3AP1;PIK3CG;PDGFRA;TLR4;LAMA2;CSF1R;ITGA8
24                                                      CXCL9;CCL5;CXCL10;IL2RB;CCR5;IL2RG;IL7R;IL10RA;ACVRL1;CCR1;CSF2RB;TNFSF13B;IL33;CD4;IL16;CSF1R
25                                                                                                          HLA-F;PTGER4;ITGA8;LAMA2;HLA-B;HLA-C;HLA-A
26                                     COL4A1;COL1A1;COL1A2;COL4A2;THBS2;COL6A3;ITGA11;PRKCB;VWF;ITGA4;FLT4;PDGFRB;LAMA4;MYLK;PDGFRA;PARVG;LAMA2;ITGA8
27                                                                              STAT1;CXCL10;PRKCB;VWF;F13A1;C1QB;C5AR1;C3AR1;C1QA;C1QC;FCGR2A;CYBB;C7
28                                                                       SPI1;PTGER4;CSF1R;LAMA2;PDGFRA;CSF2RB;MMP2;EDNRB;LAMA4;PDGFRB;FLT4;IL7R;IL2RG
Name: Lead_genes, dtype: object
</pre></div>
</div>
</div>
</div>
</section>
<section id="heatmap-visualisation">
<h2>Heatmap Visualisation<a class="headerlink" href="#heatmap-visualisation" title="Link to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">gseapy</span></code> provides a heatmap function to visualise the expression levels of the leading-edge genes. The heatmap provides a visual representation of how these genes are expressed across different samples in relation to their assigned phenotypic classes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the heatmap function from gseapy</span>
<span class="kn">from</span> <span class="nn">gseapy</span> <span class="kn">import</span> <span class="n">heatmap</span>

<span class="c1"># Select the index of the term to visualize</span>
<span class="n">i</span> <span class="o">=</span> <span class="mi">2</span>

<span class="c1"># Extract the genes contributing to the enrichment of the selected term</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">gs_res</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">Lead_genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>

<span class="c1"># Generate a heatmap of the expression levels of the leading-edge genes</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">heatmap</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">gs_res</span><span class="o">.</span><span class="n">heatmat</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">genes</span><span class="p">],</span> <span class="n">z_score</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">terms</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>

<span class="c1"># Update the x-tick labels with the class labels</span>
<span class="n">xtick_labels</span> <span class="o">=</span> <span class="p">[</span><span class="n">classes</span><span class="p">[</span><span class="n">item</span><span class="o">.</span><span class="n">get_text</span><span class="p">()]</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">()]</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xtick_labels</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[]
</pre></div>
</div>
<img alt="../_images/212444e64e2596e7953ef117166a1425910fcf2c684a00d286e3dbafb48c4635.png" src="../_images/212444e64e2596e7953ef117166a1425910fcf2c684a00d286e3dbafb48c4635.png" />
</div>
</div>
</section>
<section id="clustermap-visualisation">
<h2>Clustermap Visualisation<a class="headerlink" href="#clustermap-visualisation" title="Link to this heading">#</a></h2>
<p>The function <code class="docutils literal notranslate"><span class="pre">clustermap</span></code> from <code class="docutils literal notranslate"><span class="pre">seaborn</span></code> is used to create a clustered heatmap. It not only shows the expression levels of the leading-edge genes but also clusters them based on similarity, providing additional insights into gene expression patterns. The cluster map includes dendrograms, which show the hierarchical clustering of both genes and samples, helping to identify groups of co-expressed genes and similar samples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the clustermap function from seaborn</span>
<span class="kn">from</span> <span class="nn">seaborn</span> <span class="kn">import</span> <span class="n">clustermap</span>

<span class="c1"># Select the index of the term to visualize</span>
<span class="n">i</span> <span class="o">=</span> <span class="mi">2</span>

<span class="c1"># Extract the genes contributing to the enrichment of the selected term</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">gs_res</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">Lead_genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>

<span class="c1"># Extract the relevant subset of the heatmap data</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">gs_res</span><span class="o">.</span><span class="n">heatmat</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">genes</span><span class="p">]</span>

<span class="c1"># Rename the columns based on the class assignments</span>
<span class="n">data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">classes</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Generate a cluster map of the expression levels of the leading-edge genes</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">clustermap</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>  <span class="c1"># The data to cluster</span>
    <span class="n">method</span><span class="o">=</span><span class="s1">&#39;average&#39;</span><span class="p">,</span>  <span class="c1"># Clustering method</span>
    <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span>  <span class="c1"># Distance metric</span>
    <span class="n">z_score</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>  <span class="c1"># Standardize the data</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>  <span class="c1"># Size of the figure</span>
    <span class="n">dendrogram_ratio</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>  <span class="c1"># Ratio of the dendrogram</span>
    <span class="n">colors_ratio</span><span class="o">=</span><span class="mf">0.03</span><span class="p">,</span>  <span class="c1"># Ratio of the colors</span>
    <span class="n">cbar_pos</span><span class="o">=</span><span class="p">(</span><span class="mf">0.02</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">)</span>  <span class="c1"># Position of the color bar</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/717aec685b39aca5709559cafb20fd58e7a4f77a77c12e3e09ee9edd07365a85.png" src="../_images/717aec685b39aca5709559cafb20fd58e7a4f77a77c12e3e09ee9edd07365a85.png" />
</div>
</div>
</section>
<section id="dot-plot-visualisation">
<h2>Dot Plot Visualisation<a class="headerlink" href="#dot-plot-visualisation" title="Link to this heading">#</a></h2>
<p>Use the <code class="docutils literal notranslate"><span class="pre">dotplot</span></code> function in <code class="docutils literal notranslate"><span class="pre">gseapy</span></code> to create a visual representation of the GSEA results. Here we use â€œFDR q-valâ€ to determine the dot sizes, which represents the false discovery rate adjusted p-values. We display normalised enrichment score (NES) value as the x-axis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the dotplot function from gseapy</span>
<span class="kn">from</span> <span class="nn">gseapy</span> <span class="kn">import</span> <span class="n">dotplot</span>

<span class="c1"># Generate the dot plot for the GSEA results</span>
<span class="c1"># The dotplot function visualizes the enrichment results, focusing on the FDR q-values</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">dotplot</span><span class="p">(</span><span class="n">gs_res</span><span class="o">.</span><span class="n">res2d</span><span class="p">,</span>
             <span class="n">column</span><span class="o">=</span><span class="s2">&quot;FDR q-val&quot;</span><span class="p">,</span>  <span class="c1"># Column to be used for dot size</span>
             <span class="n">title</span><span class="o">=</span><span class="s1">&#39;KEGG_2021_Human&#39;</span><span class="p">,</span>  <span class="c1"># Title of the plot</span>
             <span class="n">cmap</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">viridis</span><span class="p">,</span>  <span class="c1"># Color map for the dots</span>
             <span class="n">size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>  <span class="c1"># Size of the dots</span>
             <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>  <span class="c1"># Size of the figure</span>
             <span class="n">cutoff</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># Cutoff for displaying the terms</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/d7b64b5b68296e79af1c1d95d8d4fbe5af3335df9f5c9466cb1e83250ac44853.png" src="../_images/d7b64b5b68296e79af1c1d95d8d4fbe5af3335df9f5c9466cb1e83250ac44853.png" />
</div>
</div>
</section>
<section id="gsea-on-clusters">
<h2>GSEA on Clusters<a class="headerlink" href="#gsea-on-clusters" title="Link to this heading">#</a></h2>
<p>Similarly to ORA, GSEA can also be performed on individual communities after clustering. This allows for a more granular analysis, revealing pathways and functions that are enriched within particular subgroups of the data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define a function to perform GSEA on a given list of genes (community)</span>
<span class="k">def</span> <span class="nf">communityGSEA</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">gene_sets</span><span class="o">=</span><span class="s1">&#39;KEGG_2021_Human&#39;</span><span class="p">,</span> <span class="n">classes</span><span class="o">=</span><span class="n">classes</span><span class="p">):</span>
    <span class="n">gs_res</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">gsea</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">tcga_gxp_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">genes</span><span class="p">],</span> <span class="n">gene_sets</span><span class="o">=</span><span class="n">gene_sets</span><span class="p">,</span> <span class="bp">cls</span><span class="o">=</span><span class="n">classes</span><span class="p">,</span> <span class="n">permutation_num</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;signal_to_noise&#39;</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">min_size</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> 
    <span class="k">return</span> <span class="n">gs_res</span>

<span class="c1"># Perform GSEA for three community</span>
<span class="n">community1_gsea</span> <span class="o">=</span> <span class="n">communityGSEA</span><span class="p">(</span><span class="n">communities</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">community2_gsea</span> <span class="o">=</span> <span class="n">communityGSEA</span><span class="p">(</span><span class="n">communities</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> 
<span class="n">community3_gsea</span> <span class="o">=</span> <span class="n">communityGSEA</span><span class="p">(</span><span class="n">communities</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Display the top results for the first community</span>
<span class="n">community1_gsea</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Name</th>
      <th>Term</th>
      <th>ES</th>
      <th>NES</th>
      <th>NOM p-val</th>
      <th>FDR q-val</th>
      <th>FWER p-val</th>
      <th>Tag %</th>
      <th>Gene %</th>
      <th>Lead_genes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>gsea</td>
      <td>Epstein-Barr virus infection</td>
      <td>0.853817</td>
      <td>1.789511</td>
      <td>0.0</td>
      <td>0.324881</td>
      <td>0.14</td>
      <td>10/11</td>
      <td>16.83%</td>
      <td>HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;HLA-DQA1;BTK;HLA-DRB5</td>
    </tr>
    <tr>
      <th>1</th>
      <td>gsea</td>
      <td>Leishmaniasis</td>
      <td>0.667361</td>
      <td>1.767749</td>
      <td>0.040816</td>
      <td>0.181183</td>
      <td>0.17</td>
      <td>11/18</td>
      <td>16.83%</td>
      <td>HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;NCF4;NCF2;HLA-DQB1;HLA-DQA1;HLA-DRB5</td>
    </tr>
    <tr>
      <th>2</th>
      <td>gsea</td>
      <td>Influenza A</td>
      <td>0.805713</td>
      <td>1.723862</td>
      <td>0.0</td>
      <td>0.191596</td>
      <td>0.2</td>
      <td>10/12</td>
      <td>16.83%</td>
      <td>HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;CASP1;HLA-DPA1;HLA-DQB1;HLA-DQA1;HLA-DRB5</td>
    </tr>
    <tr>
      <th>3</th>
      <td>gsea</td>
      <td>Phagosome</td>
      <td>0.595196</td>
      <td>1.717777</td>
      <td>0.041667</td>
      <td>0.149945</td>
      <td>0.21</td>
      <td>16/22</td>
      <td>34.65%</td>
      <td>HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;NCF4;NCF2;HLA-DQB1;HLA-DQA1;HLA-DRB5;MSR1;HLA-DOA;MARCO;ITGB2;MRC1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>gsea</td>
      <td>Cell adhesion molecules</td>
      <td>0.649813</td>
      <td>1.710744</td>
      <td>0.04</td>
      <td>0.127453</td>
      <td>0.22</td>
      <td>10/16</td>
      <td>16.83%</td>
      <td>HLA-DRB1;HLA-DMA;HLA-DPB1;HLA-DRA;HLA-DMB;HLA-DPA1;HLA-DQB1;SELPLG;HLA-DQA1;HLA-DRB5</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Extract the enriched terms from the GSEA results for the first community</span>
<span class="n">terms</span> <span class="o">=</span> <span class="n">community1_gsea</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">Term</span>

<span class="c1"># Plot the top 5 enriched terms for the first community</span>
<span class="n">axs</span> <span class="o">=</span> <span class="n">community1_gsea</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">terms</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">show_ranking</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">legend_kws</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;loc&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">0</span><span class="p">)})</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/7fc1f16fb1d70904d906e0ded3bd518c36e938a5de872297aed46fcf8d44564b.png" src="../_images/7fc1f16fb1d70904d906e0ded3bd518c36e938a5de872297aed46fcf8d44564b.png" />
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./Section2"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../Section2_intro.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Section 2: Network Analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="Part2.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Part 2: Similarity Network Fusion</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#load-networks-built-in-section-1">0. Load networks built in Section 1</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-network-analysis-with-g-exp">1. Gene Network Analysis with <code class="docutils literal notranslate"><span class="pre">G_exp</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#over-representation-analysis-ora">1.1 Over Representation Analysis (ORA)</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ora-on-gene-correlation-network">1.1.1 ORA on gene correlation network</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ora-on-gene-clusters">1.1.1 ORA on gene clusters</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualising-the-communities">Visualising the Communities</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#perform-ora-on-each-community">Perform ORA on Each Community</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-set-enrichment-analysis-gsea">1.2 Gene Set Enrichment Analysis (GSEA)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#assign-classes-based-on-phenotypic-attributes-in-metadata-e-g-smoking-status">Assign classes based on phenotypic attributes in metadata (e.g., smoking status)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualising-gsea-results">Visualising GSEA Results</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#barcode-enrichment-plot">Barcode Enrichment Plot</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#heatmap-visualisation">Heatmap Visualisation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#clustermap-visualisation">Clustermap Visualisation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dot-plot-visualisation">Dot Plot Visualisation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gsea-on-clusters">GSEA on Clusters</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By https://biomedicalinformaticsgroup.github.io/
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      Â© Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>